---
title: "The Art of Multiple Sequence Alignment in R"
author: "Erik S. Wright"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
---

# Introduction



\setlength{\intextsep}{0pt}
\begin{wrapfigure}{r}{0.32\textwidth}
\includegraphics[width=0.32\textwidth]{AlignmentSpace1}
\caption{ {#f1} The art of multiple sequence alignment.}
\end{wrapfigure}

This document is intended to illustrate the art of multiple sequence alignment in \R{} using **DECIPHER**.  Even though its beauty is often concealed, multiple sequence alignment is a form of art in more ways than one.  Take a look at Figure \@ref(f1) for an illustration of what is happening behind the scenes during multiple sequence alignment.  The practice of sequence alignment is one that requires a degree of skill, and it is that art which this vignette intends to convey.  It is simply not enough to "plug" sequences into a multiple sequence aligner and blindly trust the result.  An appreciation for the art as well a careful consideration of the results are required.

What really is multiple sequence alignment, and is there a single correct alignment?  Generally speaking, alignment seeks to perform the act of taking multiple divergent biological sequences of the same "type" and fitting them to a form that reflects some shared quality.  That quality may be how they look structurally, how they evolved from a common ancestor, or optimization of a mathematical construct.  As with most multiple sequence aligners, **DECIPHER** is "trained" to maximize scoring metrics in order to accomplish a combination of both structural alignment and evolutionary alignment.  The idea is to give the alignment a biological basis even though the molecules that the sequences represent will never meet each other and align under any natural circumstance.

The workhorse for sequence alignment in **DECIPHER** is `AlignProfiles`, which takes in two aligned sets of DNA, RNA, or amino acid (AA) sequences and returns a merged alignment.  For more than two sequences, the function `AlignSeqs` can be used to perform multiple sequence alignment in a progressive/iterative manner on sequences of the same kind.  In this case, multiple alignment works by aligning two sequences, merging with another sequence, merging with another set of sequences, and so-forth until all the sequences are aligned.  This process is iterated to further refine the alignment.  There are other functions that extend use of `AlignSeqs` for different purposes:

1. The first is `AlignTranslation`, which will align DNA/RNA sequences based on their amino acid translation and then reverse translate them back to DNA/RNA.  Aligning protein sequences is more accurate since amino acids are more conserved than their corresponding coding sequence.
2. The second function, `AlignDB`, enables generating alignments from many more sequences than are possible to fit in memory.  Its main purpose is to merge sub-alignments where each alignment alone is composed of many thousands of sequences.  This is accomplished by storing all of the aligned sequences in a database and only working with "profiles" representing the alignment.
3. The function `AdjustAlignment` takes in an existing alignment and shifts groups of gaps right and left to achieve a better alignment.  Its purpose is to eliminate artifacts that accumulate during progressive alignment, and to replace the tedious \& subjective process of manually correcting an alignment.
4. Finally, `StaggerAlignment` will create a "staggered" alignment by separating potentially non-homologous positions into separate columns.  This function will help minimize false homologies when building a phylogenetic tree, although the resulting alignment is not as aesthetically pleasing.
5. The functions `FindSynteny` and `AlignSynteny` can be used in combination to perform pairwise alignment of homologous regions from multiple genomes or non-collinear sequences.  These functions interact with a sequence database containing the genomes, which can each be comprised of multiple sequences (i.e., scaffolds, contigs, or chromosomes).



# Alignment Speed



\begin{wrapfigure}{r}{0.31\textwidth}
\includegraphics[width=0.31\textwidth]{AlignmentSpace2}
\caption{ {#f2} The possible alignment space.}
\end{wrapfigure}

The dynamic programming method used by **DECIPHER** for aligning two profiles requires order `N*M` time and memory space where `N` and `M` are the width of the pattern and subject.  Since multiple sequence alignment is an inherently challenging problem for long sequences, heuristics are employed to maximize speed while maintaining reasonable accuracy.  In this regard, the two control parameters available to the user are `restrict` and `anchor`.  The objective of the `restrict` parameter is to convert the problem from one taking quadratic time to linear time.  The goal of the `anchor` parameter is do the equivalent for memory space so that very long sequences can be efficiently aligned.

The orange diagonal line in Figure \@ref(f2) shows the optimal path for aligning two sequence profiles.  The blue segments to the left and right of the optimal path give the constraint boundaries, which the user controls with the `restrict` parameter.  Areas above and below the upper and lower (respectively) constraint boundaries are neglected from further consideration.  A higher (less negative) value of `restrict[1]` will further constrain the possible "alignment space," which represents all possible alignments between two sequences.  Since the optimal path is not known till completion of the matrix, it is risky to overly constrain the matrix.  This is particularly true in situations where the sequences are not mostly overlapping because the optimal path will likely not be diagonal, causing the path to cross a constraint boundary.  In the non-overlapping case `restrict[1]` could be set below the default to ensure that the entire "alignment space" is available.

Neglecting the "corners" of the alignment space effectively converts a quadratic time problem into a near-linear time problem.  We can see this by comparing `AlignProfiles` with and without restricting the matrix at different sequence lengths.  To extend our comparison we can include the **Biostrings** function `pairwiseAlignment`.  In this simulation, two sequences with 90% identity are aligned and the elapsed time is recorded for a variety of sequence lengths.  As can be seen in Figure \@ref(f3) below, *without* restriction `AlignProfiles` takes quadratic time in the same manner as `pairwiseAlignment`.  However, *with* restriction `AlignProfiles` takes linear time, requiring far less than a microsecond per nucleotide.

\begin{centerfig}
```{r expr0,height=6.8,width=8.5}
library(DECIPHER)
n_points <- 10
N0 <- ceiling(2^seq(5, 13, length.out=n_points))
N1 <- ceiling(2^seq(5, 12, length.out=n_points))
N2 <- ceiling(2^seq(5, 13, length.out=n_points))
N3 <- ceiling(2^seq(5, 16, length.out=n_points))
timings0 <- setNames(rep(0, length(N0)), N0)
timings1 <- setNames(rep(0, length(N1)), N1)
timings2 <- setNames(rep(0, length(N2)), N2)
timings3 <- setNames(rep(0, length(N3)), N3)
for (i in seq_len(length(N0))) {
	for (j in 0:3) {
		N <- eval(parse(text=paste("N", j, sep="")))
		# simulate sequences with 15
		string1 <- DNAStringSet(paste(sample(DNA_ALPHABET[1:4], N[i], replace = TRUE), collapse = ""))
		string2 <- replaceAt(string1,
			at=IRanges(sample(N[i], ceiling(N[i]/5)), width=1),
			sample(c(DNA_ALPHABET[1:4], ""), ceiling(N[i]/5), replace = TRUE))
		# align the sequences using two methods
		if (j==0) {
			timings0[i] <- system.time(pairwiseAlignment(string1, string2))[["user.self"]]
		} else if (j==1) {
			timings1[i] <- system.time(AlignProfiles(string1, string2, restrict=c(-1e10, 1e10, 1e10), anchor=NA))[["user.self"]]
		} else if (j==2) {
			timings2[i] <- system.time(AlignProfiles(string1, string2, anchor=NA))[["user.self"]]
		} else { # j == 3
			timings3[i] <- system.time(AlignProfiles(string1, string2))[["user.self"]]
		}
	}
}

c0 <- lm(timings0 ~ N0 + I(N0^2))
c1 <- lm(timings1 ~ N1 + I(N1^2))
c2 <- lm(timings2 ~ N2)
c3 <- lm(timings3 ~ N3)

N <- seq(1, 46340, length.out=1000) # prediction range
plot(N0, timings0,
	xlab = "Sequence length (nucleotides)",
	ylab = "Elapsed Time (sec.)",
	main = "",
	ylim=c(range(timings0,
		timings1,
		timings2,
		timings3,
		predict(c2, data.frame(N2=46340)))),
	xlim=c(0, max(N3)))
points(N, predict(c0,
		data.frame(N0 = N)),
	type="l", lty=3)
points(N1, timings1,
	col="blue", pch=0)
points(N, predict(c1,
		data.frame(N1 = N)),
	type="l", lty=3, col="blue")
points(N2, timings2,
	col="red", pch=5)
points(N, predict(c2,
		data.frame(N2 = N)),
	type="l", lty=3, col="red")
N <- seq(1, max(N3), length.out=1000) # prediction range
points(N3, timings3,
	col="green", pch=2)
points(N, predict(c3,
		data.frame(N3 = N)),
	type="l", lty=3, col="green")
legend("bottomright",
	c("Biostrings::pairwiseAlignment",
		"AlignProfiles (unrestricted, unanchored)",
		"AlignProfiles (restricted, unanchored)",
		"AlignProfiles (restricted, anchored)"),
	pch=c(1, 0, 5, 2), lty=3,
	col=c("black", "blue", "red", "green"), bg="white")
```
\caption{ {#f3} Global Pairwise Sequence Alignment Timings.}
\end{centerfig}

\medskip

The parameter `anchor` controls the fraction of sequences that must share a common region to anchor the alignment space (Fig. \@ref(f2)).  `AlignProfiles` will search for shared anchor points between the two sequence sets being aligned, and if the fraction shared is above `anchor` (70% by default) then that position is fixed in the "alignment space."  Anchors are 15-mer (for DNA/RNA) or 7-mer (for AA) exact matches between two sequences that must occur in the same order in both sequence profiles.  Anchoring generally does not affect accuracy, but can greatly diminish the amount of memory required for alignment.  In Fig. \@ref(f2), the largest white box represents the maximal memory space required with anchoring, while the entire alignment space (grey plus white areas) would be required without anchoring.  The longest pair of sequence profiles that can be aligned without anchoring is about 46 thousand nucleotides, as shown by the end of the red dotted line in Figure \@ref(f3).  If regularly spaced anchor points are available then the maximum sequence length is greatly extended.  In the vast majority of cases anchoring gives the same result as without anchoring, but with less time and memory space required.


# Alignment Accuracy



Figure \@ref(f4) compares the performance of **DECIPHER** to other sequence alignment software on structural amino acid benchmarks [@Edgar:2010].  All benchmarks have flaws, some of which can easily be found by eye in highly similar sequence sets, and therefore benchmark results should treated with care [@Iantorno:2014].  As can be seen in the figure, the performance of **DECIPHER** is similar to that of other popular alignment software such as MAFFT [@Katoh:2002] and MUSCLE [@Edgar:2004] for smaller benchmarks.  However, **DECIPHER** outperforms other programs on large sequence sets (Fig. \@ref(f5)), and its relative advantage continues increase as more sequences are aligned [@Wright:2015].  Importantly, this is because **DECIPHER** exhibits far less fall-off in accuracy as additional sequences are added.

The accuracy of protein alignment begins to drop-off when sequences in the reference alignment have less than 40% average pairwise identity (Fig. \@ref(f4)).  A similar decline in performance is observed with DNA/RNA sequences, but the drop-off occurs much earlier at around 60% sequence identity.  Therefore, it is generally preferable to align coding sequences by their translation using `AlignTranslation`.  This function first translates the input DNA/RNA sequences, then aligns the translation, and finally (conceptually) reverse translates the amino acid sequences to obtain aligned DNA/RNA sequences.  Nevertheless, even protein alignment cannot be considered reliable when the sequences being aligned differ by more than 70%.

\begin{figure}
\begin{center}
\includegraphics[width=1\textwidth]{AlignmentBenchmarks}
\caption{ {#f4} Performance comparison between different programs for multiple alignment [@Thompson:1994,Edgar:2004,Katoh:2002,Wright:2015] using amino acid structural benchmarks.  The x-axis shows percent identity between sequences in each reference alignment.  The y-axis gives the percentage of correctly aligned residues in the estimated alignment according to the reference alignment (i.e., the Q-score).  The upper-left plot is for the PREFAB (version 4) benchmark [@Edgar:2004].  The upper-right plot shows the results of the BALIBASE (version 3) benchmark [@Thompson:2005].  The lower-left plot is for SABMARK (version 1.65) [@VanWalle:2005].  The lower-right plot gives the results on the OXBENCH alignments [@Raghava:2003].  A comparison of these benchmarks can be found in reference [@Edgar:2010].}
\end{center}
\end{figure}

\begin{figure}
\begin{center}
\includegraphics[width=1\textwidth]{HOMSTRAD-modBenchmarkResults}
\caption{ {#f5} **DECIPHER** offers improved accuracy over other alignment programs ([@Edgar:2004,Katoh:2002,Pei:2007,Sievers:2011,Mirarab:2015,Wright:2015]) on large sets of input sequences.  Average accuracy on the Homstrad-mod benchmark [@Wright:2015] is shown for an increasing number of input sequences, ranging from 2 to 4,000.  All programs display a peak in accuracy at fewer than 500 sequences, but **DECIPHER** exhibits the least drop-off in accuracy as additional input sequences are added.}
\end{center}
\end{figure}

\clearpage


# Recommendations for optimal performance



**DECIPHER** has a number of alignment functions and associated parameters.  The flow-chart in Figure \@ref(f6) is intended to simplify this process for the most frequently encountered multiple sequence alignment tasks.  For more information on any of these suggestions, refer to the examples in the following sections of this vignette.

\begin{figure}
\begin{center}
\includegraphics[height=0.6\textheight,keepaspectratio]{AlignmentFlowChart}
\end{center}
\caption{ {#f6} Flow-chart depicting how to choose the best combination of alignment functions and parameters for the most common multiple sequence alignment problems.}
\end{figure}

\clearpage


# Single Gene Alignment



## Example:  Protein coding sequences


For this example we are going to align the *rplB* coding sequence from many different Bacteria.  The *rplB* gene encodes one of the primary ribosomal RNA binding proteins:  the 50S ribosomal protein L2.  We begin by loading the library and importing the sequences from a FASTA file.  Be sure to change the path names to those on your system by replacing all of the text inside quotes labeled "$<<$path to ...$>>$" with the actual path on your system.

```{r expr1,eval=TRUE,echo=TRUE}
library(DECIPHER)

# specify the path to your sequence file:
fas <- "<<path to FASTA file>>"
# OR find the example sequence file used in this tutorial:
fas <- system.file("extdata", "50S_ribosomal_protein_L2.fas", package="DECIPHER")

dna <- readDNAStringSet(fas)
dna # the unaligned sequences
```

We can align the DNA by either aligning the coding sequences directly, or their translations (amino acid sequences).  Both methods result in an aligned set of DNA sequences, unless the argument `type` is `"AAStringSet"` in `AlignTranslation`.  A quick inspection reveals that the method of translating before alignment yields a more appealing result.  In particular, the reading frame is maintained when aligning the translations.  However, if the dna did not code for a protein then the only option would be to use `AlignSeqs` because the translation would be meaningless.

```{r expr2,eval=FALSE,echo=TRUE}
AA <- AlignTranslation(dna, type="AAStringSet") # align the translation
BrowseSeqs(AA, highlight=1) # view the alignment

DNA <- AlignSeqs(dna) # align the sequences directly without translation
DNA <- AlignTranslation(dna) # align the translation then reverse translate

# write the aligned sequences to a FASTA file
writeXStringSet(DNA, file="<<path to output file>>")
```

Note that frameshift errors can greatly disrupt the alignment of protein coding sequences.  Frameshifts can be corrected by first using `CorrectFrameshifts` on the nucleotide sequences, and then using the corrected sequences as input to `AlignTranslation` with the argument `readingFrame` equal to `1`.

If the input sequences include exact replicates, then alignment can be accelerated by de-replicating the sequences before alignment.  The sequences can then be re-replicated after alignment to create a larger alignment of all the original sequences.  `AlignSeqs` does not automatically handle redundancy in the input sequences, but doing so is fairly straightforward.  In this case there aren't any exact duplicates in the example `dna` sequences.  Nonetheless, the process to de-replicate before alignment and re-replicate after alignment would look like:

```{r expr3,eval=FALSE,echo=TRUE}
u_dna <- unique(dna) # the unique input sequences
index <- match(dna, u_dna) # de-replication index

U_DNA <- AlignSeqs(u_dna) # align the sequences directly without translation
DNA <- U_DNA[index]
names(DNA) <- names(dna) # the re-replicated alignment
```

Also, when aligning nucleotide sequences (or their translations), it may be the case that the sequences are in different orientations.  If so, consider reorienting the sequences so that they all have the same directionality and complementarity by using `OrientNucleotides` prior to alignment.

## Example:  Non-coding RNA sequences


Much like proteins, non-coding RNAs often have a conserved secondary structure that can be used to improve their alignment.  The `PredictDBN` function will predict base pairings from a sequence alignment by calculating the mutual information between pairs of positions.  If RNA sequences are given as input, `AlignSeqs` will automatically use the output of `PredictDBN` to iteratively improve the alignment.  Providing an *RNAStringSet* also causes single-base and double-base substitution matrices to be used, and is preferable to providing a *DNAStringSet* when the sequences are non-coding RNA.  The type of the input sequences can easily be converted to RNA, as shown below.

```{r expr4,eval=FALSE,echo=TRUE}
# database containing 16S ribosomal RNA sequences
db <- system.file("extdata", "Bacteria_175seqs.sqlite", package="DECIPHER")

rna <- SearchDB(db, remove="all", type="RNAStringSet")
# or if starting with DNA sequences, convert to RNA with:
# rna <- RNAStringSet(dna)
# or import RNA sequences directly using:
# rna <- readRNAStringSet("<<path to FASTA file>>")

alignedRNA <- AlignSeqs(rna) # align with RNA secondary structure
```

## Example:  Aligning two aligned sequence sets


It is sometimes useful to align two or more previously-aligned sets of sequences.  Here we can use the function `AlignProfiles` to directly align profiles of the two sequence sets:

```{r expr6,eval=FALSE,echo=TRUE}
half <- floor(length(dna)/2)
dna1 <- dna[1:half] # first half
dna2 <- dna[(half + 1):length(dna)] # second half

AA1 <- AlignTranslation(dna1, type="AAStringSet")
AA2 <- AlignTranslation(dna2, type="AAStringSet")
AA <- AlignProfiles(AA1, AA2) # align two alignments
```

When the two sequence sets are very large it may be impossible to fit both sets of input sequences and the output alignment into memory at once.  The function `AlignDB` can align the sequences in two database tables, or two sets of sequences corresponding to separate identifiers in the same table.  `AlignDB` takes as input two tblNames and/or identifiers, and iteratively builds a profile for each of those respective sequence alignments in the database.  These profiles are aligned, and the insertions are iteratively applied to each of the input sequences until the completed alignment has been stored in add2tbl.

```{r expr7,eval=FALSE,echo=TRUE}
# Align DNA sequences stored in separate tables:
dbConn <- dbConnect(SQLite(), ":memory:")
Seqs2DB(AA1, "DNAStringSet", dbConn, "AA1", tblName="AA1")
Seqs2DB(AA2, "DNAStringSet", dbConn, "AA2", tblName="AA2")
AlignDB(dbConn, tblName=c("AA1", "AA2"), add2tbl="AA",
        type="AAStringSet")
AA <- SearchDB(dbConn, tblName="AA", type="AAStringSet")
BrowseDB(dbConn, tblName="AA")
dbDisconnect(dbConn)
```

The number of sequences required to fit into memory when aligning two sequence sets with `AlignDB` is controlled by the batchSize parameter.  In this way `AlignDB` can be used to align large sequence alignments with only minimal memory required.


# Advanced Options \& Features



## Example:  Building a Guide Tree


The `AlignSeqs` function uses a guide tree to decide the order in which to align pairs of sequence profiles.  The `guideTree` input is a *dendrogram* (tree) object with one leaf per input sequence.  By default this guide tree is generated directly from the input sequences using the order of shared k-mers (i.e., when the argument `guideTree` is `NULL`).  This default guide tree performs very well but requires $\mathcal{O}(n^2)$ time and memory space to construct. Therefore, it may be useful to rely on a chained guide tree when aligning hundreds of thousands of unique sequences.

\begin{figure}
\begin{center}
\includegraphics[width=1\textwidth]{DefaultVsChainedGuideTrees}
\caption{ {#f7} Comparison between the default and chained guide trees when aligning increasing numbers of Cytochrome P450 sequence sets.  The top panel shows average pairwise homology shared with the reference alignment (Q-score) and the bottom panel shows the average fraction of alignment columns that are exactly shared with the reference alignment (TC-score).}
\end{center}
\end{figure}

It has been shown that reasonably accurate alignments of tens of thousands of sequences can be obtained by using a chain guide tree [@Boyce:2014].  With a chained guide tree, sequences are added one-by-one to a growing profile representing all of the aligned sequences.  Figure \@ref(f7) shows the result of using **DECIPHER** to align increasing numbers of Cytochrome P450 sequences (in accordance with the method in reference [@Boyce:2014]), using either a chained guide tree or the default guide tree.  A chained guide tree can be easily generated, as shown below.

```{r expr5,eval=FALSE,echo=TRUE}
# form a chained guide tree
gT <- lapply(order(width(dna), decreasing=TRUE),
	function(x) {
		attr(x, "height") <- 0
		attr(x, "label") <- names(dna)[x]
		attr(x, "members") <- 1L
		attr(x, "leaf") <- TRUE
		x
	})
attr(gT, "height") <- 0.5
attr(gT, "members") <- length(dna)
class(gT) <- "dendrogram"

# use the guide tree as input for alignment
DNA <- AlignTranslation(dna,
	guideTree=gT,
	iterations=0,
	refinements=0)
```

It is also possible to read a Newick formatted tree into \R{} using the function `ReadDendrogram`, and specify this object as the input `guideTree`.

## Example:  Post-processing an existing multiple alignment


There are several steps that can be taken after alignment to verify or improve the alignment.  The most important step is to look at the result to ensure that it meets expectations.  Spurious (unalignable) sequences can then be removed and the alignment process repeated as desired.  The simplest way to view sequences with **DECIPHER** is by using the function `BrowseSeqs`.  The `highlight` parameter controls which sequence, if any, is in focus (highlighted).  A value of zero highlights the consensus sequence as shown below.

```{r expr8,eval=FALSE,echo=TRUE}
BrowseSeqs(DNA, highlight=0)
```

All **DECIPHER** multiple sequence alignments are optimized using `AdjustAlignment` (unless the input argument `FUN` is changed), with the goal of removing artifacts of the progressive alignment process.  This function will efficiently correct most obvious inaccuracies that could be found by-eye.  Therefore, making manual corrections is not recommended unless additional expert knowledge of the sequences is available.  The advantage of using `AdjustAlignment` is that it is a repeatable process that is not subjective, unlike most manual adjustments.  In order to further refine an existing alignment, `AdjustAlignment` can be called directly.

```{r expr9,eval=FALSE,echo=TRUE}
DNA_adjusted <- AdjustAlignment(DNA)
```

It is common to use alignment as a preliminary step before the creation of a phylogenetic tree.  **DECIPHER**, like the majority of alignment programs, attempts to maximize homologous positions between the sequences being aligned.  Such an alignment is particularly useful when investigating which residues are in the same structural position of a protein.  However, disparate sequence regions tend to be concentrated into the same "gappy" areas of the alignment.  When viewed from a phylogenetic perspective these homologies have highly implausible insertion/deletion scenarios.

To mitigate the problem of false homologies, `StaggerAlignment` will automatically generate a staggered version of an existing alignment.  Staggered alignments separate potentially non-homologous regions into separate columns of the alignment.  The result is an alignment that is less visually appealing, but likely more accurate from a phylogenetic perspective.

```{r expr10,eval=FALSE,echo=TRUE}
DNA_staggered <- StaggerAlignment(DNA)
```


# Aligning Homologous Regions of Multiple Genomes



The functions described so far have all required collinear sequences as input.  This requirement is frequently broken by genomes, which may include many sequence rearrangements such as inversion, duplication, and reordering.  `FindSynteny` will find homologous regions between pairs of genomes, which can then be aligned using `AlignSynteny`.  A database of sequences identified by their genome name is used as input to both functions.  This enables the alignment of genomes that are composed of many contigs, so long as they all share the same `identifier` in the database.  The example below uses a database containing five *Influenza virus A* genomes, which are each composed of eight separate segments.

```{r expr11,eval=TRUE,echo=TRUE}
db <- system.file("extdata", "Influenza.sqlite", package="DECIPHER")
synteny <- FindSynteny(db, verbose=FALSE)
synteny # an object of class `Synteny`
InfluenzaA <- AlignSynteny(synteny, db, verbose=FALSE)
unlist(InfluenzaA[[1]])
```

The output is a list, with each list component containing a `DNAStringSetList` of pairwise alignments between two genomes.  Names of the output correspond to their sequence's `identifier` in the database, and the index of the syntenic block.



It is also possible to display the blocks of synteny between all pairs of genomes.  Figure \@ref(f8) shows the different genome segments (i.e., sequences) separated by thin horizontal and vertical lines.  The syntenic blocks are diagonal lines that are composed of many homologous "hits" between the genomes.


```{r expr12,height=8.5,width=8.5}
pairs(synteny, boxBlocks=TRUE) # scatterplot matrix
```

{#f8} Dot plots showing the homologous regions among five *Influenza virus A* genomes.



# Session Information


All of the output in this vignette was produced under the following
conditions:

```{r sessinfo,echo=FALSE}
sessionInfo()
```



