---
title: "Getting Started DECIPHERing"
author: "Erik S. Wright"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Getting Started DECIPHERing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# About DECIPHER

**DECIPHER** is a software toolset that can be used for deciphering and managing biological sequences efficiently using the \R{} statistical programming language.  The program features tools falling into five categories:

* Sequence databases: import, maintain, view, and export a massive number of sequences.
* Sequence alignment: accurately align thousands of DNA, RNA, or amino acid sequences. Quickly find and align the syntenic regions of multiple genomes.
* Oligo design: test oligos in silico, or create new primer and probe sequences optimized for a variety of objectives.
* Manipulate sequences: trim low quality regions, correct frameshifts, reorient nucleotides, determine consensus, or digest with restriction enzymes.
* Analyze sequences: find chimeras, classify into a taxonomy of organisms or functions, detect repeats, predict secondary structure, create phylogenetic trees, and reconstruct ancestral states.
* Gene finding: predict coding and non-coding genes in a genome, extract them from the genome, and export them to a file.

**DECIPHER** is available under the terms of the GNU Public License version 3.


# Design Philosophy



## Curators Protect the Originals


One of the core principles of **DECIPHER** is the idea of the non-destructive workflow.  This revolves around the concept that the original sequence information should never be altered:  sequences are exported looking identical to how they were when they were first imported.  Essentially, the sequence information in the database is thought of as a backup of the original sequence file and no function is able to directly alter the sequence data.  All of the workflows simply add information to the database, which can be used to analyze, organize, and maintain the sequences.  When it comes time to export all or part of the sequences they are preserved in their original state without alteration.

## Don't Reinvent the Wheel


**DECIPHER** makes use of the **Biostrings** package that is a core part of the \href{http://www.bioconductor.org/{Bioconductor suite}}.  This package contains numerous functions for common operations such as searching, manipulating, and reverse complementing sequences.  Furthermore, **DECIPHER** makes use of the **Biostrings** interface for handling sequence data so that sequences are stored in `XStringSet` objects.  These objects are compatible with many useful packages in the Bioconductor suite.

A wide variety of user objectives necessitates that **DECIPHER** be extensible to customized projects.  \R{} provides a simple way to place the power of thousands of packages at your fingertips.  Likewise, \R{} enables direct access to the speed and efficiency of the programming language \C{} while maintaining the utility of a scripting language.  Therefore, minimal code is required to solve complex new problems.  Best of all, the \R{} statistical programming language is open source, and maintains a thriving user community so that direct collaboration with other \R{} users is available on \href{https://stat.ethz.ch/mailman/listinfo{several Internet forums}}.

## That Which is the Most Difficult, Make Fastest


A core objective of **DECIPHER** is to make massive tasks feasible in minimal time.  To this end, many of the most time consuming functions are parallelized to make use of multiple processors.  For example, the function `DistanceMatrix` gets almost a 1x speed boost for each processor core.  A modern processor with 8 cores can see a factor of close to eight times speed improvement.  Similar speedups can be achieved in many other **DECIPHER** functions by setting the `processors` argument.  This is all made possible through the use of OpenMP in \C{}-level code.

Other time consuming tasks are handled efficiently.  The function `FindChimeras` can uncover sequence chimeras by searching through a reference database of over a million sequences for thousands of 30-mer fragments in a number of minutes.  This incredible feat is accomplished by using the *PDict* class provided by **Biostrings**.  Similarly, the `SearchDB` function can obtain the one-in-a-million sequences that match a targeted query in a matter of seconds.  Such high-speed functions enable the user to find solutions to problems that previously would have been extremely difficult or nearly impossible to solve using antiquated methods.

## Stay Organized


It is no longer necessary to store related data in several different files.  **DECIPHER** is enabled by **RSQLite**, which is an \R{} interface to \href{http://www.sqlite.org/{SQLite databases}}.  **DECIPHER** creates an organized collection of sequences and their associated information known as a sequence database.  SQLite databases are flat files, meaning they can be handled just like any other file.  There is no setup required since SQLite does not require a server, unlike many other database engines.  These attributes of SQLite databases make storing, backing-up, and sharing sequence databases relatively straightforward.

Separate projects can be stored in distinct tables in the same sequence database.  Each new table is structured to include every sequence's description, identifier, and a unique key called the row_name all in one place.  The sequences are referenced by their row_names or identifier throughout most functions in the package.  Using row_names, new information created with **DECIPHER** functions can be added as additional database columns to their respective sequences' rows in the database table.  To prevent the database from seeming like a black box there is a function named `BrowseDB` that facilitates viewing of the database contents in a web browser.  A similar function is available to view sequences called `BrowseSeqs`.

The amount of DNA sequence information available is currently increasing at a phenomenal rate.  **DECIPHER** stores individual sequences using a custom compression format, called nbit, so that the database file takes up much less drive space than a standard text file of sequences.  The compressed sequences are stored in a hidden table that is linked to the main information table that the user interacts with regularly.  For example, by default sequence information is stored in the table "Seqs", and the associated sequences are stored in the table "_Seqs".  Storing the sequences in a separate table greatly improves access speed when there is a large amount of sequence information.  Separating projects into distinct tables further increases query speed over that of storing every project in a single table.


# Functionality



The functions of DECIPHER can be grouped into several categories based on intended use:

1. Primary functions for interacting with a sequence database:
    a. `Add2DB`    a. `DB2Seqs`    a. `SearchDB`    a. `Seqs2DB`
2. Secondary functions for typical database tasks:
    a. `IdentifyByRank`    a. `IdLengths`
2. Functions related to forming consensus:
    a. `ConsensusSequence`    a. `Disambiguate`    a. `IdConsensus`
2. Phylogenetics and sequence comparison:
    a. `DistanceMatrix`    a. `Clusterize`    a. `MaskAlignment`    a. `ReadDendrogram`    a. `StaggerAlignment`    a. `TreeLine`    a. `WriteDendrogram`
2. Visualization with a web browser:
    a. `BrowseDB`    a. `BrowseSeqs`
2. Manipulating sequences:
    a. `CorrectFrameshifts`    a. `OrientNucleotides`    a. `RemoveGaps`    a. `TrimDNA`
2. Analyzing sequences:
    a. `DetectRepeats`    a. `DigestDNA`    a. `PredictDBN`    a. `PredictHEC`
2. Multiple sequence alignment:
    a. `AdjustAlignment`    a. `AlignDB`    a. `AlignProfiles`    a. `AlignSeqs`    a. `AlignTranslation`    a. `ScoreAlignment`
2. Comparison of sequences that are not collinear (i.e., genomes):
    a. `AlignSynteny`    a. `FindSynteny`    a. *Synteny-class*
2. Functions related to chimeras (PCR artifacts):
    a. `CreateChimeras`    a. `FindChimeras`    a. `FormGroups`
2. Functions related to DNA microarrays:
    a. `Array2Matrix`    a. `CalculateEfficiencyArray`    a. `DesignArray`    a. `NNLS`
2. Functions related to probes for fluorescence *in situ* hybridization (FISH):
    a. `CalculateEfficiencyFISH`    a. `DesignProbes`    a. `TileSeqs`
2. Functions related to primers for polymerase chain reaction (PCR):
    a. `AmplifyDNA`    a. `CalculateEfficiencyPCR`    a. `DesignPrimers`    a. `DesignSignatures`    a. `MeltDNA`
2. Classifying sequences into a taxonomy:
    a. `LearnTaxa`    a. `IdTaxa`    a. *Taxa-class*
2. Gene calling (a.k.a. gene finding):
    a. `ExtractGenes`    a. `FindGenes`    a. `FindNonCoding`    a. *Genes-class*    a. `LearnNonCoding`    a. *NonCoding-class*    a. `WriteGenes`



# Installation

## Typical Installation (recommended)

1. Install the latest version of \R{} from <http://www.r-project.org/>.
2. Install **DECIPHER** in R by entering:

```
> if (!requireNamespace("BiocManager", quietly=TRUE))
+     install.packages("BiocManager")
> BiocManager::install("DECIPHER")
```

## Manual Installation

### All platforms

1. Install the latest \R{} version from <http://www.r-project.org/>.
2. Install **Biostrings** in R by entering:

```
> if (!requireNamespace("BiocManager", quietly=TRUE))
+     install.packages("BiocManager")
> BiocManager::install("Biostrings")
```

3. Install **RSQLite** in R by entering:

```
> install.packages("RSQLite")
```

4. Download **DECIPHER** from <http://DECIPHER.codes>.


### Mac OS X



1. First install Command Line Tools from Apple, which contains compliers that are required to build packages on the Mac.  Then, in \R{} run:

```
> install.packages("<<path to Mac OS X DECIPHER.tgz>>", repos=NULL)
```



For parallelization on Mac OS X, extra steps are required to <http://mac.r-project.org/openmp/{enable OpenMP>} before install.  In summary, run "clang -v" and then download the corresponding LLVM tar.gz file.  Run the sudo tar command as shown, and then add these two lines to your $\sim$/.R/Makevars file:

CPPFLAGS += -Xclang -fopenmp\\
LDFLAGS += -lomp

Then **DECIPHER** can be built and installed the usual way via the command line.

### Linux

In a shell enter:

```
R CMD build --no-build-vignettes "<<path to DECIPHER source>>"
R CMD INSTALL "<<path to newly built DECIPHER.tar.gz>>"
```


### Windows

Two options are available:  the first is simplest, but requires the pre-built binary (DECIPHER.zip).


1. First Option:

```
> install.packages("<<path to Windows DECIPHER.zip>>", repos=NULL)
```

2. Second Option (more difficult):
[(a)]
    a. Install Rtools from <http://cran.r-project.org/bin/windows/Rtools/>.   Be sure to check the box that says edit PATH during installation.
    b. Open a MS-DOS command prompt by clicking Start -> All Programs -> Accessories -> Command Prompt.
    c. In the command prompt enter:

```
R CMD build --no-build-vignettes "<<path to DECIPHER source>>"
R CMD INSTALL "<<path to newly built DECIPHER.zip>>"
```


# Example Workflow


To get started we need to load the **DECIPHER** package, which automatically loads several other required packages:

```{r startup}
library(DECIPHER)
```

Help for any function can be accessed through a command such as:


```
> ? DECIPHER
```


To begin, we can import a FASTA, FASTQ, or GenBank file into a sequence database.  Here we will import the sequences into an in-memory database that will be removed when we disconnect from the database.  In-memory databases are useful for temporary examples, but typically we would specify the path to a file where we want to store the database.  This is especially the case when there are many sequences, as they might not all fit into memory.

In this example, we will use a GenBank file that is included with **DECIPHER** installation.  We must set the sequences' identifier when importing with `Seqs2DB`.  Here we will identify the sequences with the word "Bacteria".  The identifier is used by many **DECIPHER** functions to reference a specific subset of sequences in the database, and can be reset at a later time using a number of different methods:

```{r expr1}
# access a sequence file included in the package:
gen <- system.file("extdata", "Bacteria_175seqs.gen", package="DECIPHER")

# connect to a database:
dbConn <- dbConnect(SQLite(), ":memory:")

# import the sequences into the sequence database
Seqs2DB(gen, "GenBank", dbConn, "Bacteria")
```

Now we can view the table of information we just added to the database in a web browser (Fig. 1):

```{r expr2}
BrowseDB(dbConn)
```

Suppose we wanted to count the number of bases in each sequence and add that information to the database:

```{r expr3}
l <- IdLengths(dbConn)
head(l)
Add2DB(l, dbConn, verbose=FALSE)
BrowseDB(dbConn, maxChars=20)
```

Next let's identify our sequences by phylum and update this information in the database:

```{r expr4}
r <- IdentifyByRank(dbConn, level=3, add2tbl=TRUE)
BrowseDB(dbConn, maxChars=20)
```

\begin{figure}
\begin{center}
\includegraphics[width=1\textwidth]{BrowseDBOutput}
\caption{ {#f1} Database table shown in web browser}
\end{center}
\end{figure}


We can now look at only those sequences that belong to the phylum Bacteroidetes (Fig. 2):

```{r expr5}
dna <- SearchDB(dbConn, identifier="Bacteroidetes")
BrowseSeqs(subseq(dna, 140, 240))
```

\begin{figure}
\begin{center}
\includegraphics[width=1\textwidth]{BrowseSeqsOutput}
\caption{ {#f2} Sequences shown in web browser}
\end{center}
\end{figure}


Let's construct a phylogenetic tree from the Bacteroidetes sequences (Fig. 3):

\begin{centerfig}
```{r expr6,fig=TRUE,height=3.5,width=6}
d <- DistanceMatrix(dna, correction="Jukes-Cantor", verbose=FALSE)
c <- TreeLine(myDistMatrix=d, method="NJ", cutoff=0.05, showPlot=TRUE, verbose=FALSE)
```
\caption{Neighbor-joining tree showing the relationships between sequences.}
\end{centerfig}



Here we created the database in-memory by connecting to ":memory:" above.  This works fine for small databases, but it is necessary to initialize larger databases with a quoted file path rather than ":memory:".  Optionally, we could use the command below to save an in-memory database to a file for long term storage.  Be sure to change the path names to those on your system by replacing all of the text inside quotes labeled "$<<$path to ...$>>$" with the actual path on your system.


```
> sqliteCopyDatabase(dbConn, "<<path to database>>")
```


Finally, we should disconnect from the database connection.  Since the sequence database was created in temporary memory, all of the information will be erased:

```{r expr7}
dbDisconnect(dbConn)
```


