---
title: "Growing phylogenetic trees with TreeLine"
author: "Erik S. Wright"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
---


# Introduction

This document describes how to grow phylogenetic trees using the `TreeLine` function in the **DECIPHER** package.  `TreeLine` takes as input a set of aligned nucleotide or amino acid sequences and returns a phylogenetic tree (i.e., *dendrogram* object) as output.  This vignette focuses on building maximum likelihood (ML) and maximum parsimony (MP) phylogenetic trees starting from sequences, but `TreeLine` can also be used to build additive trees from a distance matrix.

Why is the function called `TreeLine`?  The goal of `TreeLine` is to find the most likely/parsimonious tree for a given sequence alignment.  There are often many trees with nearly maximal likelihood/parsimony.  Therefore, `TreeLine` seeks to find a tree as close as possible to the treeline, analogous to how no trees can grow above the treeline on a mountain.

Why use `TreeLine` versus other programs?  The `TreeLine` function is designed to return an excellent phylogenetic tree with minimal user intervention.  Many tree building programs have a large set of complex options for niche applications.  In contrast, `TreeLine` simply builds a great tree when relying on its defaults.  This vignette is intended to get you started and introduce additional options/functions that might be useful.

TreeLine uses multi-start optimization followed by hill-climbing to find the highest trees on the likelihood or parsimony landscapes.  Since TreeLine is a stochastic optimizer, it optimizes many trees to prevent luck from influencing the final result.  The main difference from most other approaches to tree optimization is that TreeLine heavily uses past trees to generate new trees as it optimizes.  With any luck it'll find the treeline!


# Performance Considerations

Finding a tree with very high likelihood/parsimony is no easy feat.  `TreeLine` systematically optimizes hundreds to thousands of candidate trees before returning the best one.  This takes time, but there are things you can do to make it go faster.


* Only use the sequences you need:  `TreeLine` scales a bit worse than quadratically with the number of sequences.  Hence, limiting the number of sequences is a worthwhile consideration.  In particular, always eliminate redundant sequences, as shown below, and remove any sequences that are not necessary.  This concern is shared for all tree building programs, and `TreeLine` is no exception.
* Set a timeout:  The `maxTime` argument specifies the (approximate) maximum number of hours you are willing to let `TreeLine` run.  If you are concerned about the code running too long then simply specify this argument.
* Compile with OpenMP support:  Significant speed-ups can be achieved with multi-threading using OpenMP.  See the "Getting Started DECIPHERing" vignette for how to do this on your computer platform.  Then you only need to set the argument `processors=NULL` and `TreeLine` will use all available processors.


# Growing a Phylogenetic Tree

`TreeLine` takes as input a multiple sequence alignment when constructing a maximum likelihood or maximum parsimony phylogenetic tree.  Multiple sequence alignments can be constructed from a set of (unaligned) sequences using `AlignSeqs` or related functions.  `TreeLine` will optimize trees for amino acid (i.e., `AAStringSet`) or nucleotide (i.e., `DNAStringSet` or `RNAStringSet`) sequences.  Here, we are going to use a set of sequences that is included with **DECIPHER**.  These sequences are from the internal transcribed spacer (ITS) between the 16S and 23S ribosomal RNA genes in several *Streptomyces* species.

```{r expr1}
library(DECIPHER)

# specify the path to your sequence file:
fas <- "<<path to FASTA file>>"
# OR find the example sequence file used in this tutorial:
fas <- system.file("extdata", "Streptomyces_ITS_aligned.fas", package="DECIPHER")

seqs <- readDNAStringSet(fas) # use readAAStringSet for amino acid sequences
seqs # the aligned sequences
```

Many of these sequences are redundant or from the same genome.  We can de-replicate the sequences to accelerate tree building:

```{r expr2}
seqs <- unique(seqs) # remove duplicated sequences

ns <- gsub("^.*Streptomyces( subsp\\. | sp\\. | | sp_)([^ ]+).*$", "\\2", names(seqs))
names(seqs) <- ns # name by species
seqs <- seqs[!duplicated(ns)] # remove redundant sequences from the same species
seqs
```

Now, it's time to try our luck at finding the most likely tree.  Here, we will set a stringent time limit (0.05 hours) to make this example faster, although longer time limits (e.g., 24 hours) are advised because setting very short time limits leaves the result up to luck.

Note that `TreeLine` automatically selects a substitution model based on Akaike information criterion (by default).  It is possible to specify specific model(s) (e.g., `model="GTR+G4"`) to limit the possible selections and test your luck with fewer models.

Also, since `TreeLine` is a stochastic optimizer, it is critical to always set the random number seed for reproducibility.  You can pick any lucky number, and if you every wonder how much you pushed your luck, you can try running again from a different random number seed to see how much the result came down to luck of the draw.

```{r expr3,eval=TRUE,echo=TRUE,fig=TRUE,height=3.5,width=6}
set.seed(123) # set the random number seed

tree <- TreeLine(seqs, reconstruct=TRUE, maxTime=0.05) # default is method="ML"

set.seed(NULL) # reset seed

plot(tree)
```
{#f1} Maximum likelihood tree showing the relationships between *Streptomyces* species.



# Plotting Branch Support Values



Maybe it was just beginner's luck, but we already have a reasonable looking starting tree!  `TreeLine` automatically returns a variety of information about the tree that can be accessed with the `attributes` and `attr` functions:

```{r expr4,eval=TRUE,echo=TRUE}
attributes(tree) # view all attributes
attr(tree, "score") # best score
```

The tree is (virtually) rooted at its midpoint by default.  For maximum likelihood trees, all internal nodes include aBayes branch support values [@Anisimova:2011].  These are given as probabilities that can be used in plotting on top of each edge.  We can also italicize the species names.

\begin{centerfig}
```{r expr5,eval=TRUE,echo=TRUE,fig=TRUE,height=6,width=6}
plot(dendrapply(tree,
	function(x) {
		s <- attr(x, "probability") # choose "probability" (aBayes) or "support"
		if (!is.null(s) && !is.na(s)) {
			s <- formatC(as.numeric(s), digits=2, format="f")
			attr(x, "edgetext") <- paste(s, "\n")
		}
		attr(x, "edgePar") <- list(p.col=NA, p.lwd=1e-5, t.col="#CC55AA", t.cex=0.7)
		if (is.leaf(x))
			attr(x, "nodePar") <- list(lab.font=3, pch=NA)
		x
	}),
	horiz=TRUE,
	yaxt='n')
# add a scale bar
arrows(0, 0, 0.4, 0, code=3, angle=90, len=0.05, xpd=TRUE)
text(0.2, 0, "0.4 subs./site", pos=3, xpd=TRUE)
```
\caption{ {#f2} Tree with (aBayes) support probabilities at each internal node.}
\end{centerfig}

\clearpage

We lucked out because maximum likelihood and maximum parsimony trees both provide branch supports in the form of the fraction of optimized trees that contained a given partition (branch).  These are accessible from the "support" attribute.  As luck would have it, support values and (aBayes) probabilities are correlated, but support tends to be more conservative.

\begin{centerfig}
```{r expr6,eval=TRUE,echo=TRUE,fig=TRUE,height=3.5,width=3.5}
getSupports <- function(x) {
	if (is.leaf(x)) {
		NULL
	} else {
		rbind(cbind(attr(x, "support"), attr(x, "probability")),
			getSupports(x[[1]]), getSupports(x[[2]]))
	}
}
support <- getSupports(tree)
plot(support[, 1], support[, 2], xlab="Support", ylab="aBayes probability", asp=1)
abline(a=0, b=1, lty=2) # line of identity (y=x)
```
\caption{ {#f3} Comparison of aBayes probabilities and branch support values.}
\end{centerfig}

\clearpage


# Ancestral State Reconstruction



We're in luck \textemdash one of the advantages of maximum likelihood and maximum parsimony tree building methods is that they automatically predict states at each internal node on the tree [@Joy:2016].  This feature is enabled when `reconstruct` is set to `TRUE`.  These character states can be used by the function `MapCharacters` to determine state transitions along each edge of the tree.

\begin{centerfig}
```{r expr7,eval=TRUE,echo=TRUE,fig=TRUE,height=6,width=6}
new_tree <- MapCharacters(tree, labelEdges=TRUE)
plot(new_tree, edgePar=list(p.col=NA, p.lwd=1e-5, t.col="#55CC99", t.cex=0.7))
attr(new_tree[[1]], "change") # state changes on first branch left of (virtual) root
```
\caption{ {#f4} Edges labeled with the number of state transitions.}
\end{centerfig}

\clearpage


# Exporting the Tree



We've had a run of good luck with this tree, so we'd better save it before our luck runs out!  The functions `ReadDendrogram` and `WriteDendrogram` will import and export trees in Newick file format.  If we leave the `file` argument blank then it will print the output to the console for our viewing:

```{r expr8,eval=TRUE,echo=TRUE}
WriteDendrogram(tree, file="")
```

To keep up our lucky streak, we should probably include any model parameters in the output along with the tree.  Luckily, Newick format supports square brackets (i.e., "[]") for comments, which we can append to the end of the file for good luck:

```{r expr9,eval=TRUE,echo=TRUE}
params <- attr(tree, "parameters")
cat("[", paste(names(params), params, sep="=", collapse=","), "]", sep="", append=TRUE, file="")
```


# Session Information



All of the output in this vignette was produced under the following
conditions:

```{r sessinfo,echo=FALSE,results=tex}
toLatex(sessionInfo(), locale=FALSE)
```

\begin{thebibliography}{}

\bibitem{Anisimova:2011}
{Anisimova, M.}, {Gil, M.}, {Dufayard, J.}, {Dessimoz, C.}, \& {Gascuel, O.}
\newblock Survey of branch support methods demonstrates accuracy, power, and robustness of fast likelihood-based approximation schemes.
\newblock {\em Syst Biol.}, 60(5), 685-699.

\bibitem{Joy:2016}
{Joy, J.}, {Liang, R.}, {McCloskey, R.}, {Nguyen, T.}, \& {Poon, A.}
\newblock Ancestral Reconstruction.
\newblock {\em PLoS Comp. Biol.}, 12(7), e1004763.

\end{thebibliography}

