<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DECIPHER">
<title>Getting Started DECIPHERing • DECIPHER</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started DECIPHERing">
<meta property="og:description" content="DECIPHER">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting Started DECIPHERing</h1>
                        <h4 data-toc-skip class="author">Erik S.
Wright</h4>
            
            <h4 data-toc-skip class="date">2023-11-15</h4>
      
      
      <div class="d-none name"><code>DECIPHERing.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="about-decipher">About DECIPHER<a class="anchor" aria-label="anchor" href="#about-decipher"></a>
</h2>
<p><strong>DECIPHER</strong> is a software toolset that can be used for
deciphering and managing biological sequences efficiently using the
statistical programming language. The program features tools falling
into five categories:</p>
<ul>
<li>Sequence databases: import, maintain, view, and export a massive
number of sequences.</li>
<li>Sequence alignment: accurately align thousands of DNA, RNA, or amino
acid sequences. Quickly find and align the syntenic regions of multiple
genomes.</li>
<li>Oligo design: test oligos in silico, or create new primer and probe
sequences optimized for a variety of objectives.</li>
<li>Manipulate sequences: trim low quality regions, correct frameshifts,
reorient nucleotides, determine consensus, or digest with restriction
enzymes.</li>
<li>Analyze sequences: find chimeras, classify into a taxonomy of
organisms or functions, detect repeats, predict secondary structure,
create phylogenetic trees, and reconstruct ancestral states.</li>
<li>Gene finding: predict coding and non-coding genes in a genome,
extract them from the genome, and export them to a file.</li>
</ul>
<p><strong>DECIPHER</strong> is available under the terms of the GNU
Public License version 3.</p>
</div>
<div class="section level2">
<h2 id="design-philosophy">Design Philosophy<a class="anchor" aria-label="anchor" href="#design-philosophy"></a>
</h2>
<div class="section level3">
<h3 id="curators-protect-the-originals">Curators Protect the Originals<a class="anchor" aria-label="anchor" href="#curators-protect-the-originals"></a>
</h3>
<p>One of the core principles of <strong>DECIPHER</strong> is the idea
of the non-destructive workflow. This revolves around the concept that
the original sequence information should never be altered: sequences are
exported looking identical to how they were when they were first
imported. Essentially, the sequence information in the database is
thought of as a backup of the original sequence file and no function is
able to directly alter the sequence data. All of the workflows simply
add information to the database, which can be used to analyze, organize,
and maintain the sequences. When it comes time to export all or part of
the sequences they are preserved in their original state without
alteration.</p>
</div>
<div class="section level3">
<h3 id="dont-reinvent-the-wheel">Don’t Reinvent the Wheel<a class="anchor" aria-label="anchor" href="#dont-reinvent-the-wheel"></a>
</h3>
<p><strong>DECIPHER</strong> makes use of the
<strong>Biostrings</strong> package that is a core part of the This
package contains numerous functions for common operations such as
searching, manipulating, and reverse complementing sequences.
Furthermore, <strong>DECIPHER</strong> makes use of the
<strong>Biostrings</strong> interface for handling sequence data so that
sequences are stored in <code>XStringSet</code> objects. These objects
are compatible with many useful packages in the Bioconductor suite.</p>
<p>A wide variety of user objectives necessitates that
<strong>DECIPHER</strong> be extensible to customized projects. provides
a simple way to place the power of thousands of packages at your
fingertips. Likewise, enables direct access to the speed and efficiency
of the programming language while maintaining the utility of a scripting
language. Therefore, minimal code is required to solve complex new
problems. Best of all, the statistical programming language is open
source, and maintains a thriving user community so that direct
collaboration with other users is available on </p>
</div>
<div class="section level3">
<h3 id="that-which-is-the-most-difficult-make-fastest">That Which is the Most Difficult, Make Fastest<a class="anchor" aria-label="anchor" href="#that-which-is-the-most-difficult-make-fastest"></a>
</h3>
<p>A core objective of <strong>DECIPHER</strong> is to make massive
tasks feasible in minimal time. To this end, many of the most time
consuming functions are parallelized to make use of multiple processors.
For example, the function <code>DistanceMatrix</code> gets almost a 1x
speed boost for each processor core. A modern processor with 8 cores can
see a factor of close to eight times speed improvement. Similar speedups
can be achieved in many other <strong>DECIPHER</strong> functions by
setting the <code>processors</code> argument. This is all made possible
through the use of OpenMP in -level code.</p>
<p>Other time consuming tasks are handled efficiently. The function
<code>FindChimeras</code> can uncover sequence chimeras by searching
through a reference database of over a million sequences for thousands
of 30-mer fragments in a number of minutes. This incredible feat is
accomplished by using the <em>PDict</em> class provided by
<strong>Biostrings</strong>. Similarly, the <code>SearchDB</code>
function can obtain the one-in-a-million sequences that match a targeted
query in a matter of seconds. Such high-speed functions enable the user
to find solutions to problems that previously would have been extremely
difficult or nearly impossible to solve using antiquated methods.</p>
</div>
<div class="section level3">
<h3 id="stay-organized">Stay Organized<a class="anchor" aria-label="anchor" href="#stay-organized"></a>
</h3>
<p>It is no longer necessary to store related data in several different
files. <strong>DECIPHER</strong> is enabled by <strong>RSQLite</strong>,
which is an interface to <strong>DECIPHER</strong> creates an organized
collection of sequences and their associated information known as a
sequence database. SQLite databases are flat files, meaning they can be
handled just like any other file. There is no setup required since
SQLite does not require a server, unlike many other database engines.
These attributes of SQLite databases make storing, backing-up, and
sharing sequence databases relatively straightforward.</p>
<p>Separate projects can be stored in distinct tables in the same
sequence database. Each new table is structured to include every
sequence’s description, identifier, and a unique key called the row_name
all in one place. The sequences are referenced by their row_names or
identifier throughout most functions in the package. Using row_names,
new information created with <strong>DECIPHER</strong> functions can be
added as additional database columns to their respective sequences’ rows
in the database table. To prevent the database from seeming like a black
box there is a function named <code>BrowseDB</code> that facilitates
viewing of the database contents in a web browser. A similar function is
available to view sequences called <code>BrowseSeqs</code>.</p>
<p>The amount of DNA sequence information available is currently
increasing at a phenomenal rate. <strong>DECIPHER</strong> stores
individual sequences using a custom compression format, called nbit, so
that the database file takes up much less drive space than a standard
text file of sequences. The compressed sequences are stored in a hidden
table that is linked to the main information table that the user
interacts with regularly. For example, by default sequence information
is stored in the table “Seqs”, and the associated sequences are stored
in the table “_Seqs”. Storing the sequences in a separate table greatly
improves access speed when there is a large amount of sequence
information. Separating projects into distinct tables further increases
query speed over that of storing every project in a single table.</p>
</div>
</div>
<div class="section level2">
<h2 id="functionality">Functionality<a class="anchor" aria-label="anchor" href="#functionality"></a>
</h2>
<p>The functions of DECIPHER can be grouped into several categories
based on intended use:</p>
<ol style="list-style-type: decimal">
<li>Primary functions for interacting with a sequence database:
<ol style="list-style-type: lower-alpha">
<li>
<code>Add2DB</code> a. <code>DB2Seqs</code> a. <code>SearchDB</code>
a. <code>Seqs2DB</code>
</li>
</ol>
</li>
<li>Secondary functions for typical database tasks:
<ol style="list-style-type: lower-alpha">
<li>
<code>IdentifyByRank</code> a. <code>IdLengths</code>
</li>
</ol>
</li>
<li>Functions related to forming consensus:
<ol style="list-style-type: lower-alpha">
<li>
<code>ConsensusSequence</code> a. <code>Disambiguate</code> a.
<code>IdConsensus</code>
</li>
</ol>
</li>
<li>Phylogenetics and sequence comparison:
<ol style="list-style-type: lower-alpha">
<li>
<code>DistanceMatrix</code> a. <code>Clusterize</code> a.
<code>MaskAlignment</code> a. <code>ReadDendrogram</code> a.
<code>StaggerAlignment</code> a. <code>TreeLine</code> a.
<code>WriteDendrogram</code>
</li>
</ol>
</li>
<li>Visualization with a web browser:
<ol style="list-style-type: lower-alpha">
<li>
<code>BrowseDB</code> a. <code>BrowseSeqs</code>
</li>
</ol>
</li>
<li>Manipulating sequences:
<ol style="list-style-type: lower-alpha">
<li>
<code>CorrectFrameshifts</code> a. <code>OrientNucleotides</code> a.
<code>RemoveGaps</code> a. <code>TrimDNA</code>
</li>
</ol>
</li>
<li>Analyzing sequences:
<ol style="list-style-type: lower-alpha">
<li>
<code>DetectRepeats</code> a. <code>DigestDNA</code> a.
<code>PredictDBN</code> a. <code>PredictHEC</code>
</li>
</ol>
</li>
<li>Multiple sequence alignment:
<ol style="list-style-type: lower-alpha">
<li>
<code>AdjustAlignment</code> a. <code>AlignDB</code> a.
<code>AlignProfiles</code> a. <code>AlignSeqs</code> a.
<code>AlignTranslation</code> a. <code>ScoreAlignment</code>
</li>
</ol>
</li>
<li>Comparison of sequences that are not collinear (i.e., genomes):
<ol style="list-style-type: lower-alpha">
<li>
<code>AlignSynteny</code> a. <code>FindSynteny</code> a.
<em>Synteny-class</em>
</li>
</ol>
</li>
<li>Functions related to chimeras (PCR artifacts):
<ol style="list-style-type: lower-alpha">
<li>
<code>CreateChimeras</code> a. <code>FindChimeras</code> a.
<code>FormGroups</code>
</li>
</ol>
</li>
<li>Functions related to DNA microarrays:
<ol style="list-style-type: lower-alpha">
<li>
<code>Array2Matrix</code> a. <code>CalculateEfficiencyArray</code>
a. <code>DesignArray</code> a. <code>NNLS</code>
</li>
</ol>
</li>
<li>Functions related to probes for fluorescence <em>in situ</em>
hybridization (FISH):
<ol style="list-style-type: lower-alpha">
<li>
<code>CalculateEfficiencyFISH</code> a. <code>DesignProbes</code> a.
<code>TileSeqs</code>
</li>
</ol>
</li>
<li>Functions related to primers for polymerase chain reaction (PCR):
<ol style="list-style-type: lower-alpha">
<li>
<code>AmplifyDNA</code> a. <code>CalculateEfficiencyPCR</code> a.
<code>DesignPrimers</code> a. <code>DesignSignatures</code> a.
<code>MeltDNA</code>
</li>
</ol>
</li>
<li>Classifying sequences into a taxonomy:
<ol style="list-style-type: lower-alpha">
<li>
<code>LearnTaxa</code> a. <code>IdTaxa</code> a.
<em>Taxa-class</em>
</li>
</ol>
</li>
<li>Gene calling (a.k.a. gene finding):
<ol style="list-style-type: lower-alpha">
<li>
<code>ExtractGenes</code> a. <code>FindGenes</code> a.
<code>FindNonCoding</code> a. <em>Genes-class</em> a.
<code>LearnNonCoding</code> a. <em>NonCoding-class</em> a.
<code>WriteGenes</code>
</li>
</ol>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="typical-installation-recommended">Typical Installation (recommended)<a class="anchor" aria-label="anchor" href="#typical-installation-recommended"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Install the latest version of from <a href="http://www.r-project.org/" class="external-link uri">http://www.r-project.org/</a>.</li>
<li>Install <strong>DECIPHER</strong> in R by entering:</li>
</ol>
<pre><code>&gt; if (!requireNamespace("BiocManager", quietly=TRUE))
+     install.packages("BiocManager")
&gt; BiocManager::install("DECIPHER")</code></pre>
</div>
<div class="section level3">
<h3 id="manual-installation">Manual Installation<a class="anchor" aria-label="anchor" href="#manual-installation"></a>
</h3>
<div class="section level4">
<h4 id="all-platforms">All platforms<a class="anchor" aria-label="anchor" href="#all-platforms"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Install the latest version from <a href="http://www.r-project.org/" class="external-link uri">http://www.r-project.org/</a>.</li>
<li>Install <strong>Biostrings</strong> in R by entering:</li>
</ol>
<pre><code>&gt; if (!requireNamespace("BiocManager", quietly=TRUE))
+     install.packages("BiocManager")
&gt; BiocManager::install("Biostrings")</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Install <strong>RSQLite</strong> in R by entering:</li>
</ol>
<pre><code>&gt; install.packages("RSQLite")</code></pre>
<ol start="4" style="list-style-type: decimal">
<li>Download <strong>DECIPHER</strong> from <a href="http://DECIPHER.codes" class="external-link uri">http://DECIPHER.codes</a>.</li>
</ol>
</div>
<div class="section level4">
<h4 id="mac-os-x">Mac OS X<a class="anchor" aria-label="anchor" href="#mac-os-x"></a>
</h4>
<ol style="list-style-type: decimal">
<li>First install Command Line Tools from Apple, which contains
compliers that are required to build packages on the Mac. Then, in
run:</li>
</ol>
<pre><code>&gt; install.packages("&lt;&lt;path to Mac OS X DECIPHER.tgz&gt;&gt;", repos=NULL)</code></pre>
<p>For parallelization on Mac OS X, extra steps are required to &lt;<a href="http://mac.r-project.org/openmp/%7Benable" class="external-link uri">http://mac.r-project.org/openmp/{enable</a> OpenMP&gt;}
before install. In summary, run “clang -v” and then download the
corresponding LLVM tar.gz file. Run the sudo tar command as shown, and
then add these two lines to your <span class="math inline">\(\sim\)</span>/.R/Makevars file:</p>
<p>CPPFLAGS += -Xclang -fopenmp\ LDFLAGS += -lomp</p>
<p>Then <strong>DECIPHER</strong> can be built and installed the usual
way via the command line.</p>
</div>
<div class="section level4">
<h4 id="linux">Linux<a class="anchor" aria-label="anchor" href="#linux"></a>
</h4>
<p>In a shell enter:</p>
<pre><code>R CMD build --no-build-vignettes "&lt;&lt;path to DECIPHER source&gt;&gt;"
R CMD INSTALL "&lt;&lt;path to newly built DECIPHER.tar.gz&gt;&gt;"</code></pre>
</div>
<div class="section level4">
<h4 id="windows">Windows<a class="anchor" aria-label="anchor" href="#windows"></a>
</h4>
<p>Two options are available: the first is simplest, but requires the
pre-built binary (DECIPHER.zip).</p>
<ol style="list-style-type: decimal">
<li>First Option:</li>
</ol>
<pre><code>&gt; install.packages("&lt;&lt;path to Windows DECIPHER.zip&gt;&gt;", repos=NULL)</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Second Option (more difficult): [(a)]
<ol style="list-style-type: lower-alpha">
<li>Install Rtools from <a href="http://cran.r-project.org/bin/windows/Rtools/" class="external-link uri">http://cran.r-project.org/bin/windows/Rtools/</a>. Be sure
to check the box that says edit PATH during installation.</li>
<li>Open a MS-DOS command prompt by clicking Start -&gt; All Programs
-&gt; Accessories -&gt; Command Prompt.</li>
<li>In the command prompt enter:</li>
</ol>
</li>
</ol>
<pre><code>R CMD build --no-build-vignettes "&lt;&lt;path to DECIPHER source&gt;&gt;"
R CMD INSTALL "&lt;&lt;path to newly built DECIPHER.zip&gt;&gt;"</code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="example-workflow">Example Workflow<a class="anchor" aria-label="anchor" href="#example-workflow"></a>
</h2>
<p>To get started we need to load the <strong>DECIPHER</strong> package,
which automatically loads several other required packages:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DECIPHER</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: Biostrings</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     findMatches</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: XVector</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biostrings'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     strsplit</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: RSQLite</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: parallel</span></span></code></pre>
<p>Help for any function can be accessed through a command such as:</p>
<pre><code>&gt; ? DECIPHER</code></pre>
<p>To begin, we can import a FASTA, FASTQ, or GenBank file into a
sequence database. Here we will import the sequences into an in-memory
database that will be removed when we disconnect from the database.
In-memory databases are useful for temporary examples, but typically we
would specify the path to a file where we want to store the database.
This is especially the case when there are many sequences, as they might
not all fit into memory.</p>
<p>In this example, we will use a GenBank file that is included with
<strong>DECIPHER</strong> installation. We must set the sequences’
identifier when importing with <code>Seqs2DB</code>. Here we will
identify the sequences with the word “Bacteria”. The identifier is used
by many <strong>DECIPHER</strong> functions to reference a specific
subset of sequences in the database, and can be reset at a later time
using a number of different methods:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># access a sequence file included in the package:</span></span>
<span><span class="va">gen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.gen"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># connect to a database:</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">":memory:"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># import the sequences into the sequence database</span></span>
<span><span class="fu"><a href="../reference/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">gen</span>, <span class="st">"GenBank"</span>, <span class="va">dbConn</span>, <span class="st">"Bacteria"</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
Reading GenBank file chunk 1
## 
## 175 total sequences in table Seqs.
## Time difference of 0.08 secs</code></pre>
<p>Now we can view the table of information we just added to the
database in a web browser (Fig. 1):</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span></code></pre></div>
<p>Suppose we wanted to count the number of bases in each sequence and
add that information to the database:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IdLengths.html">IdLengths</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ================================================================================</span></span>
<span><span class="co">## Lengths counted for 175 sequences.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 0.01 secs</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   standard nonstandard width</span></span>
<span><span class="co">## 1     1222          13  1596</span></span>
<span><span class="co">## 2     1346           5  1596</span></span>
<span><span class="co">## 3     1323           3  1596</span></span>
<span><span class="co">## 4     1337           8  1596</span></span>
<span><span class="co">## 5     1318          25  1596</span></span>
<span><span class="co">## 6     1307           7  1596</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Add2DB.html">Add2DB</a></span><span class="op">(</span><span class="va">l</span>, <span class="va">dbConn</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span>, maxChars<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p>Next let’s identify our sequences by phylum and update this
information in the database:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IdentifyByRank.html">IdentifyByRank</a></span><span class="op">(</span><span class="va">dbConn</span>, level<span class="op">=</span><span class="fl">3</span>, add2tbl<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Updating column: "identifier"...</span></span>
<span><span class="co">## Formed 7 distinct groups.</span></span>
<span><span class="co">## Added to table Seqs: "identifier".</span></span>
<span><span class="co">## Time difference of 0.01 secs</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span>, maxChars<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p>We can now look at only those sequences that belong to the phylum
Bacteroidetes (Fig. 2):</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span>, identifier<span class="op">=</span><span class="st">"Bacteroidetes"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Search Expression:</span></span>
<span><span class="co">## select row_names, sequence from _Seqs where row_names in (select row_names</span></span>
<span><span class="co">## from Seqs where identifier is "Bacteroidetes")</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## DNAStringSet of length: 18</span></span>
<span><span class="co">## Time difference of 0 secs</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="fu">subseq</span><span class="op">(</span><span class="va">dna</span>, <span class="fl">140</span>, <span class="fl">240</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Let’s construct a phylogenetic tree from the Bacteroidetes sequences
(Fig. 3):</p>
<p>Here we created the database in-memory by connecting to “:memory:”
above. This works fine for small databases, but it is necessary to
initialize larger databases with a quoted file path rather than
“:memory:”. Optionally, we could use the command below to save an
in-memory database to a file for long term storage. Be sure to change
the path names to those on your system by replacing all of the text
inside quotes labeled “<span class="math inline">\(&lt;&lt;\)</span>path
to …<span class="math inline">\(&gt;&gt;\)</span>” with the actual path
on your system.</p>
<pre><code>&gt; sqliteCopyDatabase(dbConn, "&lt;&lt;path to database&gt;&gt;")</code></pre>
<p>Finally, we should disconnect from the database connection. Since the
sequence database was created in temporary memory, all of the
information will be erased:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
