<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DECIPHER">
<title>Design Microarray Probes • DECIPHER</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Design Microarray Probes">
<meta property="og:description" content="DECIPHER">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Design Microarray Probes</h1>
                        <h4 data-toc-skip class="author">Erik S.
Wright</h4>
            
            <h4 data-toc-skip class="date">2023-11-15</h4>
      
      
      <div class="d-none name"><code>DesignMicroarray.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This document describes how to design and validate sequence-specific
probes for synthesis onto a DNA microarray. As a case study, this
tutorial focuses on the development of a microarray to identify
taxonomic groups based on previously obtained 16S ribosomal RNA
sequences. The same approach could be applied to differentiate sequences
representing any number of groups based on any shared region of DNA. The
objective of microarray probe design is straightforward: to determine a
set of probes that will bind to one group of sequences (the target
consensus sequence) but no others (the non-targets). Beginning with a
set of aligned DNA sequences, the program chooses the best set of probes
for targeting each consensus sequence. More importantly, the design
algorithm is able to predict when potential cross-hybridization of the
probes may occur to non-target sequence(s). An integrated design
approach enables characterizing the probe set before fabrication, and
then assists with analysis of the experimental results.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="startup">Startup<a class="anchor" aria-label="anchor" href="#startup"></a>
</h3>
<p>To get started we need to load the <strong>DECIPHER</strong> package,
which automatically loads several other required packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DECIPHER</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: Biostrings</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     findMatches</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: XVector</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biostrings'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     strsplit</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: RSQLite</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: parallel</span></span></code></pre>
<p>Help for the <code>DesignArray</code> function can be accessed
through:</p>
<pre><code>&gt; ? DesignArray</code></pre>
<p>If <strong>DECIPHER</strong> is installed on your system, the code in
each example can be obtained via:</p>
<pre><code>&gt; browseVignettes("DECIPHER")</code></pre>
</div>
<div class="section level3">
<h3 id="creating-a-sequence-database">Creating a Sequence Database<a class="anchor" aria-label="anchor" href="#creating-a-sequence-database"></a>
</h3>
<p>We begin with a set of aligned sequences belonging to the 16S rRNA of
several samples obtained from drinking water distribution systems. Be
sure to change the path names to those on your system by replacing all
of the text inside quotes labeled “<span class="math inline">\(&lt;&lt;\)</span>path to …<span class="math inline">\(&gt;&gt;\)</span>” with the actual path on your
system.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify the path to your sequence file:</span></span>
<span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="st">"&lt;&lt;path to FASTA file&gt;&gt;"</span></span>
<span><span class="co"># OR find the example sequence file used in this tutorial:</span></span>
<span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.fas"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></code></pre></div>
<p>Next, there are two options for importing the sequences into a
database: either save a database file or maintain the database in
memory. Here we will build the database in memory because it is a small
set of sequences and we do not intend to use the database later:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify a path for where to write the sequence database</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="st">"&lt;&lt;path to write sequence database&gt;&gt;"</span></span>
<span><span class="co"># OR create the sequence database in memory</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">":memory:"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">fas</span>, <span class="st">"FASTA"</span>, <span class="va">dbConn</span>, <span class="st">"uncultured bacterium"</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
Reading FASTA file chunk 1
## 
## 175 total sequences in table Seqs.
## Time difference of 0.06 secs</code></pre>
</div>
<div class="section level3">
<h3 id="defining-groups">Defining Groups<a class="anchor" aria-label="anchor" href="#defining-groups"></a>
</h3>
<p>At this point we need to define groups of related sequences in the
database we just created. In this case we wish to cluster the sequences
into groups of at most 3% distance between sequences.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Search Expression:</span></span>
<span><span class="co">## select row_names, sequence from _Seqs where row_names in (select row_names</span></span>
<span><span class="co">## from Seqs)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## DNAStringSet of length: 175</span></span>
<span><span class="co">## Time difference of 0.01 secs</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DistanceMatrix.html">DistanceMatrix</a></span><span class="op">(</span><span class="va">dna</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TreeLine.html">TreeLine</a></span><span class="op">(</span>myDistMatrix<span class="op">=</span><span class="va">dMatrix</span>, type<span class="op">=</span><span class="st">"clusters"</span>, cutoff<span class="op">=</span><span class="fl">0.03</span>, method<span class="op">=</span><span class="st">"complete"</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Add2DB.html">Add2DB</a></span><span class="op">(</span><span class="va">clusters</span>, <span class="va">dbConn</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Now that we have identified 100 operational taxonomic units (OTUs),
we must form a set of consensus sequences that represent each OTU.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">conSeqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IdConsensus.html">IdConsensus</a></span><span class="op">(</span><span class="va">dbConn</span>, colName<span class="op">=</span><span class="st">"cluster"</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span>
<span><span class="co"># name the sequences by their cluster number</span></span>
<span><span class="va">ns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">conSeqs</span><span class="op">)</span>, <span class="st">"_"</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, <span class="va">`[`</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">conSeqs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"cluster"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">ns</span><span class="op">)</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># order the sequences by their cluster number</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">conSeqs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">conSeqs</span> <span class="op">&lt;-</span> <span class="va">conSeqs</span><span class="op">[</span><span class="va">o</span><span class="op">]</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="array-design-and-validation-steps">Array Design and Validation Steps<a class="anchor" aria-label="anchor" href="#array-design-and-validation-steps"></a>
</h2>
<div class="section level3">
<h3 id="designing-the-probe-set">Designing the Probe Set<a class="anchor" aria-label="anchor" href="#designing-the-probe-set"></a>
</h3>
<p>Next we will design the optimal set of 20 probes for targeting each
OTU. Since there are 100 OTUs, this process will result in a set of
2,000 probes. By default, probes are designed to have the optimal length
for hybridization at 46<span class="math inline">\(^\circ\)</span>C and
10% (vol/vol) formamide. We wish to allow up to 2 permutations for each
probe, which will potentially require more space on the microarray.
Since not all of the sequences span the alignment, we will design probes
between alignment positions 120 and 1,450, which is the region
encompassed by most of the sequences.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">probes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignArray.html">DesignArray</a></span><span class="op">(</span><span class="va">conSeqs</span>, maxPermutations<span class="op">=</span><span class="fl">2</span>, numProbes<span class="op">=</span><span class="fl">20</span>,</span>
<span>    start<span class="op">=</span><span class="fl">120</span>, end<span class="op">=</span><span class="fl">1450</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">probes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2000   12</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">probes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "name"          "start"         "length"        "start_aligned"</span></span>
<span><span class="co">##  [5] "end_aligned"   "permutations"  "score"         "formamide"    </span></span>
<span><span class="co">##  [9] "hyb_eff"       "target_site"   "probes"        "mismatches"</span></span></code></pre>
<p>We can see the probe sequence, target site positioning, melt point
(formamide), and predicted cross-hybridization efficiency to non-targets
(mismatches). The first probe targeting the first consensus sequence
(OTU #1) is predicted to have 84% hybridization efficiency at the
formamide concentration used in the experiment (10%). This probe is also
predicted to cross-hybridize with OTU #5 with 59% hybridization
efficiency.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">probes</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##   name start length start_aligned end_aligned permutations        score</span></span>
<span><span class="co">## 1    1     1     22           120         143            1 22.75502....</span></span>
<span><span class="co">##      formamide      hyb_eff            target_site</span></span>
<span><span class="co">## 1 16.06832.... 84.26072.... GCATCGGAACGTGTCCTAAAGT</span></span>
<span><span class="co">##                                       probes mismatches</span></span>
<span><span class="co">## 1 ACTTTAGGACACGTTCCGATGCTTTTTTTTTTTTTTTTTTTT  5 (59.4%)</span></span></code></pre>
<p>If we wished to have the probe set synthesized onto a microarray, all
we would need is the unique set of probes. Note that the predictive
model was calibrated using NimbleGen microarrays. Although predictions
are likely similar for other microarray platforms, hybridization
conditions should always be experimentally optimized.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">u</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">probes</span><span class="op">$</span><span class="va">probes</span>, <span class="st">","</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">u</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1899</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">u</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ACTTTAGGACACGTTCCGATGCTTTTTTTTTTTTTTTTTTTT"  </span></span>
<span><span class="co">## [2] "GTATTAGCGCATCTTTCGATGCTTTTTTTTTTTTTTTTTTTT"  </span></span>
<span><span class="co">## [3] "GTCTTTCGATCCCCTACTTTCCTCTTTTTTTTTTTTTTTTTTTT"</span></span>
<span><span class="co">## [4] "GGCCGCTCCAAAAGCATAAGGTTTTTTTTTTTTTTTTTTTTT"  </span></span>
<span><span class="co">## [5] "ATGGCAATTAATGACAAGGGTTGCTTTTTTTTTTTTTTTTTTTT"</span></span>
<span><span class="co">## [6] "CAGTGTGGTTGGCCATCCTCTTTTTTTTTTTTTTTTTTTTT"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="validating-the-probe-set">Validating the Probe Set<a class="anchor" aria-label="anchor" href="#validating-the-probe-set"></a>
</h3>
<p>Before fabrication onto a DNA microarray it may be useful to predict
whether the probe set will adequately discriminate between the OTUs.
This can be accomplished by simulating the hybridization process
multiple times while incorporating error. We begin by converting the
predicted cross-hybridization efficiencies into a sparse matrix that
mathematically represents the microarray (<strong>A</strong>). Here the
rows of the matrix represent each probe, and the columns of the matrix
represent each OTU. The entries of the matrix therefore give the
hybridization efficiency of probe <em>i</em> with OTU <em>j</em>. We can
neglect all hybridization efficiencies less than 5% because these will
likely not hybridize or have insufficient brightness.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Array2Matrix.html">Array2Matrix</a></span><span class="op">(</span><span class="va">probes</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">A</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">w</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">A</span><span class="op">$</span><span class="va">i</span> <span class="op">&lt;-</span> <span class="va">A</span><span class="op">$</span><span class="va">i</span><span class="op">[</span><span class="op">-</span><span class="va">w</span><span class="op">]</span></span>
<span>    <span class="va">A</span><span class="op">$</span><span class="va">j</span> <span class="op">&lt;-</span> <span class="va">A</span><span class="op">$</span><span class="va">j</span><span class="op">[</span><span class="op">-</span><span class="va">w</span><span class="op">]</span></span>
<span>    <span class="va">A</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">A</span><span class="op">$</span><span class="va">x</span><span class="op">[</span><span class="op">-</span><span class="va">w</span><span class="op">]</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We then multiply the matrix <strong>A</strong> by the amount (x) of
each OTU present to determine the corresponding brightness (b) values of
each probe. We can add a heteroskedastic error to the brightness values
to result in a more accurate simulation (b = <strong>A</strong>x +
error). Furthermore, we can introduce a 5% rate of probes that hybridize
randomly.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># simulate the case where 10</span></span>
<span><span class="va">present</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">conSeqs</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">conSeqs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">conSeqs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">[</span><span class="va">present</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">present</span><span class="op">)</span>, sd<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># determine the predicted probe brightnesses based on the present OTUS</span></span>
<span><span class="va">background</span> <span class="op">&lt;-</span> <span class="fl">0.2</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">A</span><span class="op">$</span><span class="va">x</span><span class="op">[</span><span class="va">A</span><span class="op">$</span><span class="va">j</span><span class="op">]</span><span class="op">*</span><span class="va">x</span><span class="op">[</span><span class="va">A</span><span class="op">$</span><span class="va">j</span><span class="op">]</span>, <span class="va">A</span><span class="op">$</span><span class="va">i</span>, <span class="va">sum</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="va">background</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="va">b</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span>, sd<span class="op">=</span><span class="fl">0.2</span><span class="op">*</span><span class="va">b</span><span class="op">)</span> <span class="co"># add 20</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="va">b</span> <span class="op">-</span> <span class="va">background</span> <span class="co"># background subtracted brightnesses</span></span>
<span></span>
<span><span class="co"># add in a 5</span></span>
<span><span class="va">bad_hybs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">b</span><span class="op">[</span><span class="va">bad_hybs</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">bad_hybs</span><span class="op">)</span>, sd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we can solve for the amount of each OTU present on the
microarray by solving <strong>A</strong>x = b for x using non-negative
(x <span class="math inline">\(\ge\)</span> 0) least squares. Plotting
the expected amount versus the predicted amount shows that this probe
set may result in a small number of false positives and false negatives
(Fig. 1). False negatives are the expected observations below the dashed
threshold line, which represents the minimum amount required to be
considered present. If this threshold is lowered then false negatives
will appear where no amount was expected.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># solve for the predicted amount of each OTU present on the array</span></span>
<span><span class="va">x_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NNLS.html">NNLS</a></span><span class="op">(</span><span class="va">A</span>, <span class="va">b</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="further-improving-the-result">Further Improving the Result<a class="anchor" aria-label="anchor" href="#further-improving-the-result"></a>
</h3>
<p>Least squares regression is particularly sensitive outlier
observations and heteroskedastic noise. For this reason we will decrease
the effects of outlier observations by using weighted regression. With
each iteration the weights will be refined using the residuals from the
prior solution to <strong>A</strong>x = b.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># initialize weights to one:</span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># iteratively unweight observations with high residuals:</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">{</span> <span class="co"># 10 iterations</span></span>
<span>    <span class="va">weights</span> <span class="op">&lt;-</span> <span class="va">weights</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.1</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">x_out</span><span class="op">$</span><span class="va">residuals</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">A_weighted</span> <span class="op">&lt;-</span> <span class="va">A</span></span>
<span>    <span class="va">A_weighted</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">A</span><span class="op">$</span><span class="va">x</span><span class="op">*</span><span class="va">weights</span><span class="op">[</span><span class="va">A</span><span class="op">$</span><span class="va">i</span><span class="op">]</span></span>
<span>    <span class="va">b_weighted</span> <span class="op">&lt;-</span> <span class="va">b</span><span class="op">*</span><span class="va">weights</span></span>
<span>    <span class="va">x_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NNLS.html">NNLS</a></span><span class="op">(</span><span class="va">A_weighted</span>, <span class="va">b_weighted</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Weighted regression lowered the threshold for detection so that more
OTUs would be detectable (Fig. 2). However, false negatives still remain
based on this simulation when a very small amount is expected. If the
threshold is lowered to capture all of the expected OTUs then we can
determine the false positive(s) that would result. These false positive
sequences are substantially different from the nearest sequence that is
present.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">x_out</span><span class="op">$</span><span class="va">x</span> <span class="op">&gt;=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">x_out</span><span class="op">$</span><span class="va">x</span><span class="op">[</span><span class="va">present</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">w</span> <span class="op">&lt;-</span> <span class="va">w</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">present</span>, <span class="va">w</span><span class="op">)</span><span class="op">]</span> <span class="co"># false positives</span></span>
<span><span class="va">dMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DistanceMatrix.html">DistanceMatrix</a></span><span class="op">(</span><span class="va">conSeqs</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># print distances of false positives to the nearest present OTU</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">w</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">dMatrix</span><span class="op">[</span><span class="va">i</span>, <span class="va">present</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.07703927</span></span>
<span><span class="co">## [1] 0.112588</span></span>
<span><span class="co">## [1] 0.08248233</span></span>
<span><span class="co">## [1] 0.03396524</span></span>
<span><span class="co">## [1] 0.06382979</span></span>
<span><span class="co">## [1] 0.0608565</span></span>
<span><span class="co">## [1] 0.1231603</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="extending-the-simulation">Extending the Simulation<a class="anchor" aria-label="anchor" href="#extending-the-simulation"></a>
</h3>
<p>The above simulation can be repeated multiple times and with
different initial conditions to better approximate the expected number
of false positives and false negatives (Fig. 3). In the same manner the
design parameters can be iteratively optimized to further improve the
predicted specificity of the probe set based on the simulation results.
After fabrication, validation experiments using known samples should be
used in replace of the simulated brightness values.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># simulate multiple cases where 10</span></span>
<span><span class="va">iterations</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, ncol<span class="op">=</span><span class="va">iterations</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">conSeqs</span><span class="op">)</span>, ncol<span class="op">=</span><span class="va">iterations</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">iterations</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">present</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">conSeqs</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">conSeqs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">x</span><span class="op">[</span><span class="va">present</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">present</span><span class="op">)</span>, sd<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># determine the predicted probe brightnesses based on the present OTUS</span></span>
<span>    <span class="va">b</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">A</span><span class="op">$</span><span class="va">x</span><span class="op">[</span><span class="va">A</span><span class="op">$</span><span class="va">j</span><span class="op">]</span><span class="op">*</span><span class="va">x</span><span class="op">[</span><span class="va">A</span><span class="op">$</span><span class="va">j</span>, <span class="va">i</span><span class="op">]</span>, <span class="va">A</span><span class="op">$</span><span class="va">i</span>, <span class="va">sum</span><span class="op">)</span> <span class="op">+</span> <span class="va">background</span></span>
<span>    <span class="va">b</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">b</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, sd<span class="op">=</span><span class="fl">0.2</span><span class="op">*</span><span class="va">b</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span> <span class="co"># add 20</span></span>
<span>    <span class="va">b</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">b</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">-</span> <span class="va">background</span> <span class="co"># background subtracted brightnesses</span></span>
<span>    </span>
<span>    <span class="co"># add in a 5</span></span>
<span>    <span class="va">bad_hybs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">b</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">b</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">b</span><span class="op">[</span><span class="va">bad_hybs</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">bad_hybs</span><span class="op">)</span>, sd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">b</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">x_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NNLS.html">NNLS</a></span><span class="op">(</span><span class="va">A</span>, <span class="va">b</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
