<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DECIPHER">
<title>Design Group-Specific Primers • DECIPHER</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Design Group-Specific Primers">
<meta property="og:description" content="DECIPHER">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Design Group-Specific Primers</h1>
                        <h4 data-toc-skip class="author">Erik S.
Wright</h4>
            
            <h4 data-toc-skip class="date">2023-11-15</h4>
      
      
      <div class="d-none name"><code>DesignPrimers.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This document describes how to design group-specific primers using
the <strong>DECIPHER</strong> package through the use of the
<code>DesignPrimers</code> function. As a case study, this tutorial
focuses on the Internal Transcribed Spacer (ITS) of fully sequenced
genomes belonging to different species of the genus
<em>Streptomyces</em>. The ITS resides on the chromosome between the
genes coding for the 16S and 23S ribosomal RNA. The examples in this
document are directed towards finding <strong>primers that target a
single species</strong> of <em>Streptomyces</em> among a number of
closely related <em>Streptomyces</em> species. However, a similar
strategy could be used with any set of aligned sequences that are
separated into groups. For example, genus-specific primers targeting the
16S gene that have been designed with this program are provided online
at <a href="http://DECIPHER.codes" class="external-link uri">http://DECIPHER.codes</a>.</p>
<p>A database of <strong>aligned DNA sequences</strong> separated into
groups is used as input to the program. First the function
<code>TileSeqs</code> is used to pre-process the sequences into
overlapping tiles, which will serve as the template DNA for primer
design. Second, the <code>DesignPrimers</code> function determines the
set of all possible primers that meet certain design constraints, such
as the ability to efficiently amplify the group of interest under
specified experimental conditions. Next, the complete set of primers is
scored for its predicted potential to cross-amplify with DNA belonging
to other groups. Finally, the optimal set of forward and reverse primers
is chosen that could be used in a PCR experiment to enrich for the DNA
of interest in a sample containing DNA from multiple groups.</p>
</div>
<div class="section level2">
<h2 id="the-objective-of-primer-design">The Objective of Primer Design<a class="anchor" aria-label="anchor" href="#the-objective-of-primer-design"></a>
</h2>
<p>The objective of primer design is straightforward: to determine a set
of forward and reverse primers that will amplify one group of sequences
(the target) but no others (the non-targets). If such a primer set is
unattainable then the goal becomes to accurately predict potential
cross-amplification with non-target groups. This knowledge enables an
educated assessment of the different primer options that could be used
to minimize non-target interference. The chosen primers could then be
synthesized and experimentally characterized with PCR or quantitative
PCR (qPCR).</p>
<p>Possible applications include:</p>
<ul>
<li><p>allele specific PCR</p></li>
<li><p>quantifying a subset of organisms in a sample</p></li>
<li><p>enriching for the DNA of a target group before downstream
sequencing.</p></li>
</ul>
<p>Improved specificity to the target group is achieved in two ways: 1)
by decreasing hybridization efficiency, and 2) by minimizing elongation
efficiency of the primer with non-target DNA. Hybridization efficiency
is a theoretical variable that represents the fraction of DNA template
bound to primer during the annealing step of PCR thermal cycling. In
this manner, mismatches between the primer and DNA templates are
utilized to lower the affinity of the primers to non-target DNA.
Elongation efficiency represents the efficiency of the DNA polymerase at
initiating extension of the DNA template from a mismatched 3’ primer
terminus, and is measured relative to the perfectly matched primer’s
ability to elongate. Mismatches near the 3’ end of the primer are used
to hinder extension of non-target DNA templates thereby increasing
specificity of the primers to target DNA. Therefore, this model of
terminal mismatches offers the dual benefits of decreased hybridization
and elongation efficiency, which allows for more accurate predictions of
amplification efficiency and increased specificity to the target
group.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="installing-oligoarrayaux">Installing OligoArrayAux<a class="anchor" aria-label="anchor" href="#installing-oligoarrayaux"></a>
</h3>
<p>The program OligoArrayAux (<a href="http://www.unafold.org/Dinamelt/software/oligoarrayaux.php" class="external-link uri">http://www.unafold.org/Dinamelt/software/oligoarrayaux.php</a>)
is used to predict hybridization efficiency and must be installed in a
location accessible by the system. For example, the following code
should print the installed OligoArrayAux version when executed from the
R console:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"hybrid-min -V"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="start-up">Start up<a class="anchor" aria-label="anchor" href="#start-up"></a>
</h3>
<p>To get started we need to load the <strong>DECIPHER</strong> package,
which automatically loads several other required packages.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DECIPHER</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<p>Help for the <code>DesignPrimers</code> function can be accessed
through:</p>
<pre><code>&gt; ? DesignPrimers</code></pre>
<p>If <strong>DECIPHER</strong> is installed on your system, the code in
each example can be obtained via:</p>
<pre><code>&gt; browseVignettes("DECIPHER")</code></pre>
</div>
<div class="section level3">
<h3 id="creating-a-sequence-database">Creating a Sequence Database<a class="anchor" aria-label="anchor" href="#creating-a-sequence-database"></a>
</h3>
<p>We begin with a set of aligned sequences belonging to the Internal
Transcribed Spacer (ITS) of several <em>Streptomyces</em> chromosomes.
We wish to determine if the ITS region has enough variability to be used
for distinguishing between these closely related species with qPCR. This
example uses a FASTA sequence file included as part of the
<strong>DECIPHER</strong> package, but you could follow along with your
own FASTA file of aligned sequences. Be sure to change the path names to
those on your system by replacing all of the text inside quotes labeled
“<span class="math inline">\(&lt;&lt;\)</span>path to …<span class="math inline">\(&gt;&gt;\)</span>” with the actual path on your
system.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify the path to your sequence file:</span></span>
<span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="st">"&lt;&lt;path to FASTA file&gt;&gt;"</span></span>
<span><span class="co"># OR find the example sequence file used in this tutorial:</span></span>
<span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Streptomyces_ITS_aligned.fas"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></code></pre></div>
<p>Next, there are two options for importing the sequences into a
database: either save a database file or maintain the database in
memory. Here we will build the database in memory because it is a small
set of sequences and we do not intend to use the database later:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify a path for where to write the sequence database</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="st">"&lt;&lt;path to write sequence database&gt;&gt;"</span></span>
<span><span class="co"># OR create the sequence database in memory</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">":memory:"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">fas</span>, <span class="st">"FASTA"</span>, <span class="va">dbConn</span>, <span class="st">"Streptomyces"</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
Reading FASTA file chunk 1
## 
## 88 total sequences in table Seqs.
## Time difference of 0.05 secs</code></pre>
</div>
<div class="section level3">
<h3 id="defining-groups">Defining Groups<a class="anchor" aria-label="anchor" href="#defining-groups"></a>
</h3>
<p>At this point we need to define groups of related sequences in the
database we just created. In this case we wish to define a unique group
for the sequences belonging to each species of <em>Streptomyces</em>. In
order to accomplish this we must parse the information contained in the
description of each FASTA sequence record that we just imported.
Alternatively we could use the functions <code>IdentifyByRank</code>,
<code>FormGroups</code>, <code>Clusterize</code>, or
<code>TreeLine</code> to define groups.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the FASTA record description</span></span>
<span><span class="va">desc</span> <span class="op">&lt;-</span> <span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">dbConn</span>, <span class="st">"select description from Seqs"</span><span class="op">)</span></span>
<span><span class="co"># parse the sequence description to obtain the species name</span></span>
<span><span class="va">desc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">desc</span><span class="op">$</span><span class="va">description</span>, <span class="st">"Streptomyces "</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">desc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"sp. "</span>, <span class="st">""</span>, <span class="va">desc</span>, perl<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">desc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"sp_"</span>, <span class="st">""</span>, <span class="va">desc</span>, perl<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">desc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">desc</span>, <span class="st">" "</span>, fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">desc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "albus"             "clavuligerus"      "ghanaensis"       </span></span>
<span><span class="co">##  [4] "griseoflavus"      "lividans"          "pristinaespiralis"</span></span>
<span><span class="co">##  [7] "Mg1"               "SPB78"             "AA4"              </span></span>
<span><span class="co">## [10] "SPB74"             "SirexAA-E"         "scabiei"          </span></span>
<span><span class="co">## [13] "griseus"           "coelicolor"        "cattleya"         </span></span>
<span><span class="co">## [16] "bingchenggensis"   "avermitilis"       "C"                </span></span>
<span><span class="co">## [19] "Tu6071"</span></span></code></pre>
<p>Now that we have our 19 different species names, we must add them to
the database as the identifier of each sequence.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Add2DB.html">Add2DB</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>identifier<span class="op">=</span><span class="va">desc</span>, stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>, <span class="va">dbConn</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Expression:</span></span>
<span><span class="co">## update Seqs set identifier = :identifier where row_names = :row_names</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Added to table Seqs:  "identifier".</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 0 secs</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="primer-design-steps">Primer Design Steps<a class="anchor" aria-label="anchor" href="#primer-design-steps"></a>
</h2>
<div class="section level3">
<h3 id="tiling-sequences">Tiling Sequences<a class="anchor" aria-label="anchor" href="#tiling-sequences"></a>
</h3>
<p>We continue by creating a set of k-mers, known as “tiles”, that
represent the sequences in each group. Here we must make several
decisions that will affect primer design in the future. The defaults are
to create tiles of length 26-27 nucleotides with up to 10 permutations
that represent at least 90% of the permutations present in each target
site. These parameters will generally fit most primer designs, but it is
recommended to read the help file for the <code>TileSeqs</code> function
to make sure that it is doing what is desired. We will save the
resulting tiles back to the database as a new table named “Tiles” in
case we wish to access them in the future.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TileSeqs.html">TileSeqs</a></span><span class="op">(</span><span class="va">dbConn</span>, add2tbl<span class="op">=</span><span class="st">"Tiles"</span>, minCoverage<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 8.44 secs</span></span></code></pre>
<p>Now we can examine the first few tiles by using the <code>head</code>
command. Here we see that some tiles have been flagged as
<code>misprime=TRUE</code>, which in this case was caused by some
permutations of this target site containing degenerate bases (N’s).
Ambiguity codes are not used in primer design and are thus excluded from
consideration as potential target sites for primers. Other reasons for
mispriming include runs of a single base and di-nucleotide repeats,
which increase primer entropy and could result in non-specific
annealing. If at least a <code>minCoverage</code> fraction of the group
could not be represented in <code>maxPermutations</code> then
<code>misprime</code> is also marked as <code>TRUE</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tiles</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   row_names start end start_aligned end_aligned misprime width    id coverage</span></span>
<span><span class="co">## 1         1     1  27             1          28     TRUE   627 albus      0.6</span></span>
<span><span class="co">## 2         2     2  28             2          29     TRUE   627 albus      0.6</span></span>
<span><span class="co">## 3         3     3  29             3          30     TRUE   627 albus      0.6</span></span>
<span><span class="co">## 4         4     4  30             4          31     TRUE   627 albus      0.6</span></span>
<span><span class="co">## 5         5     5  31             5          32    FALSE   627 albus      1.0</span></span>
<span><span class="co">## 6         6     6  32             6          33    FALSE   627 albus      1.0</span></span>
<span><span class="co">##   groupCoverage                 target_site</span></span>
<span><span class="co">## 1           0.6 TGTACACACCGCCCGTCACGTCACGAA</span></span>
<span><span class="co">## 2           0.6 GTACACACCGCCCGTCACGTCACGAAA</span></span>
<span><span class="co">## 3           0.6 TACACACCGCCCGTCACGTCACGAAAG</span></span>
<span><span class="co">## 4           0.6 ACACACCGCCCGTCACGTCACGAAAGT</span></span>
<span><span class="co">## 5           1.0 CACACCGCCCGTCACGTCACGAAAGTC</span></span>
<span><span class="co">## 6           1.0 ACACCGCCCGTCACGTCACGAAAGTCG</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="designing-all-possible-primers">Designing All Possible Primers<a class="anchor" aria-label="anchor" href="#designing-all-possible-primers"></a>
</h3>
<p>Next we wish to design primers for the group <em>S. avermitilis</em>.
As an example, we will begin by designing primers for every possible
target site that meet certain design criteria. For example, the default
reagent conditions are defined as: 3 mM [dNTPs], 70 mM [<span class="math inline">\(K^{+}\)</span>], 3 mM [<span class="math inline">\(Mg^{2+}\)</span>], and a primer concentration of
400 nM. Also, the defaults specify that up to 4 forward or reverse
primer permutations can be used so long as they cover 90% of the
variants present in their target site. In this case we wish to encompass
100% of each species’ sequences, so we will use
<code>minCoverage = 1</code> and <code>minGroupCoverage = 1</code>
rather than the default 90% and 20%, respectively. Other default
parameters include an annealing temperature of <span class="math inline">\(64^{\circ}C\)</span>, and a minimum and maximum
product size of 75 and 1,200 base pairs, respectively. Note that the
input parameters to <code>DesignPrimers</code> should be carefully
considered in order adequately represent the actual PCR conditions that
will be used.</p>
<p>It is worth noting that the estimates of elongation efficiency are
applicable to standard <em>Taq</em> DNA Polymerase, and do not apply to
high-fidelity polymerases. Such polymerases, commonly used in cloning
and sequencing applications, have a 3’ to 5’ exonuclease activity that
will attempt to remove mismatches located at the primer’s 3’ terminus.
If you intend to use a high-fidelity polymerase then you should specify
<code>taqEfficiency=FALSE</code> when issuing commands to
<code>DesignPrimers</code> or <code>CalculateEfficiencyPCR</code>. In
this case elongation efficiency will not be included in the estimate of
amplification efficiency and only hybridization efficiency will be
used.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignPrimers.html">DesignPrimers</a></span><span class="op">(</span><span class="va">tiles</span>, identifier<span class="op">=</span><span class="st">"avermitilis"</span>,</span>
<span>    minCoverage<span class="op">=</span><span class="fl">1</span>, minGroupCoverage<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## avermitilis (499 candidate primers):</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 4.9 secs</span></span></code></pre>
<p>Not that this command designs individual primers and a different
command is required to design pairs of forward and reverse primers (see
below). Let’s examine the first primer that is available, out of 499
potential target sites. We can see that only one permutation was
required to cover 100% of the species in this target site. However, the
score of this primer is relatively bad (negative) because it has a high
efficiency of amplification predicted with other groups (i.e.,
species).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">primers</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 499 × 17</span></span></span>
<span><span class="co">##    identifier  start_forward start_reverse start_aligned_forward</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;I&lt;chr&gt;&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;I&lt;int&gt;&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;I&lt;dbl&gt;&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;I&lt;int&gt;&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> avermitilis            11            17                     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> avermitilis            11            19                     2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> avermitilis            11            19                     3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> avermitilis            12            20                     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> avermitilis            12            21                     5</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> avermitilis            13            22                     6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> avermitilis            14            23                     7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> avermitilis            14            24                     8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> avermitilis            16            25                     9</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> avermitilis            16            26                    10</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 489 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 13 more variables: start_aligned_reverse &lt;I&lt;int&gt;&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   permutations_forward &lt;I&lt;dbl&gt;&gt;, permutations_reverse &lt;I&lt;dbl&gt;&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   score_forward &lt;I&lt;dbl&gt;&gt;, score_reverse &lt;I&lt;dbl&gt;&gt;, forward_primer &lt;chr[,4]&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   reverse_primer &lt;chr[,4]&gt;, forward_efficiency &lt;dbl[,4]&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   reverse_efficiency &lt;dbl[,4]&gt;, forward_coverage &lt;dbl[,4]&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   reverse_coverage &lt;dbl[,4]&gt;, mismatches_forward &lt;I&lt;chr&gt;&gt;, …</span></span></span></code></pre>
<p>The predicted forward and reverse primer efficiencies are
respectively 84.4% and 88.7% at the default annealing temperature of
<span class="math inline">\(64^{\circ}C\)</span>. The efficiency metric
is useful because it approximates the fraction of available templates
that will be copied in each cycle of PCR. In the
<code>mismatches_forward</code> and <code>mismatches_reverse</code>
fields the predicted amplification efficiency of using the primers with
each non-target group is shown. Here we can see that most non-target
groups have the same predicted efficiency as the target, because they
have DNA that perfectly matches the primer sequences.</p>
</div>
<div class="section level3">
<h3 id="designing-the-best-possible-primer-set">Designing the Best Possible Primer Set<a class="anchor" aria-label="anchor" href="#designing-the-best-possible-primer-set"></a>
</h3>
<p>We could potentially examine all primers one-by-one to find the best
forward and reverse primers to amplify only the target group. Another
option is to allow the program to find the best primer sets by searching
through all combinations of forward and reverse primers. To do this we
simply change the parameter <code>numPrimerSets</code> to a positive
number from its default of zero. In doing so the
<code>DesignPrimers</code> function will search through the best forward
and reverse primers to find the option that minimizes forward and
reverse primer overlap. Furthermore, the program will search potential
target sites within 20 nucleotides of where the primer might bind
upstream and downstream of its target site in the alignment.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignPrimers.html">DesignPrimers</a></span><span class="op">(</span><span class="va">tiles</span>, identifier<span class="op">=</span><span class="st">"avermitilis"</span>, minCoverage<span class="op">=</span><span class="fl">1</span>,</span>
<span>    minGroupCoverage<span class="op">=</span><span class="fl">1</span>, numPrimerSets<span class="op">=</span><span class="fl">5</span>, maxSearchSize<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## avermitilis (499 candidate primers):</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## Determining Best Primer Pairs:</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 6.7 secs</span></span></code></pre>
<p>Now each row of the output (<code>primers</code>) contains one pair
of forward and reverse primers. Examining the output we see that several
primer sets are predicted to only amplify <em>S. avermitilis</em> and
not any other species, as shown by the empty column
<em>mismatches_set</em>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">primers</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 5 × 20</span></span></span>
<span><span class="co">##   identifier  start_forward start_reverse product_size start_aligned_forward</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;I&lt;chr&gt;&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;I&lt;int&gt;&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;I&lt;dbl&gt;&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;I&lt;dbl&gt;&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;I&lt;int&gt;&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> avermitilis           245           325           81                   247</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> avermitilis           245           322           78                   247</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> avermitilis           245           326           82                   247</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> avermitilis           245           324           80                   247</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> avermitilis           245           324           80                   247</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 15 more variables: start_aligned_reverse &lt;I&lt;int&gt;&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   permutations_forward &lt;I&lt;dbl&gt;&gt;, permutations_reverse &lt;I&lt;dbl&gt;&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   score_forward &lt;I&lt;dbl&gt;&gt;, score_reverse &lt;I&lt;dbl&gt;&gt;, score_set &lt;I&lt;dbl&gt;&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   forward_primer &lt;I&lt;chr[,4]&gt;[,4]&gt;, reverse_primer &lt;I&lt;chr[,4]&gt;[,4]&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   forward_efficiency &lt;I&lt;dbl[,4]&gt;[,4]&gt;, reverse_efficiency &lt;I&lt;dbl[,4]&gt;[,4]&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   forward_coverage &lt;I&lt;dbl[,4]&gt;[,4]&gt;, reverse_coverage &lt;I&lt;dbl[,4]&gt;[,4]&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   mismatches_forward &lt;I&lt;chr&gt;&gt;, mismatches_reverse &lt;I&lt;chr&gt;&gt;, …</span></span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="additional-primer-analysis">Additional Primer Analysis<a class="anchor" aria-label="anchor" href="#additional-primer-analysis"></a>
</h2>
<div class="section level3">
<h3 id="denaturation-plot">Denaturation Plot<a class="anchor" aria-label="anchor" href="#denaturation-plot"></a>
</h3>
<p>After choosing forward and reverse primers we can graph their melt
curves in order to predict how they will behave at different annealing
temperatures. Recall that in this example there is only a single forward
or reverse primer permutation necessary to cover the entire group.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_range</span> <span class="op">&lt;-</span> <span class="fl">60</span><span class="op">:</span><span class="fl">75</span></span>
<span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CGTTGATTATTCGGCACACTCGAC"</span>, <span class="st">"CCCTCGCCCTCCCATGT"</span><span class="op">)</span> <span class="co"># forward and reverse</span></span>
<span><span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">temp</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/CalculateEfficiencyPCR.html">CalculateEfficiencyPCR</a></span><span class="op">(</span><span class="va">ps</span>, <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/reverseComplement.html" class="external-link">reverseComplement</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                         <span class="va">temp</span>, P<span class="op">=</span><span class="fl">4e-7</span>, ions<span class="op">=</span><span class="fl">.225</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">efficiency</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">temp_range</span>, <span class="va">f</span><span class="op">)</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">2</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">temp_range</span>, <span class="va">efficiency</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, ylab<span class="op">=</span><span class="st">"Hybridization Efficiency"</span>,</span>
<span>     xlab<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Temperature ("</span>, <span class="va">degree</span>, <span class="st">"C)"</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span><span class="op">)</span>,</span>
<span>     type<span class="op">=</span><span class="st">"l"</span>, lwd<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="st">"Blue"</span>, main<span class="op">=</span><span class="st">"Denaturation Plot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">temp_range</span>, <span class="va">efficiency</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"Red"</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fl">0.5</span>, lty<span class="op">=</span><span class="fl">2</span>, lwd<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="st">"Orange"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fl">64</span>, lty<span class="op">=</span><span class="fl">2</span>, lwd<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="st">"Green"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Forward Primer"</span>, <span class="st">"Reverse Primer"</span>, <span class="st">"50% Efficiency"</span>,</span>
<span>                            <span class="st">"Annealing Temperature"</span><span class="op">)</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Blue"</span>, <span class="st">"Red"</span>, <span class="st">"Orange"</span>, <span class="st">"Green"</span><span class="op">)</span>,</span>
<span>       lwd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, lty<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="DesignPrimers_files/figure-html/expr12-1.png" width="700"></p>
<p>We originally designed the primers to have at least 80%
<code>minEfficiency</code> at the annealing temperature of <span class="math inline">\(64^{\circ}C\)</span>, and it is clear from the
graph that both primers are predicted to have similarly high
efficiencies at this temperature. However, it appears that the forward
primer will denature more quickly than the reverse primer as the
temperature is increased. This is due to the fact that these primers
have different enthalpy and entropy, but similar free energies of duplex
formation near the annealing temperature where their melt curves
intersect.</p>
</div>
<div class="section level3">
<h3 id="visualizing-the-target-sites">Visualizing the Target Sites<a class="anchor" aria-label="anchor" href="#visualizing-the-target-sites"></a>
</h3>
<p>Often it is useful to visualize the sequence region where the primers
will anneal on both the target and non-targets side-by-side. We can
accomplish this by looking at the amplicon sequences and highlighting
the region of both primers. First we query the database for the
sequences of interest and proceed to trim them to the region around the
amplicon given by the <code>start_aligned_forward</code> and
<code>start_aligned_reverse</code> outputs of
<code>DesignPrimers</code>. Next we name and order the unique sequences
in the set such that the target (<em>S. avermitilis</em>) appears at the
top.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Search Expression:</span></span>
<span><span class="co">## select row_names, sequence from _Seqs where row_names in (select row_names</span></span>
<span><span class="co">## from Seqs)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## DNAStringSet of length: 88</span></span>
<span><span class="co">## Time difference of 0 secs</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span>
<span><span class="va">amplicon</span> <span class="op">&lt;-</span> <span class="fu">subseq</span><span class="op">(</span><span class="va">dna</span>, <span class="fl">247</span>, <span class="fl">348</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">amplicon</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">desc</span></span>
<span><span class="co"># only show unique sequences</span></span>
<span><span class="va">u_amplicon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">amplicon</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">u_amplicon</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">amplicon</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">u_amplicon</span>, <span class="va">amplicon</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">amplicon</span> <span class="op">&lt;-</span> <span class="va">u_amplicon</span></span>
<span><span class="co"># move the target group to the top</span></span>
<span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">amplicon</span><span class="op">)</span><span class="op">==</span><span class="st">"avermitilis"</span><span class="op">)</span></span>
<span><span class="va">amplicon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">amplicon</span><span class="op">[</span><span class="va">w</span><span class="op">]</span>, <span class="va">amplicon</span><span class="op">[</span><span class="op">-</span><span class="va">w</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Then we can use the function <code>BrowseSeqs</code> to visualize the
target site for each primer in the amplicon. This command will open a
new window in the default web browser. The top sequence shows the target
(<em>S. avermitilis</em>), and each primer’s target site is colored. We
can highlight the target sequence so that only mismatches are colored in
the non-target sequences. It is clear why this target site was chosen,
because all other <em>Streptomyces</em> species have multiple mismatches
to both the forward and reverse primers. The non-target mismatches,
especially those located nearest the 3’ end (towards the center of the
amplicon), will act to lower both hybridization and elongation
efficiency of non-target amplification.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">amplicon</span>, colorPatterns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">27</span>, <span class="fl">76</span>, <span class="fl">94</span><span class="op">)</span>, highlight<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="finishing-up">Finishing Up<a class="anchor" aria-label="anchor" href="#finishing-up"></a>
</h3>
<p>Finally, we can order the forward and reverse primers for synthesis
and try them out in a PCR reaction! The forward and reverse primers
should be synthesized in the same orientation and sense that they are
listed in the output (<code>primers</code>). Initially it may be useful
to perform a temperature gradient PCR reaction with the target DNA in
order to experimentally determine the melt point. For example, in this
case we would run several reactions with the target species at
increasing temperature, from about <span class="math inline">\(56^{\circ}C\)</span> to <span class="math inline">\(72^{\circ}C\)</span>. As the annealing temperature
is increased the hybridization efficiency will eventually decrease until
no amplification is observed. In subsequent experiments the annealing
temperature should be set just below the highest temperature where
robust amplification was observed.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>All of the output in this vignette was produced under the following
conditions:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=zh_CN.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=zh_CN.UTF-8        LC_COLLATE=zh_CN.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=zh_CN.UTF-8    LC_MESSAGES=zh_CN.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=zh_CN.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=zh_CN.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Asia/Shanghai</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] dplyr_1.1.3         DECIPHER_2.28.0     RSQLite_2.3.1      </span></span>
<span><span class="co">## [4] Biostrings_2.68.1   GenomeInfoDb_1.36.2 XVector_0.40.0     </span></span>
<span><span class="co">## [7] IRanges_2.34.1      S4Vectors_0.38.1    BiocGenerics_0.46.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] utf8_1.2.4              sass_0.4.7              generics_0.1.3         </span></span>
<span><span class="co">##  [4] bitops_1.0-7            stringi_1.7.12          digest_0.6.33          </span></span>
<span><span class="co">##  [7] magrittr_2.0.3          evaluate_0.23           fastmap_1.1.1          </span></span>
<span><span class="co">## [10] blob_1.2.4              rprojroot_2.0.4         jsonlite_1.8.7         </span></span>
<span><span class="co">## [13] DBI_1.1.3               fansi_1.0.5             purrr_1.0.2            </span></span>
<span><span class="co">## [16] textshaping_0.3.6       jquerylib_0.1.4         cli_3.6.1              </span></span>
<span><span class="co">## [19] rlang_1.1.2             crayon_1.5.2            bit64_4.0.5            </span></span>
<span><span class="co">## [22] cachem_1.0.8            yaml_2.3.7              tools_4.3.2            </span></span>
<span><span class="co">## [25] memoise_2.0.1           GenomeInfoDbData_1.2.10 vctrs_0.6.4            </span></span>
<span><span class="co">## [28] R6_2.5.1                lifecycle_1.0.3         zlibbioc_1.46.0        </span></span>
<span><span class="co">## [31] stringr_1.5.0           fs_1.6.3                bit_4.0.5              </span></span>
<span><span class="co">## [34] ragg_1.2.5              pkgconfig_2.0.3         desc_1.4.2             </span></span>
<span><span class="co">## [37] pkgdown_2.0.7           bslib_0.5.1             pillar_1.9.0           </span></span>
<span><span class="co">## [40] glue_1.6.2              systemfonts_1.0.5       highr_0.10             </span></span>
<span><span class="co">## [43] tidyselect_1.2.0        tibble_3.2.1            xfun_0.41              </span></span>
<span><span class="co">## [46] rstudioapi_0.15.0       knitr_1.45              htmltools_0.5.7        </span></span>
<span><span class="co">## [49] rmarkdown_2.25          compiler_4.3.2          RCurl_1.98-1.12</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
