<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Merges the two separate sequence alignments in a database.  The aligned
sequences must have separate identifiers in the same table or be located in
different database tables."><title>Align Two Sets of Aligned Sequences in a Sequence Database — AlignDB • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Align Two Sets of Aligned Sequences in a Sequence Database — AlignDB"><meta property="og:description" content="Merges the two separate sequence alignments in a database.  The aligned
sequences must have separate identifiers in the same table or be located in
different database tables."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Align Two Sets of Aligned Sequences in a Sequence Database</h1>
      
      <div class="d-none name"><code>AlignDB.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Merges the two separate sequence alignments in a database.  The aligned
sequences must have separate identifiers in the same table or be located in
different database tables.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">AlignDB</span><span class="op">(</span></span>
<span>  <span class="va">dbFile</span>,</span>
<span>  tblName <span class="op">=</span> <span class="st">"Seqs"</span>,</span>
<span>  identifier <span class="op">=</span> <span class="st">""</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"DNAStringSet"</span>,</span>
<span>  add2tbl <span class="op">=</span> <span class="st">"Seqs"</span>,</span>
<span>  batchSize <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  perfectMatch <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  misMatch <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  gapOpening <span class="op">=</span> <span class="op">-</span><span class="fl">14</span>,</span>
<span>  gapExtension <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  gapPower <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  terminalGap <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  normPower <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  substitutionMatrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  processors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>dbFile</dt>
<dd><p>A SQLite connection object or a character string specifying
the path to the database file.</p></dd>


<dt>tblName</dt>
<dd><p>Character string specifying the table(s) where the sequences
are located.  If two <code>tblName</code>s are provided then the sequences in both
tables will be aligned.</p></dd>


<dt>identifier</dt>
<dd><p>Optional character string used to narrow the search
results to those matching a specific identifier.  If "" then all identifiers
are selected.  If two <code>identifier</code>s are provided then the set of
sequences matching each <code>identifier</code> will be aligned.</p></dd>


<dt>type</dt>
<dd><p>The type of <code>XStringSet</code> being processed.  This should be
(an abbreviation of) one of <code>"AAStringSet"</code>, <code>"DNAStringSet"</code>, or
<code>"RNAStringSet"</code>.</p></dd>


<dt>add2tbl</dt>
<dd><p>Character string specifying the table name in which to add
the aligned sequences.</p></dd>


<dt>batchSize</dt>
<dd><p>Integer specifying the number of sequences to process at a
time.</p></dd>


<dt>perfectMatch</dt>
<dd><p>Numeric giving the reward for aligning two matching
nucleotides in the alignment.  Only used when <code>type</code> is
<code>DNAStringSet</code> or <code>RNAStringSet</code>.</p></dd>


<dt>misMatch</dt>
<dd><p>Numeric giving the cost for aligning two mismatched
nucleotides in the alignment.  Only used when <code>type</code> is
<code>DNAStringSet</code> or <code>RNAStringSet</code>.</p></dd>


<dt>gapOpening</dt>
<dd><p>Numeric giving the cost for opening a gap in the
alignment.</p></dd>


<dt>gapExtension</dt>
<dd><p>Numeric giving the cost for extending an open gap in the
alignment.</p></dd>


<dt>gapPower</dt>
<dd><p>Numeric specifying the exponent to use in the gap cost
function.</p></dd>


<dt>terminalGap</dt>
<dd><p>Numeric giving the cost for allowing leading and trailing
gaps ("-" or "." characters) in the alignment.  Either two numbers, the
first for leading gaps and the second for trailing gaps, or a single number
for both.</p></dd>


<dt>normPower</dt>
<dd><p>Numeric giving the exponent that controls the degree of
normalization applied to scores by column occupancy.  If two numerics are
provided, the first controls the normalization power of terminal gaps, while
the second controls that of internal gaps.  A <code>normPower</code> of <code>0</code>
does not normalize the scores, which results in all columns of the profiles
being weighted equally, and is the optimal value for aligning fragmentary
sequences.  A <code>normPower</code> of <code>1</code> normalizes the score for aligning
two positions by their column occupancy (1 - fraction of gaps).  A
<code>normPower</code> greater than <code>1</code> more strongly discourages aligning
with ``gappy'' regions of the alignment.</p></dd>


<dt>standardize</dt>
<dd><p>Logical determining whether scores are standardized to be
in units of per matching site. Standardization effectively divides the score
of each possible alignment by its length so that scores are relative rather
than absolute.</p></dd>


<dt>substitutionMatrix</dt>
<dd><p>Either a substitution matrix representing the
substitution scores for an alignment or the name of the amino acid
substitution matrix to use in alignment.  The latter may be one of the
following: <code>BLOSUM45'', </code>BLOSUM50'', <code>BLOSUM62'', </code>BLOSUM80'',
<code>BLOSUM100'', </code>PAM30'', <code>PAM40'', </code>PAM70'', <code>PAM120'', </code>PAM250'', or
``MIQS''.  The default (NULL) will use the <code>perfectMatch</code> and
<code>misMatch</code> penalties for DNA/RNA or <code>PFASUM50</code> for AA.</p></dd>


<dt>processors</dt>
<dd><p>The number of processors to use, or <code>NULL</code> to
automatically detect and use all available processors.</p></dd>


<dt>verbose</dt>
<dd><p>Logical indicating whether to display progress.</p></dd>


<dt>...</dt>
<dd><p>Further arguments to be passed directly to <code><a href="Codec.html">Codec</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Returns the number of newly aligned sequences added to the database.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Sometimes it is useful to align two large sets of sequences, where each set
of sequences is already aligned but the two sets are not aligned to each
other.  <code>AlignDB</code> first builds a profile of each sequence set in
increments of <code>batchSize</code> so that the entire sequence set is not
required to fit in memory.  Next the two profiles are aligned using dynamic
programming.  Finally, the new alignment is applied to all the sequences as
they are incrementally added to the <code>add2tbl</code>.</p>
<p>Two <code>identifier</code>s or <code>tblName</code>s must be provided, indicating the
two sets of sequences to align.  The sequences corresponding to the first
<code>identifier</code> and <code>tblName</code> will be aligned to those of the second
<code>identifier</code> or <code>tblName</code>.  The aligned sequences are added to
<code>add2tbl</code> under a new identifier formed from the concatenation of the
two <code>identifier</code>s or <code>tblName</code>s.  (See examples section below.)</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Wright, E. S. (2015). DECIPHER: harnessing local sequence
context to improve protein multiple sequence alignment. BMC Bioinformatics,
16, 322. http://doi.org/10.1186/s12859-015-0749-z</p>
<p>Wright, E. S. (2020). RNAconTest: comparing tools for noncoding RNA multiple
sequence alignment based on structural consistency. RNA 2020, 26, 531-540.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="AlignProfiles.html">AlignProfiles</a></code>, <code><a href="AlignSeqs.html">AlignSeqs</a></code>,
<code><a href="AlignTranslation.html">AlignTranslation</a></code>, <code><a href="PFASUM.html">PFASUM</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.gen"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.fas"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Align two tables and place result into a third</span></span></span>
<span class="r-in"><span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu">SQLite</span><span class="op">(</span><span class="op">)</span>, <span class="st">":memory:"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">gen</span>, <span class="st">"GenBank"</span>, <span class="va">dbConn</span>, <span class="st">"Seqs1"</span>, tblName<span class="op">=</span><span class="st">"Set1"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading GenBank file chunk 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 175 total sequences in table Set1.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.05 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">fas</span>, <span class="st">"FASTA"</span>, <span class="va">dbConn</span>, <span class="st">"Seqs2"</span>, tblName<span class="op">=</span><span class="st">"Set2"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading FASTA file chunk 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 175 total sequences in table Set2.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.02 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">AlignDB</span><span class="op">(</span><span class="va">dbConn</span>, tblName<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Set1"</span>, <span class="st">"Set2"</span><span class="op">)</span>, add2tbl<span class="op">=</span><span class="st">"AlignedSets"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Added 350 aligned sequences to table AlignedSets with identifier 'Set1_Set2'.</span>
<span class="r-in"><span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="IdLengths.html">IdLengths</a></span><span class="op">(</span><span class="va">dbConn</span>, <span class="st">"AlignedSets"</span>, add2tbl<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Lengths counted for 350 sequences.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Added to AlignedSets:  "standard", "nonstandard", and "width".</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.01 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span>, tblName<span class="op">=</span><span class="st">"AlignedSets"</span><span class="op">)</span> <span class="co"># all sequences have the same width</span></span></span>
<span class="r-in"><span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Align two identifiers and place the result in the same table</span></span></span>
<span class="r-in"><span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu">SQLite</span><span class="op">(</span><span class="op">)</span>, <span class="st">":memory:"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">gen</span>, <span class="st">"GenBank"</span>, <span class="va">dbConn</span>, <span class="st">"Seqs1"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading GenBank file chunk 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 175 total sequences in table Seqs.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.02 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">fas</span>, <span class="st">"FASTA"</span>, <span class="va">dbConn</span>, <span class="st">"Seqs2"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading FASTA file chunk 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Added 175 new sequences to table Seqs.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 350 total sequences in table Seqs.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.02 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">AlignDB</span><span class="op">(</span><span class="va">dbConn</span>, identifier<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Seqs1"</span>, <span class="st">"Seqs2"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Added 350 aligned sequences to table Seqs with identifier 'Seqs1_Seqs2'.</span>
<span class="r-in"><span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="IdLengths.html">IdLengths</a></span><span class="op">(</span><span class="va">dbConn</span>, add2tbl<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Lengths counted for 700 sequences.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Added to Seqs:  "standard", "nonstandard", and "width".</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.02 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span> <span class="co"># note the sequences with a new identifier</span></span></span>
<span class="r-in"><span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

