<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Predicts the amplification efficiency of theoretical PCR products
(amplicons) given one or more primer sequences."><title>Simulate Amplification of DNA by PCR — AmplifyDNA • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate Amplification of DNA by PCR — AmplifyDNA"><meta property="og:description" content="Predicts the amplification efficiency of theoretical PCR products
(amplicons) given one or more primer sequences."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulate Amplification of DNA by PCR</h1>
      
      <div class="d-none name"><code>AmplifyDNA.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Predicts the amplification efficiency of theoretical PCR products
(amplicons) given one or more primer sequences.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">AmplifyDNA</span><span class="op">(</span></span>
<span>  <span class="va">primers</span>,</span>
<span>  <span class="va">myDNAStringSet</span>,</span>
<span>  <span class="va">maxProductSize</span>,</span>
<span>  <span class="va">annealingTemp</span>,</span>
<span>  <span class="va">P</span>,</span>
<span>  ions <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  includePrimers <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  minEfficiency <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>primers</dt>
<dd><p>A <code>DNAStringSet</code> object or character vector with one or
more unaligned primer sequences in 5' to 3' orientation.</p></dd>


<dt>myDNAStringSet</dt>
<dd><p>A <code>DNAStringSet</code> object or character vector with
unaligned template DNA sequences in 5' to 3' orientation.</p></dd>


<dt>maxProductSize</dt>
<dd><p>Integer specifying the maximum length of PCR products
(amplicons) in nucleotides.</p></dd>


<dt>annealingTemp</dt>
<dd><p>Numeric specifying the annealing temperature used in
the PCR reaction.</p></dd>


<dt>P</dt>
<dd><p>Numeric giving the molar concentration of primers in the reaction.</p></dd>


<dt>ions</dt>
<dd><p>Numeric giving the molar sodium equivalent ionic concentration.
Values may range between 0.01M and 1M.</p></dd>


<dt>includePrimers</dt>
<dd><p>Logical indicating whether to include the primer
sequences in the theoretical PCR products.  (See details section below.)</p></dd>


<dt>...</dt>
<dd><p>Additional arguments to be passed directly to
<code><a href="CalculateEfficiencyPCR.html">CalculateEfficiencyPCR</a></code>, including <code>batchSize</code>,
<code>taqEfficiency</code>, <code>maxDistance</code>, <code>maxGaps</code>, and
<code>processors</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code>DNAStringSet</code> object containing potential PCR products,
sorted from highest-to-lowest amplification efficiency.  The sequences are
named by their predicted amplification efficiency followed by the index of
each primer in <code>primers</code> and the name (or index if names are missing)
of the amplified sequence in <code>myDNAStringSet</code>.  (See examples section
below.)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Exponential amplification in PCR requires the annealing and elongation of
two primers from target sites on opposing strands of the template DNA.  If
the template DNA sequence (e.g., chromosome) is known then predictions of
theoretical amplicons can be obtained from in silico simulations of
amplification.  <code>AmplifyDNA</code> first searches for primer target sites on
the template DNA, and then calculates an amplification efficiency from each
target site using <code><a href="CalculateEfficiencyPCR.html">CalculateEfficiencyPCR</a></code>.  Ambiguity codes
(<code>IUPAC_CODE_MAP</code>) are supported in the <code>primers</code>, but not in
<code>myDNAStringSet</code> to prevent trivial matches (e.g., runs of N's).</p>
<p>If <code>taqEfficiency</code> is <code>TRUE</code> (the default), the amplification
efficiency of each primer is defined as the product of hybridization
efficiency and elongation efficiency.  Amplification efficiency must be at
least <code>minEfficiency</code> for a primer to be amplified in silico.  Overall
amplification efficiency of the PCR product is then calculated as the
geometric mean of the two (i.e., forward and reverse) primers' efficiencies.
Finally, amplicons are generated if the two primers are within
<code>maxProductSize</code> nucleotides downstream of each other.</p>
<p>Potential PCR products are returned, either with or without including the
primer sequences in the amplicon.  The default (<code>includePrimers=TRUE</code>)
is to incorporate the primer sequences as would normally occur during
amplification.  The alternative is to return the complete template sequence
including the target sites, which may not exactly match the primer
sequences.  Note that amplicons may be duplicated when different input
<code>primers</code> can amplify the same region of DNA.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>The program OligoArrayAux
(<a href="http://www.unafold.org/Dinamelt/software/oligoarrayaux.php" class="external-link">http://www.unafold.org/Dinamelt/software/oligoarrayaux.php</a>) must be
installed in a location accessible by the system.  For example, the
following code should print the installed OligoArrayAux version when
executed from the R console:</p>
<p><code>system("hybrid-min -V")</code></p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>ES Wright et al. (2013) "Exploiting Extension Bias in PCR to
Improve Primer Specificity in Ensembles of Nearly Identical DNA Templates."
Environmental Microbiology, doi:10.1111/1462-2920.12259.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="CalculateEfficiencyPCR.html">CalculateEfficiencyPCR</a></code>, <code><a href="DesignPrimers.html">DesignPrimers</a></code>,
<code><a href="DesignSignatures.html">DesignSignatures</a></code>, <code><a href="MeltDNA.html">MeltDNA</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">yeastSEQCHR1</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># not run (must have OligoArrayAux installed first):</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># match a single primer that acts as both the forward and reverse</span></span></span>
<span class="r-in"><span><span class="va">primer1</span> <span class="op">&lt;-</span> <span class="st">"TGGAAGCTGAAACG"</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="fu">AmplifyDNA</span><span class="op">(</span><span class="va">primer1</span>, <span class="va">yeastSEQCHR1</span>, annealingTemp<span class="op">=</span><span class="fl">55</span>, P<span class="op">=</span><span class="fl">4e-7</span>, maxProductSize<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># perform a typical amplification with two primer sequences:</span></span></span>
<span class="r-in"><span><span class="va">primer2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GGCTGTTGTTGGTGTT"</span>, <span class="st">"TGTCATCAGAACACCAA"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="fu">AmplifyDNA</span><span class="op">(</span><span class="va">primer2</span>, <span class="va">yeastSEQCHR1</span>, annealingTemp<span class="op">=</span><span class="fl">55</span>, P<span class="op">=</span><span class="fl">4e-7</span>, maxProductSize<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># perform a multiplex PCR amplification with multiple primers:</span></span></span>
<span class="r-in"><span><span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">primer1</span>, <span class="va">primer2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="fu">AmplifyDNA</span><span class="op">(</span><span class="va">primers</span>, <span class="va">yeastSEQCHR1</span>, annealingTemp<span class="op">=</span><span class="fl">55</span>, P<span class="op">=</span><span class="fl">4e-7</span>, maxProductSize<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

