<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Forms a consensus sequence representing a set of sequences."><title>Create a Consensus Sequence — ConsensusSequence • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a Consensus Sequence — ConsensusSequence"><meta property="og:description" content="Forms a consensus sequence representing a set of sequences."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Create a Consensus Sequence</h1>
      
      <div class="d-none name"><code>ConsensusSequence.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Forms a consensus sequence representing a set of sequences.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ConsensusSequence</span><span class="op">(</span></span>
<span>  <span class="va">myXStringSet</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ambiguity <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  noConsensusChar <span class="op">=</span> <span class="st">"+"</span>,</span>
<span>  minInformation <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">threshold</span>,</span>
<span>  ignoreNonBases <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  includeTerminalGaps <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>myXStringSet</dt>
<dd><p>An <code>AAStringSet</code>, <code>DNAStringSet</code>, or
<code>RNAStringSet</code> object of aligned sequences.</p></dd>


<dt>threshold</dt>
<dd><p>Numeric specifying that less than <code>threshold</code> fraction
of sequence information can be lost at any position of the consensus
sequence.</p></dd>


<dt>ambiguity</dt>
<dd><p>Logical specifying whether to consider ambiguity as split
between their respective nucleotides.  Degeneracy codes are specified in the
<code>IUPAC_CODE_MAP</code>.</p></dd>


<dt>noConsensusChar</dt>
<dd><p>Single character from the sequence's alphabet giving
the base to use when there is no consensus in a position.</p></dd>


<dt>minInformation</dt>
<dd><p>Minimum fraction of information required to form
consensus in each position.</p></dd>


<dt>ignoreNonBases</dt>
<dd><p>Logical specifying whether to count gap ("-"), mask
("+"), and unknown (".") characters towards the consensus.</p></dd>


<dt>includeTerminalGaps</dt>
<dd><p>Logical specifying whether or not to include
terminal gaps ("-" or "." characters on each end of the sequence) into the
formation of consensus.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>An <code>XStringSet</code> with a single consensus sequence matching the
input type.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>ConsensusSequence</code> removes the least frequent characters at each
position, so long as they represent less than <code>threshold</code> fraction of
the sequences in total.  If necessary, <code>ConsensusSequence</code> represents
the remaining characters using a degeneracy code from the
<code>IUPAC_CODE_MAP</code>.  Degeneracy codes are always used in cases where
multiple characters are equally abundant.</p>
<p>Two key parameters control the degree of consensus: <code>threshold</code> and
<code>minInformation</code>.  The default <code>threshold</code> (<code>0.05</code>) means
that at less than 5\<!-- % of sequences will not be represented by the consensus -->
sequence at any given position.  The default <code>minInformation</code> (<code>1
- 0.05</code>) specifies that at least 95\<!-- % of sequences must contain the -->
information in the consensus, otherwise the <code>noConsensusChar</code> is used.
This enables an alternative character (e.g., "+") to be substituted at
positions that would otherwise yield an ambiguity code.</p>
<p>If <code>ambiguity = TRUE</code> (the default) then degeneracy codes in
<code>myXStringSet</code> are split between their respective bases according to
the <code>IUPAC_CODE_MAP</code> for DNA/RNA and <code>AMINO_ACID_CODE</code> for AA.
For example, an <code>R'' in a \code{DNAStringSet} would count as half an </code>A''
and half a <code>G''.  If \code{ambiguity = FALSE} then degeneracy codes are not considered in forming the consensus.  For an \code{AAStringSet} input, the lack of degeneracy codes generally results in </code>X'' at positions with
mismatches, unless the <code>threshold</code> is set to a higher value than the
default.</p>
<p>If <code>includeNonBases = TRUE</code> (the default) then gap ("-"), mask ("+"),
and unknown (".") characters are counted towards the consensus, otherwise
they are omitted from calculation of the consensus.  Note that gap ("-") and
unknown (".") characters are treated interchangeably as gaps when forming
the consensus sequence.  For this reason, the consensus of a position with
all unknown (".") characters will be a gap ("-").  Also, note that if
consensus is formed between different length sequences then it will
represent only the longest sequences at the end.  For this reason the
consensus sequence is generally based on a sequence alignment such that all
of the sequences have equal lengths.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="Disambiguate.html">Disambiguate</a></code>, <code><a href="IdConsensus.html">IdConsensus</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.sqlite"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">db</span>, limit<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Search Expression:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> select row_names, sequence from _Seqs where row_names in (select row_names</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> from Seqs) limit 10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet of length: 10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span> <span class="co"># consensus at bottom</span></span></span>
<span class="r-in"><span><span class="fu"><a href="BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">dna</span>, threshold<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="co"># consensus at bottom</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># controlling the degree of consensus</span></span></span>
<span class="r-in"><span><span class="va">AAAT</span> <span class="op">&lt;-</span> <span class="fu">DNAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"T"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">AAAT</span><span class="op">)</span> <span class="co"># "W"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 <span style="color: #EEEEEE; background-color: #808080;">W</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">AAAT</span>, threshold<span class="op">=</span><span class="fl">0.3</span><span class="op">)</span> <span class="co"># "A"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 <span style="color: #000000; background-color: #FFAFAF;">A</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">AAAT</span>, threshold<span class="op">=</span><span class="fl">0.3</span>, minInformation<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span> <span class="co"># "+"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 +</span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">AAAT</span>, threshold<span class="op">=</span><span class="fl">0.3</span>, minInformation<span class="op">=</span><span class="fl">0.8</span>, noConsensusChar<span class="op">=</span><span class="st">"N"</span><span class="op">)</span> <span class="co"># "N"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 <span style="color: #EEEEEE; background-color: #B2B2B2;">N</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># switch between degenerate-based and majority-based consensus</span></span></span>
<span class="r-in"><span><span class="va">majority</span> <span class="op">&lt;-</span> <span class="fu">DNAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GTT"</span>, <span class="st">"GAA"</span>, <span class="st">"CTG"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">majority</span><span class="op">)</span> <span class="co"># degenerate-based</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     3 <span style="color: #EEEEEE; background-color: #808080;">SWD</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">majority</span>, threshold<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="co"># majority-based</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     3 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #EEEEEE; background-color: #808080;">D</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">majority</span>, threshold<span class="op">=</span><span class="fl">0.5</span>, minInformation<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     3 ++<span style="color: #EEEEEE; background-color: #808080;">D</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># behavior in the case of a tie</span></span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">DNAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># "W"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 <span style="color: #EEEEEE; background-color: #808080;">W</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">DNAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span><span class="op">)</span><span class="op">)</span>, threshold<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="co"># "W"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 <span style="color: #EEEEEE; background-color: #808080;">W</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">AAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># "X"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 <span style="color: #EEEEEE; background-color: #B2B2B2;">X</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">AAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span><span class="op">)</span><span class="op">)</span>, threshold<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="co"># "X"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 <span style="color: #EEEEEE; background-color: #B2B2B2;">X</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">AAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"I"</span>, <span class="st">"L"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># "J"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 <span style="color: #EEEEEE; background-color: #808080;">J</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">AAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"I"</span>, <span class="st">"L"</span><span class="op">)</span><span class="op">)</span>, threshold<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="co"># "J"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     1 <span style="color: #EEEEEE; background-color: #808080;">J</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># handling terminal gaps</span></span></span>
<span class="r-in"><span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu">DNAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ANGCT-"</span>,<span class="st">"-ACCT-"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">dna</span><span class="op">)</span> <span class="co"># "ANSCT-"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     6 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #EEEEEE; background-color: #B2B2B2;">N</span><span style="color: #EEEEEE; background-color: #808080;">S</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span>-</span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">dna</span>, includeTerminalGaps<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co"># "+NSCT-"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     6 +<span style="color: #EEEEEE; background-color: #B2B2B2;">N</span><span style="color: #EEEEEE; background-color: #808080;">S</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span>-</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># the "." character is treated is a "-"</span></span></span>
<span class="r-in"><span><span class="va">aa</span> <span class="op">&lt;-</span> <span class="fu">AAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ANQIH-"</span>, <span class="st">"ADELW."</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">aa</span><span class="op">)</span> <span class="co"># "ABZJX-"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     6 <span style="color: #000000; background-color: #FFD7D7;">A</span><span style="color: #EEEEEE; background-color: #808080;">BZJ</span><span style="color: #EEEEEE; background-color: #B2B2B2;">X</span>-</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># internal non-bases are included by default</span></span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">DNAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A-+.A"</span>, <span class="st">"AAAAA"</span><span class="op">)</span><span class="op">)</span>, noConsensusChar<span class="op">=</span><span class="st">"N"</span><span class="op">)</span> <span class="co"># "ANNNA"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     5 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #EEEEEE; background-color: #B2B2B2;">NNN</span><span style="color: #000000; background-color: #FFAFAF;">A</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">DNAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A-+.A"</span>, <span class="st">"AAAAA"</span><span class="op">)</span><span class="op">)</span>, ignoreNonBases<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co"># "AAAAA"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     5 <span style="color: #000000; background-color: #FFAFAF;">AAAAA</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># degeneracy codes in the input are considered by default</span></span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">DNAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AWNDA"</span>, <span class="st">"AAAAA"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># "AWNDA"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     5 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #EEEEEE; background-color: #808080;">W</span><span style="color: #EEEEEE; background-color: #B2B2B2;">N</span><span style="color: #EEEEEE; background-color: #808080;">D</span><span style="color: #000000; background-color: #FFAFAF;">A</span></span>
<span class="r-in"><span><span class="fu">ConsensusSequence</span><span class="op">(</span><span class="fu">DNAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AWNDA"</span>, <span class="st">"AAAAA"</span><span class="op">)</span><span class="op">)</span>, ambiguity<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="co"># "AAAAA"</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     width seq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1]     5 <span style="color: #000000; background-color: #FFAFAF;">AAAAA</span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

