<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Corrects the reading frame to mitigate the impact of frameshift errors
caused by insertions or deletions in unaligned nucleotide sequences."><title>Corrects Frameshift Errors In Protein Coding Sequences — CorrectFrameshifts • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Corrects Frameshift Errors In Protein Coding Sequences — CorrectFrameshifts"><meta property="og:description" content="Corrects the reading frame to mitigate the impact of frameshift errors
caused by insertions or deletions in unaligned nucleotide sequences."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Corrects Frameshift Errors In Protein Coding Sequences</h1>
      
      <div class="d-none name"><code>CorrectFrameshifts.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Corrects the reading frame to mitigate the impact of frameshift errors
caused by insertions or deletions in unaligned nucleotide sequences.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CorrectFrameshifts</span><span class="op">(</span></span>
<span>  <span class="va">myXStringSet</span>,</span>
<span>  <span class="va">myAAStringSet</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"indels"</span>,</span>
<span>  acceptDistance <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  rejectDistance <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  maxComparisons <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  gapOpening <span class="op">=</span> <span class="op">-</span><span class="fl">13</span>,</span>
<span>  gapExtension <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  frameShift <span class="op">=</span> <span class="op">-</span><span class="fl">15</span>,</span>
<span>  geneticCode <span class="op">=</span> <span class="va">GENETIC_CODE</span>,</span>
<span>  substitutionMatrix <span class="op">=</span> <span class="st">"PFASUM50"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  processors <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>myXStringSet</dt>
<dd><p>A <code>DNAStringSet</code> or <code>RNAStringSet</code> of
unaligned protein coding sequences in 5' to 3' orientation.</p></dd>


<dt>myAAStringSet</dt>
<dd><p>An <code>AAStringSet</code> of reference protein sequences.
Ideally this would consist of a small set of diverse amino acid sequences
belonging to the same group of protein coding sequences as
<code>myXStringSet</code>.</p></dd>


<dt>type</dt>
<dd><p>Character string indicating the type of result desired.  This
should be (an abbreviation of) one of <code>"indels"</code>, <code>"sequences"</code>,
or <code>"both"</code>.  (See details section below.)</p></dd>


<dt>acceptDistance</dt>
<dd><p>Numeric giving the maximum distance from a reference
sequence that is acceptable to skip any remaining comparisons.</p></dd>


<dt>rejectDistance</dt>
<dd><p>Numeric giving the maximum distance from a reference
sequence that is allowed when correcting frameshifts.  Sequences in
<code>myXStringSet</code> that are greater than <code>rejectDistance</code> from the
nearest reference sequence will only have their length trimmed from the
3'-end to a multiple of three nucleotides without any frameshift correction.</p></dd>


<dt>maxComparisons</dt>
<dd><p>The number of reference comparisons to make before
stopping the search for a closer reference sequence.</p></dd>


<dt>gapOpening</dt>
<dd><p>Numeric giving the cost for opening a gap between the
query and reference sequences.</p></dd>


<dt>gapExtension</dt>
<dd><p>Numeric giving the cost for extending an open gap
between the query and reference sequences.</p></dd>


<dt>frameShift</dt>
<dd><p>Numeric giving the cost for shifting between frames of the
query sequence.</p></dd>


<dt>geneticCode</dt>
<dd><p>Named character vector in the same format as
<code>GENETIC_CODE</code> (the default), which represents the standard genetic
code.</p></dd>


<dt>substitutionMatrix</dt>
<dd><p>Either a substitution matrix representing the
substitution scores for matching two amino acids or the name of the amino
acid substitution matrix.  The latter may be one of the following:
<code>BLOSUM45'', </code>BLOSUM50'', <code>BLOSUM62'', </code>BLOSUM80'', <code>BLOSUM100'', </code>PAM30'', <code>PAM40'', </code>PAM70'', <code>PAM120'', </code>PAM250'', <code>PFASUM50'' (the default), or </code>MIQS''.</p></dd>


<dt>verbose</dt>
<dd><p>Logical indicating whether to display progress.</p></dd>


<dt>processors</dt>
<dd><p>The number of processors to use, or <code>NULL</code> to
automatically detect and use all available processors.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>If <code>type</code> is <code>"indels"</code> then the returned object is a list
with the same length as <code>myXStringSet</code>.  Each element is a list with
four components:</p>
<dl><dt>"insertions"</dt>
<dd><p>Approximate positions of inserted
nucleotides, which could be removed to correct the reading frame, or excess
nucleotides at the 3'-end that make the length longer than a multiple of
three.</p></dd>
 <dt>"deletions"</dt>
<dd><p>Approximate positions of deleted nucleotides,
which could be added back to correct the reading frame.</p></dd>
 <dt>"distance"</dt>
<dd><p>The amino acid distance from the nearest reference sequence, between 0 and</p><ol><li><p></p></li>
</ol><p>"index" The integer index of the reference sequence that was
used for frame correction, or <code>0</code> if no reference sequence was within
<code>rejectDistance</code>.  Note that positions in <code>insertions</code> and
<code>deletions</code> are sometimes repeated to indicate that the same position
needs to be shifted successively more than once to correct the reading
frame.</p></dd>


</dl><p>If <code>type</code> is <code>"sequences"</code> then the returned object is an
<code>XStringSet</code> of the same type as the input (<code>myXStringSet</code>).
Nucleotides are added or deleted as necessary to correct for frameshifts.
The returned sequences all have a reading frame starting from position 1, so
that they can be translated directly.</p>


<p>If <code>type</code> is <code>"both"</code> then the returned object is a list with two
components: one for the <code>"indels"</code> and the other for the
<code>"sequences"</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Accurate translation of protein coding sequences can be greatly disrupted by
one or two nucleotide phase shifts that occasionally occur during DNA
sequencing.  These frameshift errors can potentially be corrected through
comparison with other unshifted protein sequences.  This function uses a set
of reference amino acid sequences (<code>AAStringSet</code>) to find and correct
frameshift errors in a set of nucleotide sequences (<code>myXStringSet</code>).
First, three frame translation of the nucleotide sequences is performed, and
the nearest reference sequence is selected.  Then the optimal reading frame
at each position is determined based on a variation of the Guan &amp; Uberbacher
(1996) method.  Putative insertions and/or deletions (indels) are returned
in the result, typically with close proximity to the true indel locations.
For a comparison of this method to others, see Wang et al. (2013).</p>
<p>If <code>type</code> is <code>"sequences"</code> or <code>"both"</code>, then frameshifts are
corrected by adding <code>N</code>'s and/or removing nucleotides.  Note that this
changes the nucleotide sequence, and the new sequence often has minor errors
because the exact location of the indel(s) cannot be determined.  However,
the original frameshifts that disrupted the entire downstream sequence are
reduced to local perturbations.  All of the returned nucleotide sequences
will have a reading frame starting from the first position.  This allows
direct translation, and in practice works well if there is a similar
reference <code>myAAStringSet</code> with the correct reading frame.  Hence it is
more important that <code>myAAStringSet</code> contain a wide variety of sequences
than it is that it contain a lot of sequences.</p>
<p>Multiple inputs control the time required for frameshift correction.  The
number of sequences in the reference set (<code>myAAStringSet</code>) will affect
the speed of the first search for similar sequences.  Assessing frameshifts
in the second step requires order <code>N*M</code> time, where <code>N</code> and
<code>M</code> are the lengths of the query (<code>myXStringSet</code>) and reference
sequences.  Two parameters control the number of assessments that are made
for each sequence: (1) <code>maxComparisons</code> determines the maximum number
of reference sequences to compare to each query sequence, and (2)
<code>acceptDist</code> defines the maximum distance between a query and reference
that is acceptable before continuing to the next query sequence.  A lower
value for <code>maxComparisons</code> or a higher value for <code>acceptDist</code> will
accelerate frameshift correction, potentially at the expense of some
accuracy.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Guan, X., &amp; Uberbacher, E. C. (1996). Alignments of DNA and
protein sequences containing frameshift errors. Computer Applications in the
Biosciences : CABIOS, <b>12(1)</b>, 31-40.</p>
<p>Wang, Q., et al. (2013). Ecological Patterns of nifH Genes in Four
Terrestrial Climatic Zones Explored with Targeted Metagenomics Using
FrameBot, a New Informatics Tool. mBio, <b>4(5)</b>, e00592-13-e00592-13.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="AlignTranslation.html">AlignTranslation</a></code>, <code><a href="OrientNucleotides.html">OrientNucleotides</a></code>,
<code><a href="PFASUM.html">PFASUM</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"50S_ribosomal_protein_L2.fas"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="va">fas</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># introduce artificial indels</span></span></span>
<span class="r-in"><span><span class="va">n_ins</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="co"># insertions per sequence</span></span></span>
<span class="r-in"><span><span class="va">shifted</span> <span class="op">&lt;-</span> <span class="fu">replaceAt</span><span class="op">(</span><span class="va">dna</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">dna</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="va">sample</span>,</span></span>
<span class="r-in"><span>    <span class="va">n_ins</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">DNA_BASES</span>,</span></span>
<span class="r-in"><span>    <span class="va">n_ins</span>,</span></span>
<span class="r-in"><span>    replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">n_dels</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># deletions per sequence</span></span></span>
<span class="r-in"><span><span class="va">shifted</span> <span class="op">&lt;-</span> <span class="fu">replaceAt</span><span class="op">(</span><span class="va">shifted</span>,</span></span>
<span class="r-in"><span>  <span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">shifted</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>      <span class="fu">IRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">x</span>,</span></span>
<span class="r-in"><span>          <span class="va">n_dels</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        width<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">}</span><span class="op">)</span>, <span class="st">"IRangesList"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># to make frameshift correction more challenging,</span></span></span>
<span class="r-in"><span><span class="co"># only supply 20 reference amino acid sequences</span></span></span>
<span class="r-in"><span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">CorrectFrameshifts</span><span class="op">(</span><span class="va">shifted</span>,</span></span>
<span class="r-in"><span>  <span class="fu">translate</span><span class="op">(</span><span class="va">dna</span><span class="op">[</span><span class="va">s</span><span class="op">]</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  type<span class="op">=</span><span class="st">"both"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Finding the closest reference amino acid sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.36 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Assessing frameshifts in nucleotide sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 8.03 secs</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># there was a wide range of distances</span></span></span>
<span class="r-in"><span><span class="co"># to the nearest reference sequence</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">`[`</span>, <span class="st">"distance"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0%         25%         50%         75%        100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.000000000 0.007272727 0.167883212 0.260073260 0.481848185 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># none of the sequences were &gt; rejectDistance</span></span></span>
<span class="r-in"><span><span class="co"># from the nearest reference sequence</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">`[`</span>, <span class="st">"index"</span><span class="op">)</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># the number of indels was generally correct</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">insertions</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">shifted</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0          1          2          4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.03470032 0.11671924 0.82965300 0.01892744 </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">deletions</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">shifted</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          0          1          2          3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.11671924 0.82965300 0.03470032 0.01892744 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># align and display the translations</span></span></span>
<span class="r-in"><span><span class="va">AA</span> <span class="op">&lt;-</span> <span class="fu"><a href="AlignTranslation.html">AlignTranslation</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">sequences</span>,</span></span>
<span class="r-in"><span>  readingFrame<span class="op">=</span><span class="fl">1</span>,</span></span>
<span class="r-in"><span>  type<span class="op">=</span><span class="st">"AAStringSet"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Determining distance matrix based on shared 5-mers:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.75 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Clustering into groups by similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.02 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Aligning Sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 1.13 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Iteration 1 of 2:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Determining distance matrix based on alignment:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.07 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reclustering into groups by similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.03 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Realigning Sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.9 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Iteration 2 of 2:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Determining distance matrix based on alignment:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.07 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reclustering into groups by similarity:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.03 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Realigning Sequences:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.04 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu"><a href="BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">AA</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

