<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Finds chimeras present in a database of sequences.  Makes use of a reference
database of (presumed to be) good quality sequences."><title>Find Chimeras in a Sequence Database — FindChimeras • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Find Chimeras in a Sequence Database — FindChimeras"><meta property="og:description" content="Finds chimeras present in a database of sequences.  Makes use of a reference
database of (presumed to be) good quality sequences."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Find Chimeras in a Sequence Database</h1>
      
      <div class="d-none name"><code>FindChimeras.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Finds chimeras present in a database of sequences.  Makes use of a reference
database of (presumed to be) good quality sequences.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">FindChimeras</span><span class="op">(</span></span>
<span>  <span class="va">dbFile</span>,</span>
<span>  tblName <span class="op">=</span> <span class="st">"Seqs"</span>,</span>
<span>  identifier <span class="op">=</span> <span class="st">""</span>,</span>
<span>  <span class="va">dbFileReference</span>,</span>
<span>  tblNameReference <span class="op">=</span> <span class="st">"Seqs"</span>,</span>
<span>  batchSize <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  minNumFragments <span class="op">=</span> <span class="fl">20000</span>,</span>
<span>  tb.width <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  multiplier <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  minLength <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  minCoverage <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  overlap <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  minSuspectFragments <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  showPercentCoverage <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  add2tbl <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  maxGroupSize <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  minGroupSize <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  excludeIDs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  processors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>dbFile</dt>
<dd><p>A SQLite connection object or a character string specifying
the path to the database file to be checked for chimeric sequences.</p></dd>


<dt>tblName</dt>
<dd><p>Character string specifying the table in which to check for
chimeras.</p></dd>


<dt>identifier</dt>
<dd><p>Optional character string used to narrow the search
results to those matching a specific identifier.  If "" then all identifiers
are selected.</p></dd>


<dt>dbFileReference</dt>
<dd><p>A SQLite connection object or a character string
specifying the path to the reference database file of (presumed to be) good
quality sequences.  A 16S reference database is available from
<a href="http://DECIPHER.codes" class="external-link">http://DECIPHER.codes</a>.</p></dd>


<dt>tblNameReference</dt>
<dd><p>Character string specifying the table with reference
sequences.</p></dd>


<dt>batchSize</dt>
<dd><p>Number sequences to tile with fragments at a time.</p></dd>


<dt>minNumFragments</dt>
<dd><p>Number of suspect fragments to accumulate before
searching through other groups.</p></dd>


<dt>tb.width</dt>
<dd><p>A single integer 1..14 giving the number of nucleotides at
the start of each fragment that are part of the trusted band.</p></dd>


<dt>multiplier</dt>
<dd><p>A single integer specifying the multiple of fragments
found out-of-group greater than fragments found in-group in order to
consider a sequence a chimera.</p></dd>


<dt>minLength</dt>
<dd><p>Minimum length of a chimeric region in order to be
considered as a chimera.</p></dd>


<dt>minCoverage</dt>
<dd><p>Minimum fraction of coverage necessary in a chimeric
region.</p></dd>


<dt>overlap</dt>
<dd><p>Number of nucleotides at the end of the sequence that the
chimeric region must overlap in order to be considered a chimera.</p></dd>


<dt>minSuspectFragments</dt>
<dd><p>Minimum number of suspect fragments belonging to
another group required to consider a sequence a chimera.</p></dd>


<dt>showPercentCoverage</dt>
<dd><p>Logical indicating whether to list the percent
coverage of suspect fragments in each chimeric region in the output.</p></dd>


<dt>add2tbl</dt>
<dd><p>Logical or a character string specifying the table name in
which to add the result.</p></dd>


<dt>maxGroupSize</dt>
<dd><p>Maximum number of sequences searched in a group.  A
value of less than 0 means the search is unlimited.</p></dd>


<dt>minGroupSize</dt>
<dd><p>The minimum number of sequences in a group to be
considered as part of the search for chimeras.  May need to be set to a
small value for reference databases with mostly small groups.</p></dd>


<dt>excludeIDs</dt>
<dd><p>Optional character vector of <code>identifier</code>(s) to
exclude from database searches, or <code>NULL</code> (the default) to not exclude
any.</p></dd>


<dt>processors</dt>
<dd><p>The number of processors to use, or <code>NULL</code> to
automatically detect and use all available processors.</p></dd>


<dt>verbose</dt>
<dd><p>Logical indicating whether to display progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code>data.frame</code> containing only the sequences that meet the
specifications for being chimeric.  The chimera column contains information
on the chimeric region and to which group it belongs.  The <code>row.names</code></p>


<p>of the <code>data.frame</code> correspond to those of the sequences in the
<code>dbFile</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>FindChimeras</code> works by finding suspect fragments that are uncommon in
the group where the sequence belongs, but very common in another group where
the sequence does not belong.  Each sequence in the <code>dbFile</code> is tiled
into short sequence segments called fragments.  If the fragments are
infrequent in their respective group in the <code>dbFileReference</code> then they
are considered suspect.  If enough suspect fragments from a sequence meet
the specified constraints then the sequence is flagged as a chimera.</p>
<p>The default parameters are optimized for full-length 16S sequences (&gt; 1,000
nucleotides).  Shorter 16S sequences require two parameters that are
different than the defaults: <code>minCoverage = 0.2</code>, and
<code>minSuspectFragments = 2</code>.</p>
<p>Groups are determined by the identifier present in each database.  For this
reason, the groups in the <code>dbFile</code> should exist in the groups of the
<code>dbFileReference</code>.  The reference database is assumed to contain many
sequences of only good quality.</p>
<p>If a reference database is not present then it is feasible to create a
reference database by using the input database as the reference database.
Removing chimeras from the reference database and then iteratively repeating
the process can result in a clean reference database.</p>
<p>For non-16S sequences it may be necessary to optimize the parameters for the
particular sequences.  The simplest way to perform an optimization is to
experiment with different input parameters on artificial chimeras such as
those created using <code><a href="CreateChimeras.html">CreateChimeras</a></code>.  Adjusting input
parameters until the maximum number of artificial chimeras are identified is
the easiest way to determine new defaults.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>ES Wright et al. (2012) "DECIPHER: A Search-Based Approach to
Chimera Identification for 16S rRNA Sequences." Applied and Environmental
Microbiology, doi:10.1128/AEM.06516-11.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="CreateChimeras.html">CreateChimeras</a></code>, <code><a href="Add2DB.html">Add2DB</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.sqlite"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># It is necessary to set dbFileReference to the file path of the</span></span></span>
<span class="r-in"><span><span class="co"># 16S reference database available from http://DECIPHER.codes</span></span></span>
<span class="r-in"><span><span class="va">chimeras</span> <span class="op">&lt;-</span> <span class="fu">FindChimeras</span><span class="op">(</span><span class="va">db</span>, dbFileReference<span class="op">=</span><span class="va">db</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ================================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No chimeras found.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.03 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

