<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Predicts 3-state protein secondary structure based on the primary (amino
acid) sequence using the GOR IV method (Garnier et al., 1996)."><title>Predict Protein Secondary Structure as Helix, Beta-Sheet, or Coil — PredictHEC • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Predict Protein Secondary Structure as Helix, Beta-Sheet, or Coil — PredictHEC"><meta property="og:description" content="Predicts 3-state protein secondary structure based on the primary (amino
acid) sequence using the GOR IV method (Garnier et al., 1996)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Predict Protein Secondary Structure as Helix, Beta-Sheet, or Coil</h1>
      
      <div class="d-none name"><code>PredictHEC.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Predicts 3-state protein secondary structure based on the primary (amino
acid) sequence using the GOR IV method (Garnier et al., 1996).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">PredictHEC</span><span class="op">(</span></span>
<span>  <span class="va">myAAStringSet</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"states"</span>,</span>
<span>  windowSize <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  background <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>H <span class="op">=</span> <span class="op">-</span><span class="fl">0.12</span>, E <span class="op">=</span> <span class="op">-</span><span class="fl">0.25</span>, C <span class="op">=</span> <span class="fl">0.23</span><span class="op">)</span>,</span>
<span>  HEC_MI1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  HEC_MI2 <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>myAAStringSet</dt>
<dd><p>An <code>AAStringSet</code> object of sequences.</p></dd>


<dt>type</dt>
<dd><p>Character string indicating the type of results desired.  This
should be (an unambiguous abbreviation of) one of <code>"states"</code>,
<code>"scores"</code>, or <code>"probabilities"</code>.</p></dd>


<dt>windowSize</dt>
<dd><p>Numeric specifying the number of residues to the left or
right of the center position to use in the prediction.</p></dd>


<dt>background</dt>
<dd><p>Numeric vector with the background ``scores'' for each of
the three states (H, E, and C).</p></dd>


<dt>HEC_MI1</dt>
<dd><p>An array of dimensions 20 x 21 x 3 giving the mutual
information for single residues.</p></dd>


<dt>HEC_MI2</dt>
<dd><p>An array of dimensions 20 x 20 x 21 x 21 x 3 giving the
mutual information for pairs of residues.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>If <code>type</code> is <code>"states"</code> (the default), then the output is
a character vector with the secondary structure assignment ("H", "E", or
"C") for each residue in <code>myAAStringSet</code>.</p>


<p>Otherwise, the output is a list with one element for each sequence in
<code>myAAStringSet</code>.  Each list element contains a matrix of dimension 3
(H, E, or C) by the number of residues in the sequence.  If <code>type</code> is
<code>"scores"</code>, then values in the matrix represent log-odds ``scores''.
If <code>type</code> is <code>"probabilities"</code> then the values represent the
normalized probabilities of the three states at a position.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The GOR (Garnier-Osguthorpe-Robson) method is an information-theory method
for prediction of secondary structure based on the primary sequence of a
protein.  Version IV of the method makes 3-state predictions based on the
mutual information contained in single residues and pairs of residues within
<code>windowSize</code> residues of the position being assigned.  This approach is
about 65\<!-- % accurate, and is one of the most accurate methods for assigning -->
secondary structure that only use a single sequence.  This implementation of
GOR IV does not use decision constants or the number of contiguous states
when assigning the final state.  Note that characters other than the
standard 20 amino acids are not assigned a state.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Garnier, J., Gibrat, J. F., &amp; Robson, B. (1996). GOR method for
predicting protein secondary structure from amino acid sequence.
<em>Methods in Enzymology</em>, <b>266</b>, 540-553.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="HEC_MI.html">HEC_MI1</a></code>, <code><a href="HEC_MI.html">HEC_MI2</a></code>,
<code><a href="PredictDBN.html">PredictDBN</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"50S_ribosomal_protein_L2.fas"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="va">fas</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">aa</span> <span class="op">&lt;-</span> <span class="fu">translate</span><span class="op">(</span><span class="va">dna</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">hec</span> <span class="op">&lt;-</span> <span class="fu">PredictHEC</span><span class="op">(</span><span class="va">aa</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hec</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "CCCCCCCCCCCCHHHHHHHCCCCCCCCCCCCEEEECCCCCCCCCCCCEEEEEECCCCCCEEEEEHHHCCCCCCHHHHHHHECCCCCCHHHEEEEECCCCCCEEEECCCCCCCCCEEEHHHHHHHHHCCCCCCCCCCCCCEECHHEEECCCCCEEEEECCCCEEEECCCCCHEEEHHCCCCCEEEECCCCEEECCCCCCCCHHHEHCCCCCCEEECCCCCCEEEEECCCCCCCCCCCCEECCCCCCECCCCECCCCCEEECCCCCCHEEHHCCC" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2] "CCCCECCCCCCCCCEEEEEEECEEECCCCCCHEEEEECCCCCCCCCCCEEEEEECCCCCCEEEEEEECCCCCCCCCEEEEEEECCCCCHHHEEHHHHCCCCEEEECCCCCCEEEEEECCCCCCHHHCCCCCCCCCCCCEEEEEEECCCCCCHEEEEHCCCHHEECCCCCCEEEEECCCCCCEEEEEECEEEEEECCCCCHHHHHHCCCCCEEEECCCCCCCEEEECCCCCCCCCCCCEECCCCCCCCHHHHHHCCCCCCCCCCCCHHHHHHHCC"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3] "CCCCECCCCCCCCCEEEEEEECEEECCCCCCHEEEEECCCCCCCCCCCEEEEEECCCCCCEEEEEEECCCCCCCCCEEEEEEECCCCCHHHEEHHHHCCCCEEEECCCCCCEEEEEECCCCCCHHHCCCCCCCCCCCCEEEEEEECCCCCCHEEEEHCCCHHEECCCCCCEEEEECCCCCCEEEEEECEEEEEECCCCCHHHHHHCCCCCEEEECCCCCCCEEEECCCCCCCCCCCCEECCCCCCCCHHHHHHCCCCCCCCCCCCHHHHHHHCC"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4] "CCCCECCCCCCCCCEEEEEEECEEECCCCCCHEEEEECCCCCCCCCCCEEEEEECCCCCCEEEEEEECCCCCCCCCEEEEEEECCCCCHHHEEHHHHCCCCEEEECCCCCCEEEEEECCCCCCHHHCCCCCCCCCCCCEEEEEEECCCCCCHEEEEHCCCHHEECCCCCCEEEEECCCCCCEEEEEECEEEEEECCCCCHHHHHHCCCCCEEEECCCCCCCEEEECCCCCCCCCCCCEECCCCCCCCHHHHHHCCCCCCCCCCCCHHHHHHHCC"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5] "CCEEEECCCCCCCEEEEEEECCCCCCCCCCCCEECCCCCCCCCCCCCEEEEEEECCCCCCEEEEHHHHCCCCCCHHHHHHHCCCCCCHHHHEEEEHCCCCHEEEECCCCCCCCCEEECCCCCCEECCCCCCCCCCCCCCEEEEEEECCCCCCEEEHHCCHHEEEEECCCCCEEEEHCCCHHHHHHHHHCEEECCCCCCHHHHHHHHHCCCCCECCCCCCEEECECCCCCCCCCCCCCCECCCCCCCCEEEECCCCCEECCCCCCEEEEECCCC" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6] "CCEEEECCCCCCCEEEEEEECCCCCCCCCCCCEECCCCCCCCCCCCCEEEEEEECCCCCCEEEEHHHHCCCCCCHHHHHHHCCCCCCHHHHEEEEHCCCCHEEEECCCCCCCCCEEECCCCCCEECCCCCCCCCCCCCCEEEEEEECCCCCCEEEHHCCHHEEEEECCCCCEEEEHCCCHHHHHHHHHCEEECCCCCCHHHHHHHHHCCCCCECCCCCCEEECECCCCCCCCCCCCCCECCCCCCCCEEEECCCCCEECCCCCCEEEEECCCC" </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

