<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Makes small adjustments by shifting groups of gaps left and right to find
their optimal positioning in a multiple sequence alignment."><title>Improve An Existing Alignment By Adjusting Gap Placements — AdjustAlignment • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Improve An Existing Alignment By Adjusting Gap Placements — AdjustAlignment"><meta property="og:description" content="Makes small adjustments by shifting groups of gaps left and right to find
their optimal positioning in a multiple sequence alignment."><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Improve An Existing Alignment By Adjusting Gap Placements</h1>
      
      <div class="d-none name"><code>AdjustAlignment.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Makes small adjustments by shifting groups of gaps left and right to find
their optimal positioning in a multiple sequence alignment.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">AdjustAlignment</span><span class="op">(</span></span>
<span>  <span class="va">myXStringSet</span>,</span>
<span>  perfectMatch <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  misMatch <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  gapLetter <span class="op">=</span> <span class="op">-</span><span class="fl">3</span>,</span>
<span>  gapOpening <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span>,</span>
<span>  gapExtension <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  substitutionMatrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shiftPenalty <span class="op">=</span> <span class="op">-</span><span class="fl">0.2</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  weight <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  processors <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>myXStringSet</dt>
<dd><p>An <code>AAStringSet</code>, <code>DNAStringSet</code>, or
<code>RNAStringSet</code> object of aligned sequences.</p></dd>


<dt>perfectMatch</dt>
<dd><p>Numeric giving the reward for aligning two matching
nucleotides in the alignment.  Only used for <code>DNAStringSet</code> or
<code>RNAStringSet</code> inputs.</p></dd>


<dt>misMatch</dt>
<dd><p>Numeric giving the cost for aligning two mismatched
nucleotides in the alignment.  Only used for <code>DNAStringSet</code> or
<code>RNAStringSet</code> inputs.</p></dd>


<dt>gapLetter</dt>
<dd><p>Numeric giving the cost for aligning gaps to letters.  A
lower value (more negative) encourages the overlapping of gaps across
different sequences in the alignment.</p></dd>


<dt>gapOpening</dt>
<dd><p>Numeric giving the cost for opening or closing a gap in
the alignment.</p></dd>


<dt>gapExtension</dt>
<dd><p>Numeric giving the cost for extending an open gap in the
alignment.</p></dd>


<dt>substitutionMatrix</dt>
<dd><p>Either a substitution matrix representing the
substitution scores for an alignment or the name of the amino acid
substitution matrix to use in alignment.  The latter may be one of the
following: <code>BLOSUM45'', </code>BLOSUM50'', <code>BLOSUM62'', </code>BLOSUM80'',
<code>BLOSUM100'', </code>PAM30'', <code>PAM40'', </code>PAM70'', <code>PAM120'', </code>PAM250'', or
``MIQS''.  The default (NULL) will use the <code>perfectMatch</code> and
<code>misMatch</code> penalties for DNA/RNA or <code>PFASUM50</code> for AA.</p></dd>


<dt>shiftPenalty</dt>
<dd><p>Numeric giving the cost for every additional position
that a group of gaps is shifted.</p></dd>


<dt>threshold</dt>
<dd><p>Numeric specifying the improvement in score required to
permanently apply an adjustment to the alignment.</p></dd>


<dt>weight</dt>
<dd><p>A numeric vector of weights for each sequence, or a single
number implying equal weights.</p></dd>


<dt>processors</dt>
<dd><p>The number of processors to use, or <code>NULL</code> to
automatically detect and use all available processors.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>An <code>XStringSet</code> of aligned sequences.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The process of multiple sequence alignment often results in the integration
of small imperfections into the final alignment.  Some of these errors are
obvious by-eye, which encourages manual refinement of automatically
generated alignments.  However, the manual refinement process is inherently
subjective and time consuming.  <code>AdjustAlignment</code> refines an existing
alignment in a process similar to that which might be applied manually, but
in a repeatable and must faster fashion.  This function shifts all of the
gaps in an alignment to the left and right to find their optimal
positioning.  The optimal position is defined as the position that maximizes
the alignment ``score'', which is determined by the input parameters.  The
resulting alignment will be similar to the input alignment but with many
imperfections eliminated.  Note that the affine gap penalties here are
different from the more flexible penalties used in
<code><a href="AlignProfiles.html">AlignProfiles</a></code>, and have been optimized independently.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Wright, E. S. (2015). DECIPHER: harnessing local sequence
context to improve protein multiple sequence alignment. BMC Bioinformatics,
16, 322. http://doi.org/10.1186/s12859-015-0749-z</p>
<p>Wright, E. S. (2020). RNAconTest: comparing tools for noncoding RNA multiple
sequence alignment based on structural consistency. RNA 2020, 26, 531-540.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="AlignSeqs.html">AlignSeqs</a></code>, <code><a href="AlignTranslation.html">AlignTranslation</a></code>,
<code><a href="PFASUM.html">PFASUM</a></code>, <code><a href="StaggerAlignment.html">StaggerAlignment</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># a trivial example</span></span></span>
<span class="r-in"><span><span class="va">aa</span> <span class="op">&lt;-</span> <span class="fu">AAStringSet</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ARN-PK"</span>, <span class="st">"ARRP-K"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">aa</span> <span class="co"># input alignment</span></span></span>
<span class="r-in"><span><span class="fu">AdjustAlignment</span><span class="op">(</span><span class="va">aa</span><span class="op">)</span> <span class="co"># output alignment</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># specifying an alternative substitution matrix</span></span></span>
<span class="r-in"><span><span class="fu">AdjustAlignment</span><span class="op">(</span><span class="va">aa</span>, substitutionMatrix<span class="op">=</span><span class="st">"BLOSUM62"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># a real example</span></span></span>
<span class="r-in"><span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Streptomyces_ITS_aligned.fas"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="va">fas</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dna</span> <span class="co"># input alignment</span></span></span>
<span class="r-in"><span><span class="va">adjustedDNA</span> <span class="op">&lt;-</span> <span class="fu">AdjustAlignment</span><span class="op">(</span><span class="va">dna</span><span class="op">)</span> <span class="co"># output alignment</span></span></span>
<span class="r-in"><span><span class="fu"><a href="BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">adjustedDNA</span>, highlight<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">adjustedDNA</span><span class="op">==</span><span class="va">dna</span> <span class="co"># most sequences were adjusted (those marked FALSE)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

