<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Aligns two sets of one or more aligned sequences by first generating
representative profiles, then aligning the profiles with dynamic
programming, and finally merging the two aligned sequence sets."><title>Align Two Sets of Aligned Sequences — AlignProfiles • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Align Two Sets of Aligned Sequences — AlignProfiles"><meta property="og:description" content="Aligns two sets of one or more aligned sequences by first generating
representative profiles, then aligning the profiles with dynamic
programming, and finally merging the two aligned sequence sets."><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Align Two Sets of Aligned Sequences</h1>
      
      <div class="d-none name"><code>AlignProfiles.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Aligns two sets of one or more aligned sequences by first generating
representative profiles, then aligning the profiles with dynamic
programming, and finally merging the two aligned sequence sets.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">AlignProfiles</span><span class="op">(</span></span>
<span>  <span class="va">pattern</span>,</span>
<span>  <span class="va">subject</span>,</span>
<span>  p.weight <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  s.weight <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  p.struct <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  s.struct <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  perfectMatch <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  misMatch <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  gapOpening <span class="op">=</span> <span class="op">-</span><span class="fl">14</span>,</span>
<span>  gapExtension <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  gapPower <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  terminalGap <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  restrict <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1000</span>, <span class="fl">2</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  anchor <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  normPower <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  substitutionMatrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  structureMatrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  processors <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>pattern</dt>
<dd><p>An <code>AAStringSet</code>, <code>DNAStringSet</code>, or
<code>RNAStringSet</code> object of aligned sequences to use as the pattern.</p></dd>


<dt>subject</dt>
<dd><p>A <code>XStringSet</code> object of aligned sequences to use as the
subject.  Must match the type of the pattern.</p></dd>


<dt>p.weight</dt>
<dd><p>A numeric vector of weights for each sequence in the pattern
to use in generating a profile, or a single number implying equal weights.</p></dd>


<dt>s.weight</dt>
<dd><p>A numeric vector of weights for each sequence in the subject
to use in generating a profile, or a single number implying equal weights.</p></dd>


<dt>p.struct</dt>
<dd><p>Either <code>NULL</code> (the default), a matrix, or a list of
matrices with one list element per sequence in the pattern.  (See details
section below.)</p></dd>


<dt>s.struct</dt>
<dd><p>Either <code>NULL</code> (the default), a matrix, or a list of
matrices with one list element per sequence in the subject.  (See details
section below.)</p></dd>


<dt>perfectMatch</dt>
<dd><p>Numeric giving the reward for aligning two matching
nucleotides in the alignment.  Only applicable for <code>DNAStringSet</code> or
<code>RNAStringSet</code> inputs.</p></dd>


<dt>misMatch</dt>
<dd><p>Numeric giving the cost for aligning two mismatched
nucleotides in the alignment.  Only applicable for <code>DNAStringSet</code> or
<code>RNAStringSet</code> inputs.</p></dd>


<dt>gapOpening</dt>
<dd><p>Numeric giving the cost for opening a gap in the
alignment.</p></dd>


<dt>gapExtension</dt>
<dd><p>Numeric giving the cost for extending an open gap in the
alignment.</p></dd>


<dt>gapPower</dt>
<dd><p>Numeric specifying the exponent to use in the gap cost
function.  (See details section below.)</p></dd>


<dt>terminalGap</dt>
<dd><p>Numeric giving the cost for allowing leading and trailing
gaps ("-" or "." characters) in the alignment.  Either two numbers, the
first for leading gaps and the second for trailing gaps, or a single number
for both.</p></dd>


<dt>restrict</dt>
<dd><p>Numeric vector of length three controlling the degree of
restriction around ridge lines in the dynamic programming matrix.  The first
element determines the span of the region around a ridge that is considered
during alignment.  The default (<code>-1000</code>) will align most inputs that
can reasonably be globally aligned without any loss in accuracy.  Input
sequences with high similarity could be more restricted (e.g., <code>-500</code>),
whereas a <code>pattern</code> and <code>subject</code> with little overlap should be
less restricted (e.g., <code>-10000</code>).  The second element sets the minimum
slope to either side of a ridge that is required to allow restriction at any
point.  The third element sets the minimum duration of the ridge required to
begin restricting the matrix around the ridge.  The duration of the ridge is
defined as the number of consecutive positions meeting the first two
conditions for restriction.  (See details section below.)</p></dd>


<dt>anchor</dt>
<dd><p>Numeric giving the fraction of sequences with identical k-mers
required to become an anchor point, or <code>NA</code> to not use anchors.
Alternatively, a matrix specifying anchor regions.  (See details section
below.)</p></dd>


<dt>normPower</dt>
<dd><p>Numeric giving the exponent that controls the degree of
normalization applied to scores by column occupancy.  If two numerics are
provided, the first controls the normalization power of terminal gaps, while
the second controls that of internal gaps.  A <code>normPower</code> of <code>0</code>
does not normalize the scores, which results in all columns of the profiles
being weighted equally, and is the optimal value for aligning fragmentary
sequences.  A <code>normPower</code> of <code>1</code> normalizes the score for aligning
two positions by their column occupancy (1 - fraction of gaps).  A
<code>normPower</code> greater than <code>1</code> more strongly discourages aligning
with ``gappy'' regions of the alignment.  (See details section below.)</p></dd>


<dt>standardize</dt>
<dd><p>Logical determining whether scores are standardized to be
in units of per matching site. Standardization effectively divides the score
of each possible alignment by its length so that scores are relative rather
than absolute.</p></dd>


<dt>substitutionMatrix</dt>
<dd><p>Either a substitution matrix representing the
substitution scores for an alignment or the name of the amino acid
substitution matrix to use in alignment.  The latter may be one of the
following: <code>BLOSUM45'', </code>BLOSUM50'', <code>BLOSUM62'', </code>BLOSUM80'',
<code>BLOSUM100'', </code>PAM30'', <code>PAM40'', </code>PAM70'', <code>PAM120'', </code>PAM250'', or
``MIQS''.  The default (NULL) will use the <code>perfectMatch</code> and
<code>misMatch</code> penalties for DNA/RNA or <code>PFASUM50</code> for AA.  (See
examples section below.)</p></dd>


<dt>structureMatrix</dt>
<dd><p>A structure matrix if <code>p.struct</code> and
<code>s.struct</code> are supplied, or <code>NULL</code> otherwise.</p></dd>


<dt>processors</dt>
<dd><p>The number of processors to use, or <code>NULL</code> to
automatically detect and use all available processors.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>An <code>XStringSet</code> of aligned sequences.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Profiles are aligned using dynamic programming, a variation of the
Needleman-Wunsch algorithm for global alignment.  The dynamic programming
method requires order <code>N*M</code> time and memory space where <code>N</code> and
<code>M</code> are the width of the pattern and subject.  This method works by
filling in a matrix of the possible ``alignment space'' by considering all
matches, insertions, and deletions between two sequence profiles.  The
highest scoring alignment is then used to add gaps to each of the input
sequence sets.</p>
<p>Heuristics can be useful to improve performance on long input sequences.
The <code>restrict</code> parameter can be used to dynamically constrain the
possible ``alignment space'' to only paths that will likely include the
final alignment, which in the best case can improve the speed from quadratic
time to nearly linear time.  The degree of restriction is important, and the
default value of <code>restrict</code> is reasonable in the vast majority of
cases.  It is also possible to prevent restriction by setting
<code>restrict</code> to such extreme values that these requirements will never be
met (e.g., <code>c(-1e10, 1e10, 1e10)</code>).</p>
<p>The argument <code>anchor</code> can be used to split the global alignment into
multiple sub-alignments.  This can greatly decrease the memory requirement
for long sequences when appropriate anchor points can be found.  Anchors are
15-mer (for DNA/RNA) or 7-mer (for AA) subsequences that are shared between
at least <code>anchor</code> fraction of <code>pattern</code>(s) and <code>subject</code>(s).
Anchored ranges are extended along the length of each sequence in a manner
designed to split the alignment into sub-alignments that can be separately
solved.  For most input sequences, the default anchoring has no effect on
accuracy, but anchoring can be disabled by setting <code>anchor=NA</code>.</p>
<p>Alternatively, <code>anchor</code> can be a matrix with <code>4</code> rows and one
column per anchor.  The first two rows correspond to the anchor start and
end positions in the pattern sequence(s), and the second two rows are the
equivalent anchor region in the subject sequence(s).  Anchors specified in
this manner must be strictly increasing (non-overlapping) in both sequences,
and have an anchor width of at least two positions.  Note that the anchors
do not have to be equal length, in which case the anchor regions will also
be aligned.  Manually splitting the alignment into more subtasks can
sometimes make it more efficient, but typically automatic anchoring is
effective.</p>
<p>The argument <code>normPower</code> determines how the distribution of information
is treated during alignment.  Higher values of <code>normPower</code> encourage
alignment between columns with higher occupancy (1 - fraction of gaps), and
de-emphasize the alignment of columns containing many gaps.  A
<code>normPower</code> of <code>0</code> will treat all columns equally regardless of
occupancy, which can be useful when the <code>pattern</code> or <code>subject</code>
contain many incomplete (fragment) sequences.  For example, <code>normPower</code>
should be set to <code>0</code> when aligning many short reads to a longer
reference sequence.</p>
<p>The arguments <code>p.struct</code> and <code>s.struct</code> may be used to provide
secondary structure probabilities in the form of a list containing matrices
or a single matrix.  If the input is a list, then each list element must
contain a matrix with dimensions <code>q*w</code>, where <code>q</code> is the number of
possible secondary structure states, and <code>w</code> is the width of the
unaligned pattern sequence.  Values in each matrix represent the probability
of the given state at that position in the sequence.  Alternatively, a
single matrix can be used as input if <code>w</code> is the width of the entire
<code>pattern</code> or <code>subject</code> alignment.  A <code>structureMatrix</code> must
be supplied along with the structures.  The functions <code>PredictHEC</code> and
<code>PredictDBN</code> can be used to predict secondary structure probabilities
in the format required by <code>AlignProfiles</code> (for amino acid and RNA
sequences, respectively).</p>
<p>The gap cost function is based on the observation that gap lengths are best
approximated by a Zipfian distribution (Chang &amp; Benner, 2004).  The cost of
inserting a gap of length <code>L</code> is equal to: <code>gapOpening +
gapExtension*sum(seq_len(L - 1)^gapPower)</code> when <code>L &gt; 1</code>, and
<code>gapOpen</code> when <code>L = 1</code>.  This function effectively penalizes
shorter gaps significantly more than longer gaps when <code>gapPower &lt; 0</code>,
and is equivalent to the affine gap penalty when <code>gapPower</code> is
<code>0</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Chang, M. S. S., &amp; Benner, S. A. (2004). Empirical Analysis of
Protein Insertions and Deletions Determining Parameters for the Correct
Placement of Gaps in Protein Sequence Alignments. Journal of Molecular
Biology, <b>341(2)</b>, 617-631.</p>
<p>Needleman, S., &amp; Wunsch, C. (1970). A general method applicable to the
search for similarities in the amino acid sequence of two proteins. Journal
of Molecular Biology, <b>48(3)</b>, 443-453.</p>
<p>Wright, E. S. (2015). DECIPHER: harnessing local sequence context to improve
protein multiple sequence alignment. BMC Bioinformatics, 16, 322.
http://doi.org/10.1186/s12859-015-0749-z</p>
<p>Wright, E. S. (2020). RNAconTest: comparing tools for noncoding RNA multiple
sequence alignment based on structural consistency. RNA 2020, 26, 531-540.</p>
<p>Yu, Y.-K., et al. (2015). Log-odds sequence logos. Bioinformatics, 31(3),
324-331. http://doi.org/10.1093/bioinformatics/btu634</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="AlignDB.html">AlignDB</a></code>, <code><a href="AlignSeqs.html">AlignSeqs</a></code>,
<code><a href="AlignSynteny.html">AlignSynteny</a></code>, <code><a href="AlignTranslation.html">AlignTranslation</a></code>,
<code><a href="PFASUM.html">PFASUM</a></code>, <code><a href="MIQS.html">MIQS</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># align two sets of sequences</span></span></span>
<span class="r-in"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.sqlite"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dna1</span> <span class="op">&lt;-</span> <span class="fu"><a href="SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">db</span>, remove<span class="op">=</span><span class="st">"common"</span>, limit<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="co"># the first 100 sequences</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Search Expression:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> select row_names, sequence from _Seqs where row_names in (select row_names</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> from Seqs) limit 100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet of length: 100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0.02 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">dna2</span> <span class="op">&lt;-</span> <span class="fu"><a href="SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">db</span>, remove<span class="op">=</span><span class="st">"common"</span>, limit<span class="op">=</span><span class="st">"100,100"</span><span class="op">)</span> <span class="co"># the rest</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Search Expression:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> select row_names, sequence from _Seqs where row_names in (select row_names</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> from Seqs) limit 100,100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet of length: 75</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 0 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">alignedDNA</span> <span class="op">&lt;-</span> <span class="fu">AlignProfiles</span><span class="op">(</span><span class="va">dna1</span>, <span class="va">dna2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">alignedDNA</span>, highlight<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># specify a DNA substitution matrix</span></span></span>
<span class="r-in"><span><span class="va">subMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>,</span></span>
<span class="r-in"><span>                    nrow<span class="op">=</span><span class="fl">4</span>, ncol<span class="op">=</span><span class="fl">4</span>,</span></span>
<span class="r-in"><span>                    dimnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">DNA_BASES</span>, <span class="va">DNA_BASES</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">subMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co"># perfectMatch</span></span></span>
<span class="r-in"><span><span class="va">alignedDNA.defaultSubM</span> <span class="op">&lt;-</span> <span class="fu">AlignProfiles</span><span class="op">(</span><span class="va">dna1</span>, <span class="va">dna2</span>, substitutionMatrix<span class="op">=</span><span class="va">subMatrix</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">alignedDNA.defaultSubM</span><span class="op">==</span><span class="va">alignedDNA</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># specify a different DNA substitution matrix</span></span></span>
<span class="r-in"><span><span class="va">subMatrix2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">12</span>, <span class="fl">3</span>, <span class="fl">5</span>, <span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">12</span>, <span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">5</span>, <span class="fl">3</span>, <span class="fl">11</span>, <span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">3</span>, <span class="fl">9</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    nrow<span class="op">=</span><span class="fl">4</span>, ncol<span class="op">=</span><span class="fl">4</span>,</span></span>
<span class="r-in"><span>                    dimnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">DNA_BASES</span>, <span class="va">DNA_BASES</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">alignedDNA.alterSubM</span> <span class="op">&lt;-</span> <span class="fu">AlignProfiles</span><span class="op">(</span><span class="va">dna1</span>, <span class="va">dna2</span>, substitutionMatrix<span class="op">=</span><span class="va">subMatrix2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">alignedDNA.alterSubM</span><span class="op">==</span><span class="va">alignedDNA</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># anchors are found automatically by default, but it is also</span></span></span>
<span class="r-in"><span><span class="co"># possible to specify anchor regions between the sequences</span></span></span>
<span class="r-in"><span><span class="va">anchors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">774</span>, <span class="fl">788</span>, <span class="fl">752</span>, <span class="fl">766</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">anchors</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [,1]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]  774</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]  788</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,]  752</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,]  766</span>
<span class="r-in"><span><span class="fu">subseq</span><span class="op">(</span><span class="va">dna1</span>, <span class="va">anchors</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">anchors</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 100:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       width seq                                             names               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ...   ... ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [96]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 96</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [97]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 97</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [98]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 98</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [99]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 99</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [100]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                 100</span>
<span class="r-in"><span><span class="fu">subseq</span><span class="op">(</span><span class="va">dna2</span>, <span class="va">anchors</span><span class="op">[</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">anchors</span><span class="op">[</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DNAStringSet object of length 75:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      width seq                                              names               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  101</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  102</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  103</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  104</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ...   ... ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [71]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  171</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [72]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  172</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [73]    15 <span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  173</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [74]    15 <span style="color: #000000; background-color: #AFFFAF;">CCC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  174</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [75]    15 <span style="color: #000000; background-color: #AFFFAF;">CCC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                  175</span>
<span class="r-in"><span><span class="va">alignedDNA2</span> <span class="op">&lt;-</span> <span class="fu">AlignProfiles</span><span class="op">(</span><span class="va">dna1</span>, <span class="va">dna2</span>, anchor<span class="op">=</span><span class="va">anchors</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

