<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Calculates the amplification efficiency of primers from their hybridization
efficiency and elongation efficiency at the target site."><title>Predict Amplification Efficiency of Primer Sequences — CalculateEfficiencyPCR • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Predict Amplification Efficiency of Primer Sequences — CalculateEfficiencyPCR"><meta property="og:description" content="Calculates the amplification efficiency of primers from their hybridization
efficiency and elongation efficiency at the target site."><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Predict Amplification Efficiency of Primer Sequences</h1>
      
      <div class="d-none name"><code>CalculateEfficiencyPCR.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Calculates the amplification efficiency of primers from their hybridization
efficiency and elongation efficiency at the target site.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CalculateEfficiencyPCR</span><span class="op">(</span></span>
<span>  <span class="va">primer</span>,</span>
<span>  <span class="va">target</span>,</span>
<span>  <span class="va">temp</span>,</span>
<span>  <span class="va">P</span>,</span>
<span>  <span class="va">ions</span>,</span>
<span>  batchSize <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  taqEfficiency <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  maxDistance <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  maxGaps <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  processors <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>primer</dt>
<dd><p>A <code>DNAStringSet</code> object or character vector with
unaligned primer sequences in 5' to 3' orientation.</p></dd>


<dt>target</dt>
<dd><p>A <code>DNAStringSet</code> object or character vector with
unaligned target or non-target sequences in 5' to 3' orientation.</p></dd>


<dt>temp</dt>
<dd><p>Numeric specifying the annealing temperature used in the PCR
reaction.</p></dd>


<dt>P</dt>
<dd><p>Numeric giving the molar concentration of primers in the reaction.</p></dd>


<dt>ions</dt>
<dd><p>Numeric giving the molar sodium equivalent ionic concentration.
Values may range between 0.01M and 1M.</p></dd>


<dt>batchSize</dt>
<dd><p>Integer specifying the number of primers to simulate
hybridization per batch.  See the Description section below.</p></dd>


<dt>taqEfficiency</dt>
<dd><p>Logical determining whether to make use of elongation
efficiency and maxDistance to increase predictive accuracy for <em>Taq</em>
DNA Polymerase amplifying primers with mismatches near the 3' terminus.
Note that this should be set to FALSE if using a high-fidelity polymerase
with 3' to 5' exonuclease activity.</p></dd>


<dt>maxDistance</dt>
<dd><p>Numeric specifying the maximal fraction of mismatched
base pairings on a rolling basis beginning from the 3' end of the primer.
Only used if <code>taqEfficiency</code> is <code>TRUE</code>.</p></dd>


<dt>maxGaps</dt>
<dd><p>Integer specifying the maximum number of insertions or
deletions (indels) in the primer/target alignment.  Only used if
<code>taqEfficiency</code> is <code>TRUE</code>.</p></dd>


<dt>processors</dt>
<dd><p>The number of processors to use, or <code>NULL</code> to
automatically detect and use all available processors.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A vector of predicted efficiencies for amplifying each
<code>primer</code>/<code>target</code> pair of sequences.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Amplification of pairwise <code>primer</code>/<code>target</code> pairs is simulated
<em>in silico</em>.  A complex model of hybridization is employed that takes
into account the side reactions resulting from probe-folding,
target-folding, and primer-dimer formation.  The resulting hybridization
efficiency is multiplied by the elongation efficiency to predict the overall
efficiency of amplification.</p>
<p>Free energy is obtained from system calls to OligoArrayAux, which must be
properly installed (see the Notes section below).  Primer/target pairs are
sent to OligoArrayAux in batches of <code>batchSize</code>, which prevents systems
calls from being too many characters.  Note that OligoArrayAux does not
support degeneracy codes (non-base letters), although they are accepted
without error.  Any sequences with ambiguity should be expanded into
multiple permutations with <code><a href="Disambiguate.html">Disambiguate</a></code> before input.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>The program OligoArrayAux
(<a href="http://www.unafold.org/Dinamelt/software/oligoarrayaux.php" class="external-link">http://www.unafold.org/Dinamelt/software/oligoarrayaux.php</a>) must be
installed in a location accessible by the system.  For example, the
following code should print the installed OligoArrayAux version when
executed from the R console:</p>
<p><code>system("hybrid-min -V")</code></p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>ES Wright et al. (2013) "Exploiting Extension Bias in PCR to
Improve Primer Specificity in Ensembles of Nearly Identical DNA Templates."
Environmental Microbiology, doi:10.1111/1462-2920.12259.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="AmplifyDNA.html">AmplifyDNA</a></code>, <code><a href="DesignPrimers.html">DesignPrimers</a></code>,
<code><a href="DesignSignatures.html">DesignSignatures</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AAAAACGGGGAGCGGGGGG"</span>, <span class="st">"AAAAACTCAACCCGAGGAGCGCGT"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu">reverseComplement</span><span class="op">(</span><span class="fu">DNAStringSet</span><span class="op">(</span><span class="va">primers</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># not run (must have OligoArrayAux installed first):</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="fu">CalculateEfficiencyPCR</span><span class="op">(</span><span class="va">primers</span>, <span class="va">targets</span>, temp<span class="op">=</span><span class="fl">75</span>, P<span class="op">=</span><span class="fl">4e-7</span>, ions<span class="op">=</span><span class="fl">0.225</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

