<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Aids the design of pairs of primers for amplifying a unique ``signature''
from each group of sequences.  Signatures are distinct PCR products that can
be differentiated by their length, melt temperature, or sequence."><title>Design PCR Primers for Amplifying Group-Specific Signatures — DesignSignatures • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Design PCR Primers for Amplifying Group-Specific Signatures — DesignSignatures"><meta property="og:description" content="Aids the design of pairs of primers for amplifying a unique ``signature''
from each group of sequences.  Signatures are distinct PCR products that can
be differentiated by their length, melt temperature, or sequence."><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Design PCR Primers for Amplifying Group-Specific Signatures</h1>
      
      <div class="d-none name"><code>DesignSignatures.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Aids the design of pairs of primers for amplifying a unique ``signature''
from each group of sequences.  Signatures are distinct PCR products that can
be differentiated by their length, melt temperature, or sequence.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DesignSignatures</span><span class="op">(</span></span>
<span>  <span class="va">dbFile</span>,</span>
<span>  tblName <span class="op">=</span> <span class="st">"Seqs"</span>,</span>
<span>  identifier <span class="op">=</span> <span class="st">""</span>,</span>
<span>  focusID <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"melt"</span>,</span>
<span>  resolution <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  levels <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  enzymes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minLength <span class="op">=</span> <span class="fl">17</span>,</span>
<span>  maxLength <span class="op">=</span> <span class="fl">26</span>,</span>
<span>  maxPermutations <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  annealingTemp <span class="op">=</span> <span class="fl">64</span>,</span>
<span>  P <span class="op">=</span> <span class="fl">4e-07</span>,</span>
<span>  monovalent <span class="op">=</span> <span class="fl">0.07</span>,</span>
<span>  divalent <span class="op">=</span> <span class="fl">0.003</span>,</span>
<span>  dNTPs <span class="op">=</span> <span class="fl">8e-04</span>,</span>
<span>  minEfficiency <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  ampEfficiency <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  numPrimerSets <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  minProductSize <span class="op">=</span> <span class="fl">70</span>,</span>
<span>  maxProductSize <span class="op">=</span> <span class="fl">400</span>,</span>
<span>  kmerSize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  searchPrimers <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  maxDictionary <span class="op">=</span> <span class="fl">20000</span>,</span>
<span>  primerDimer <span class="op">=</span> <span class="fl">1e-07</span>,</span>
<span>  pNorm <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  taqEfficiency <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  processors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>dbFile</dt>
<dd><p>A SQLite connection object or a character string specifying
the path to the database file.</p></dd>


<dt>tblName</dt>
<dd><p>Character string specifying the table where the DNA sequences
are located.</p></dd>


<dt>identifier</dt>
<dd><p>Optional character string used to narrow the search
results to those matching a specific identifier.  Determines the target
group(s) for which primers will be designed.  If "" then all identifiers are
selected.</p></dd>


<dt>focusID</dt>
<dd><p>Optional character string specifying which of the
<code>identifier</code>s will be used in the initial step of designing primers.
If <code>NA</code> (the default), then the <code>identifier</code> with the most
sequence information is used as the <code>focusID</code>.</p></dd>


<dt>type</dt>
<dd><p>Character string indicating the type of signature being used to
differentiate the PCR products from each group.  This should be (an
abbreviation of) one of <code>"melt"</code>, <code>"length"</code>, or
<code>"sequence"</code>.</p></dd>


<dt>resolution</dt>
<dd><p>Numeric specifying the ``resolution'' of the experiment,
or a vector giving the boundaries of bins.  (See details section below.)</p></dd>


<dt>levels</dt>
<dd><p>Numeric giving the number of ``levels'' that can be
distinguished in each bin.  (See details section below.)</p></dd>


<dt>enzymes</dt>
<dd><p>Named character vector providing the cut sites of one or more
restriction enzymes.  Cut sites must be delineated in the same format as
<code><a href="RESTRICTION_ENZYMES.html">RESTRICTION_ENZYMES</a></code>.</p></dd>


<dt>minLength</dt>
<dd><p>Integer providing the minimum length of primers to consider
in the design.</p></dd>


<dt>maxLength</dt>
<dd><p>Integer providing the maximum length of primers to consider
in the design.</p></dd>


<dt>maxPermutations</dt>
<dd><p>Integer providing the maximum number of permutations
allowed in a forward or reverse primer to attain greater coverage of
sequences.</p></dd>


<dt>annealingTemp</dt>
<dd><p>Numeric indicating the desired annealing temperature
that will be used in the PCR experiment.</p></dd>


<dt>P</dt>
<dd><p>Numeric giving the molar concentration of primers in the reaction.</p></dd>


<dt>monovalent</dt>
<dd><p>The molar concentration of monovalent (Na and K) ions
in solution that will be used to determine a sodium equivalent
concentration.</p></dd>


<dt>divalent</dt>
<dd><p>The molar concentration of divalent (Mg) ions in solution
that will be used to determine a sodium equivalent concentration.</p></dd>


<dt>dNTPs</dt>
<dd><p>Numeric giving the molar concentration of free nucleotides
added to the solution that will be used to determine a sodium equivalent
concentration.</p></dd>


<dt>minEfficiency</dt>
<dd><p>Numeric giving the minimum efficiency of hybridization
desired for the primer set.</p></dd>


<dt>ampEfficiency</dt>
<dd><p>Numeric giving the minimum efficiency required for
theoretical amplification of the primers.  Note that <code>ampEfficiency</code>
must be less than or equal to <code>minEfficiency</code>.  Lower values of
<code>ampEfficiency</code> will allow for more PCR products, although very low
values are unrealistic experimentally.</p></dd>


<dt>numPrimerSets</dt>
<dd><p>Integer giving the optimal number of primer sets
(forward and reverse primer sets) to design.</p></dd>


<dt>minProductSize</dt>
<dd><p>Integer giving the minimum number of nucleotides
desired in the PCR product.</p></dd>


<dt>maxProductSize</dt>
<dd><p>Integer giving the maximum number of nucleotides
desired in the PCR product.</p></dd>


<dt>kmerSize</dt>
<dd><p>Integer giving the size of k-mers to use in the preliminary
search for potential primers.</p></dd>


<dt>searchPrimers</dt>
<dd><p>Numeric specifying the number of forward and reverse
primers to use in searching for potential PCR products.  A lower value will
result in a faster search, but potentially neglect some useful primers.</p></dd>


<dt>maxDictionary</dt>
<dd><p>Numeric giving the maximum number of primers to search
for simultaneously in any given step.</p></dd>


<dt>primerDimer</dt>
<dd><p>Numeric giving the maximum amplification efficiency of
potential primer-dimer products.</p></dd>


<dt>pNorm</dt>
<dd><p>Numeric specifying the power (p &gt; 0) used in calculating the
\(L\textsuperscript{p}\)-norm when scoring primer pairs.  By default (p =
1), the score is equivalent to the average difference between pairwise
signatures.  When p &lt; 1, many small differences will be preferred over fewer
large differences, and vise-versa when p &gt; 1.  This enables prioritizing
primer pairs that will yield a greater number of unique signatures (p &lt; 1),
or fewer distinct, but more dissimilar, signatures (p &gt; 1).</p></dd>


<dt>taqEfficiency</dt>
<dd><p>Logical determining whether to make use of elongation
efficiency to increase predictive accuracy for <em>Taq</em> DNA Polymerase
amplifying primers with mismatches near the 3' terminus.  Note that this
should be set to FALSE if using a high-fidelity polymerase with 3' to 5'
exonuclease activity.</p></dd>


<dt>processors</dt>
<dd><p>The number of processors to use, or <code>NULL</code> to
automatically detect and use all available processors.</p></dd>


<dt>verbose</dt>
<dd><p>Logical indicating whether to display progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code>data.frame</code> with the top-scoring pairs of forward and
reverse primers, their score, the total number of PCR products, and
associated columns for the restriction enzyme (if <code>enzyme</code> is not
<code>NULL</code>).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Signatures are group-specific PCR products that can be differentiated by
either their melt temperature profile, length, or sequence.
<code>DesignSignatures</code> assists in finding the optimal pair of forward and
reverse primers for obtaining a distinguishable signature from each group of
sequences.  Groups are delineated by their unique <code>identifier</code> in the
database.  The algorithm works by progressively narrowing the search for
optimal primers: (1) the most frequent k-mers are found; (2) these are used
to design primers initially matching the <code>focusID</code> group; (3) the most
common forward and reverse primers are selected based on all of the groups,
and ambiguity is added up to <code>maxPermutations</code>; (4) a final search is
performed to find the optimal forward and reverse primer.  Pairs of primers
are scored by the distance between the signatures generated for each group,
which depends on the <code>type</code> of experiment.</p>
<p>The arguments <code>resolution</code> and <code>levels</code> control the theoretical
resolving power of the experiment.  The <code>signature</code> for a group is
discretized or grouped into <code>bins'' each with a certain magnitude of the signal.  Here \code{resolution} determines the separation between distinguishable </code>bins'', and <code>levels</code> controls the range of values in
each bin.  A high-accuracy experiment would have many bins and/or many
levels.  While <code>levels</code> is interpreted similarly for every <code>type</code>
of experiment, <code>resolution</code> is treated differently depending on
<code>type</code>.  If <code>type</code> is <code>"melt"</code>, then <code>resolution</code> can be
either a vector of different melt temperatures, or a single number giving
the change in temperatures that can be differentiated.  A high-resolution
melt (HRM) assay would typically have a resolution between 0.25 and 1 degree
Celsius.  If <code>type</code> is <code>"length"</code> then resolution is either the
number of bins between the <code>minProductSize</code> and <code>maxProductSize</code>,
or the bin boundaries.  For example, <code>resolution</code> can be lower (wider
bins) at long lengths, and higher (narrower bins) at shorter lengths.  If
<code>type</code> is <code>"sequence"</code> then <code>resolution</code> sets the k-mer size
used in differentiating amplicons.  Oftentimes, 4 to 6-mers are used for the
classification of amplicons.</p>
<p>The signatures can be diversified by using a restriction enzyme to digest
the PCR products when <code>type</code> is <code>"melt"</code> or <code>"length"</code>.  If
<code>enzymes</code> are supplied then the an additional search is made to find
the best enzyme to use with each pair of primers.  In this case, the output
includes all of the primer pairs, as well as any <code>enzymes</code> that will
digest the PCR products of that primer pair.  The output is re-scored to
rank the top primer pair and enzyme combination.  Note that <code>enzymes</code>
is inapplicable when <code>type</code> is <code>"sequence"</code> because restriction
enzymes do not alter the sequence of the DNA.  Also, it is recommended that
only a subset of the available <code>RESTRICTION_ENZYMES</code> are used as input
<code>enzymes</code> in order to accelerate the search for the best enzyme.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Wright, E.S. &amp; Vetsigian, K.H. (2016) "DesignSignatures: a tool
for designing primers that yields amplicons with distinct signatures."
Bioinformatics, doi:10.1093/bioinformatics/btw047.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="AmplifyDNA.html">AmplifyDNA</a></code>, <code><a href="CalculateEfficiencyPCR.html">CalculateEfficiencyPCR</a></code>,
<code><a href="DesignPrimers.html">DesignPrimers</a></code>, <code><a href="DigestDNA.html">DigestDNA</a></code>,
<code><a href="Disambiguate.html">Disambiguate</a></code>, <code><a href="MeltDNA.html">MeltDNA</a></code>,
<code><a href="RESTRICTION_ENZYMES.html">RESTRICTION_ENZYMES</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># below are suggested inputs for different types of experiments</span></span></span>
<span class="r-in"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.sqlite"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># High Resolution Melt (HRM) assay:</span></span></span>
<span class="r-in"><span><span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu">DesignSignatures</span><span class="op">(</span><span class="va">db</span>,</span></span>
<span class="r-in"><span>           resolution<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">100</span>, <span class="fl">0.25</span><span class="op">)</span>, <span class="co"># degrees Celsius</span></span></span>
<span class="r-in"><span>           minProductSize<span class="op">=</span><span class="fl">55</span>, <span class="co"># base pairs</span></span></span>
<span class="r-in"><span>           maxProductSize<span class="op">=</span><span class="fl">400</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Primers for next-generation sequencing:</span></span></span>
<span class="r-in"><span><span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu">DesignSignatures</span><span class="op">(</span><span class="va">db</span>,</span></span>
<span class="r-in"><span>           type<span class="op">=</span><span class="st">"sequence"</span>,</span></span>
<span class="r-in"><span>           minProductSize<span class="op">=</span><span class="fl">300</span>, <span class="co"># base pairs</span></span></span>
<span class="r-in"><span>           maxProductSize<span class="op">=</span><span class="fl">700</span>,</span></span>
<span class="r-in"><span>           resolution<span class="op">=</span><span class="fl">5</span>, <span class="co"># 5-mers</span></span></span>
<span class="r-in"><span>           levels<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Primers for community fingerprinting:</span></span></span>
<span class="r-in"><span><span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu">DesignSignatures</span><span class="op">(</span><span class="va">db</span>,</span></span>
<span class="r-in"><span>           type<span class="op">=</span><span class="st">"length"</span>,</span></span>
<span class="r-in"><span>           levels<span class="op">=</span><span class="fl">2</span>, <span class="co"># presence/absence</span></span></span>
<span class="r-in"><span>           minProductSize<span class="op">=</span><span class="fl">200</span>, <span class="co"># base pairs</span></span></span>
<span class="r-in"><span>           maxProductSize<span class="op">=</span><span class="fl">1400</span>,</span></span>
<span class="r-in"><span>           resolution<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">700</span>, <span class="fl">3</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                        <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">705</span>, <span class="fl">1000</span>, <span class="fl">5</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                        <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1010</span>, <span class="fl">1400</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Primers for restriction fragment length polymorphism (RFLP):</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">RESTRICTION_ENZYMES</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">myEnzymes</span> <span class="op">&lt;-</span> <span class="va">RESTRICTION_ENZYMES</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EcoRI"</span>, <span class="st">"HinfI"</span>, <span class="st">"SalI"</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu">DesignSignatures</span><span class="op">(</span><span class="va">db</span>,</span></span>
<span class="r-in"><span>           type<span class="op">=</span><span class="st">"length"</span>,</span></span>
<span class="r-in"><span>           levels<span class="op">=</span><span class="fl">2</span>, <span class="co"># presence/absence</span></span></span>
<span class="r-in"><span>           minProductSize<span class="op">=</span><span class="fl">200</span>, <span class="co"># base pairs</span></span></span>
<span class="r-in"><span>           maxProductSize<span class="op">=</span><span class="fl">600</span>,</span></span>
<span class="r-in"><span>           resolution<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">100</span>, <span class="fl">3</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                        <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">105</span>, <span class="fl">200</span>, <span class="fl">5</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                        <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">210</span>, <span class="fl">600</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>           enzymes<span class="op">=</span><span class="va">myEnzymes</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

