<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Assists in the design of single or dual probes targeting a specific group of
sequences while minimizing the potential to cross-hybridize with other
groups of sequences."><title>Design FISH Probes Targeting a Specific Group of Sequences — DesignProbes • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Design FISH Probes Targeting a Specific Group of Sequences — DesignProbes"><meta property="og:description" content="Assists in the design of single or dual probes targeting a specific group of
sequences while minimizing the potential to cross-hybridize with other
groups of sequences."><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Design FISH Probes Targeting a Specific Group of Sequences</h1>
      
      <div class="d-none name"><code>DesignProbes.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Assists in the design of single or dual probes targeting a specific group of
sequences while minimizing the potential to cross-hybridize with other
groups of sequences.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DesignProbes</span><span class="op">(</span></span>
<span>  <span class="va">tiles</span>,</span>
<span>  identifier <span class="op">=</span> <span class="st">""</span>,</span>
<span>  start <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  end <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minLength <span class="op">=</span> <span class="fl">17</span>,</span>
<span>  maxLength <span class="op">=</span> <span class="fl">26</span>,</span>
<span>  maxPermutations <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  minCoverage <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  minGroupCoverage <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  hybTemp <span class="op">=</span> <span class="fl">46</span>,</span>
<span>  P <span class="op">=</span> <span class="fl">2.5e-07</span>,</span>
<span>  Na <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  FA <span class="op">=</span> <span class="fl">35</span>,</span>
<span>  minEfficiency <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  worstScore <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,</span>
<span>  numProbeSets <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  batchSize <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  target <span class="op">=</span> <span class="st">"SSU"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>tiles</dt>
<dd><p>A set of tiles representing each group of sequences, as in the
format created by the function <code>TileSeqs</code>.</p></dd>


<dt>identifier</dt>
<dd><p>Optional character string used to narrow the search
results to those matching a specific identifier.  Determines the target
group(s) for which probes will be designed.  If "" then all identifiers are
selected.</p></dd>


<dt>start</dt>
<dd><p>Integer specifying the starting position in the alignment where
potential target sites begin.  Preferably a position that is included in
most sequences in the alignment.</p></dd>


<dt>end</dt>
<dd><p>Integer specifying the ending position in the alignment where
potential target sites end.  Preferably a position that is included in most
sequences in the alignment.</p></dd>


<dt>minLength</dt>
<dd><p>Integer providing the minimum length of probes to consider
in the design.</p></dd>


<dt>maxLength</dt>
<dd><p>Integer providing the maximum length of probes to consider
in the design, which must be less than or equal to the <code>maxLength</code> of
tiles.</p></dd>


<dt>maxPermutations</dt>
<dd><p>Integer providing the maximum number of probe
permutations required to reach the desired coverage of a target site.</p></dd>


<dt>minCoverage</dt>
<dd><p>Numeric giving the minimum fraction of the target group's
sequences that must be covered by the designed probe(s).</p></dd>


<dt>minGroupCoverage</dt>
<dd><p>Numeric giving the minimum fraction of the target
group that must have sequence information (not terminal gaps) in the target
site's region.</p></dd>


<dt>hybTemp</dt>
<dd><p>Numeric specifying the hybridization temperature, typically
<code>46</code> degrees Celsius.</p></dd>


<dt>P</dt>
<dd><p>Numeric giving the molar concentration of probes during
hybridization.</p></dd>


<dt>Na</dt>
<dd><p>Numeric giving the molar sodium concentration in the hybridization
buffer.  Values may range between 0.01M and 1M.  Note that salt correction
from <code>1</code> molar is not available for the thermodynamic rules of RNA/RNA
interactions.</p></dd>


<dt>FA</dt>
<dd><p>Numeric concentration (as percent v/v) of the denaturant formamide
in the hybridization buffer.</p></dd>


<dt>minEfficiency</dt>
<dd><p>Numeric giving the minimum equilibrium hybridization
efficiency desired for designed probe(s) at the defined experimental
conditions.</p></dd>


<dt>worstScore</dt>
<dd><p>Numeric specifying the score cutoff to remove target sites
from consideration.  For example, a <code>worstScore</code> of -5 will remove all
probes scoring below -5, although this may eventually result in no probes
meeting the design criteria.</p></dd>


<dt>numProbeSets</dt>
<dd><p>Integer giving the optimal number of dual probe sets to
design.  If set to zero then all potential single probes are returned, and
the probe sets minimizing potential false cross-hybridizations are not
chosen.</p></dd>


<dt>batchSize</dt>
<dd><p>Integer specifying the number of probes to simulate
hybridization per batch that is passed to <code>CalculateEfficiencyFISH</code>.</p></dd>


<dt>target</dt>
<dd><p>The target molecule used in the generation of tiles.  Either
<code>"SSU"</code> for the small-subunit rRNA, <code>"LSU"</code> for the large-subunit
rRNA, or "Other".  Used to determine the domain for <code>dG3</code> calculations,
which is plus or minus 200 nucleotides of the target site if <code>"Other"</code>.</p></dd>


<dt>verbose</dt>
<dd><p>Logical indicating whether to display progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A different <code>data.frame</code> will be returned depending on number
of primer sets requested.  If no probe sets are required then columns
contain the designed probes for every possible position scored by their
potential to cross-hybridize with other identified groups.  If one or more
probe sets are requested then columns contain information for the optimal
set of probes (probe one and probe two) that could be used in combination to
give the fewest potential cross-hybridizations.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Probes are designed to maximize sensitivity and specificity to the target
group(s) (<code>identifier</code>(s)).  If <code>numProbeSets &gt; 0</code> then that many
pairs of probes with minimal cross-hybridization overlap are returned,
enabling increased specificity with a dual-color approach.</p>
<p>Probes are designed from a set of tiles to target each <code>identifier</code>
while minimizing affinity for all other tiled groups.  Arguments provide
constraints that ensure the designed probes meet the specified criteria as
well as being optimized for the particular experimental conditions.  A
search is conducted through all tiles in the same alignment position to
estimate the chance of cross-hybridization with a non-target group.</p>
<p>Two models are used in design, both of which were experimentally calibrated
using denaturation profiles from 5 organisms belonging to all three domains
of life.  Probe lengths are chosen to meet the <code>minEfficiency</code> using a
fast model of probe-target hybridization.  Candidate probes are then
confirmed using a slower model that also takes into account probe-folding
and target-folding.  Finally, probes are scored for their inability to
cross-hybridize with non-target groups by using the fast model and taking
into account any mismatches.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>The program OligoArrayAux
(<a href="http://www.unafold.org/Dinamelt/software/oligoarrayaux.php" class="external-link">http://www.unafold.org/Dinamelt/software/oligoarrayaux.php</a>) must be
installed in a location accessible by the system.  For example, the
following code should print the installed OligoArrayAux version when
executed from the R console:</p>
<p><code>system("hybrid-min -V")</code></p>
<p>To install OligoArrayAux from the downloaded source folder on Unix-like
platforms, open the shell (or Terminal on Mac OS) and type:</p>
<p><code>cd oligoarrayaux # change directory to the correct folder name</code></p>
<p><code>./configure</code></p>
<p><code>make</code></p>
<p><code>sudo make install</code></p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>ES Wright et al. (2014) "Automated Design of Probes for
rRNA-Targeted Fluorescence In Situ Hybridization Reveals the Advantages of
Using Dual Probes for Accurate Identification." Applied and Environmental
Microbiology, doi:10.1128/AEM.01685-14.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="CalculateEfficiencyFISH.html">CalculateEfficiencyFISH</a></code>, <code><a href="TileSeqs.html">TileSeqs</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.sqlite"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># not run (must have OligoArrayAux installed first):</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="va">tiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="TileSeqs.html">TileSeqs</a></span><span class="op">(</span><span class="va">db</span>, identifier<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Enterobacteriales"</span>,<span class="st">"Pseudomonadales"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="va">probes</span> <span class="op">&lt;-</span> <span class="fu">DesignProbes</span><span class="op">(</span><span class="va">tiles</span>, identifier<span class="op">=</span><span class="st">"Enterobacteriales"</span>, start<span class="op">=</span><span class="fl">280</span>, end<span class="op">=</span><span class="fl">420</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

