<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Assists in the design of primer sets targeting a specific group of sequences
while minimizing the potential to cross-amplify other groups of sequences."><title>Design PCR Primers Targeting a Specific Group of Sequences — DesignPrimers • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Design PCR Primers Targeting a Specific Group of Sequences — DesignPrimers"><meta property="og:description" content="Assists in the design of primer sets targeting a specific group of sequences
while minimizing the potential to cross-amplify other groups of sequences."><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Design PCR Primers Targeting a Specific Group of Sequences</h1>
      
      <div class="d-none name"><code>DesignPrimers.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Assists in the design of primer sets targeting a specific group of sequences
while minimizing the potential to cross-amplify other groups of sequences.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DesignPrimers</span><span class="op">(</span></span>
<span>  <span class="va">tiles</span>,</span>
<span>  identifier <span class="op">=</span> <span class="st">""</span>,</span>
<span>  start <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  end <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minLength <span class="op">=</span> <span class="fl">17</span>,</span>
<span>  maxLength <span class="op">=</span> <span class="fl">26</span>,</span>
<span>  maxPermutations <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  minCoverage <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  minGroupCoverage <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  annealingTemp <span class="op">=</span> <span class="fl">64</span>,</span>
<span>  P <span class="op">=</span> <span class="fl">4e-07</span>,</span>
<span>  monovalent <span class="op">=</span> <span class="fl">0.07</span>,</span>
<span>  divalent <span class="op">=</span> <span class="fl">0.003</span>,</span>
<span>  dNTPs <span class="op">=</span> <span class="fl">8e-04</span>,</span>
<span>  minEfficiency <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  worstScore <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,</span>
<span>  numPrimerSets <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  minProductSize <span class="op">=</span> <span class="fl">75</span>,</span>
<span>  maxProductSize <span class="op">=</span> <span class="fl">1200</span>,</span>
<span>  maxSearchSize <span class="op">=</span> <span class="fl">1500</span>,</span>
<span>  batchSize <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  maxDistance <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  primerDimer <span class="op">=</span> <span class="fl">1e-07</span>,</span>
<span>  ragged5Prime <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  taqEfficiency <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  induceMismatch <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  processors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>tiles</dt>
<dd><p>A set of tiles representing each group of sequences, as in the
format created by the function <code>TileSeqs</code>.</p></dd>


<dt>identifier</dt>
<dd><p>Optional character string used to narrow the search
results to those matching a specific identifier.  Determines the target
group(s) for which primers will be designed.  If "" then all identifiers are
selected.</p></dd>


<dt>start</dt>
<dd><p>Integer specifying the starting position in the alignment where
potential forward primer target sites begin.  Preferably a position that is
included in most sequences in the alignment.</p></dd>


<dt>end</dt>
<dd><p>Integer specifying the ending position in the alignment where
potential reverse primer target sites end.  Preferably a position that is
included in most sequences in the alignment.</p></dd>


<dt>minLength</dt>
<dd><p>Integer providing the minimum length of primers to consider
in the design.</p></dd>


<dt>maxLength</dt>
<dd><p>Integer providing the maximum length of primers to consider
in the design, which must be less than or equal to the <code>maxLength</code> of
tiles.</p></dd>


<dt>maxPermutations</dt>
<dd><p>Integer providing the maximum number of permutations
considered as part of a forward or reverse primer set.</p></dd>


<dt>minCoverage</dt>
<dd><p>Numeric giving the minimum fraction of the target group's
sequences that must be covered with the primer set.</p></dd>


<dt>minGroupCoverage</dt>
<dd><p>Numeric giving the minimum fraction of the target
group that must have sequence information (not terminal gaps) in the region
covered by the primer set.</p></dd>


<dt>annealingTemp</dt>
<dd><p>Numeric indicating the desired annealing temperature
that will be used in the PCR experiment.</p></dd>


<dt>P</dt>
<dd><p>Numeric giving the molar concentration of primers in the reaction.</p></dd>


<dt>monovalent</dt>
<dd><p>The molar concentration of monovalent (Na and K) ions
in solution that will be used to determine a sodium equivalent
concentration.</p></dd>


<dt>divalent</dt>
<dd><p>The molar concentration of divalent (Mg) ions in solution
that will be used to determine a sodium equivalent concentration.</p></dd>


<dt>dNTPs</dt>
<dd><p>Numeric giving the molar concentration of free nucleotides
added to the solution that will be used to determine a sodium equivalent
concentration.</p></dd>


<dt>minEfficiency</dt>
<dd><p>Numeric giving the minimum efficiency of hybridization
desired for the primer set.  Note that an efficiency of 99\<!-- % (0.99) will -->
greatly lower predicted specificity of the primer set, however an efficiency
of 50\<!-- % (0.5) may be too low in actuality to amplify the target group due to -->
error in melt temperature predictions.</p></dd>


<dt>worstScore</dt>
<dd><p>Numeric specifying the score cutoff to remove target sites
from consideration.  For example, a <code>worstScore</code> of -5 will remove all
primer sets scoring below -5, although this may eventually result in no
primer sets meeting the design criteria.</p></dd>


<dt>numPrimerSets</dt>
<dd><p>Integer giving the optimal number of primer sets
(forward and reverse primer sets) to design.  If set to zero then all
possible forward and reverse primers are returned, but the primer sets
minimizing potential cross-amplifications are not chosen.</p></dd>


<dt>minProductSize</dt>
<dd><p>Integer giving the minimum number of nucleotides
desired in the PCR product.</p></dd>


<dt>maxProductSize</dt>
<dd><p>Integer giving the maximum number of nucleotides
desired in the PCR product.</p></dd>


<dt>maxSearchSize</dt>
<dd><p>Integer giving the maximum number of nucleotides to
search for false priming upstream and downstream of the expected binding
site.</p></dd>


<dt>batchSize</dt>
<dd><p>Integer specifying the number of primers to simulate
hybridization per batch that is passed to <code>CalculateEfficiencyPCR</code>.</p></dd>


<dt>maxDistance</dt>
<dd><p>Numeric specifying the maximal fraction of mismatched
base pairings on a rolling basis beginning from the 3' end of the primer.</p></dd>


<dt>primerDimer</dt>
<dd><p>Numeric giving the maximum amplification efficiency of
potential primer-dimer products.</p></dd>


<dt>ragged5Prime</dt>
<dd><p>Logical specifying whether the 5' end or 3' end of
primer permutations targeting the same site should be varying lengths.</p></dd>


<dt>taqEfficiency</dt>
<dd><p>Logical determining whether to make use of elongation
efficiency and maxDistance to increase predictive accuracy for <em>Taq</em>
DNA Polymerase amplifying primers with mismatches near the 3' terminus.
Note that this should be set to FALSE if using a high-fidelity polymerase
with 3' to 5' exonuclease activity.</p></dd>


<dt>induceMismatch</dt>
<dd><p>Logical or integer specifying whether to induce a
mismatch in the primer with the template DNA.  If <code>TRUE</code> then a
mismatch is induced at the 6th primer position.  If an integer value is
provided between 2 and 6 then a mismatch is induced in that primer position,
where the 3'-end is defined as position 1.</p></dd>


<dt>processors</dt>
<dd><p>The number of processors to use, or <code>NULL</code> to
automatically detect and use all available processors.</p></dd>


<dt>verbose</dt>
<dd><p>Logical indicating whether to display progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A different <code>data.frame</code> will be returned depending on number
of primer sets requested.  If no primer sets are required then columns
contain the forward and reverse primers for every possible position scored
by their potential to amplify other identified groups.  If one or more
primer sets are requested then columns contain information for the optimal
set of forward and reverse primers that could be used in combination to give
the fewest potential cross-amplifications.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Primers are designed for use with <em>Taq</em> DNA Polymerase to maximize
sensitivity and specificity for the target group of sequences.  The design
makes use of <em>Taq</em>'s bias against certain 3' terminal mismatch types in
order to increase specificity further than can be achieve with hybridization
efficiency alone.</p>
<p>Primers are designed from a set of tiles to target each <code>identifier</code>
while minimizing affinity for all other tiled groups.  Arguments provide
constraints that ensure the designed primer sets meet the specified criteria
as well as being optimized for the particular experimental conditions.  A
search is conducted through all tiles in the same alignment position to
estimate the chance of cross-amplification with a non-target group.</p>
<p>If <code>numPrimers</code> is greater than or equal to one then the set of forward
and reverse primers that minimizes potential false positive overlap is
returned.  This will also initiate a thorough search through all target
sites upstream and downstream of the expected binding sites to ensure that
the primers do not bind to nearby positions.  Lowering the
<code>maxSearchSize</code> will speed up the thorough search at the expense of
potentially missing an unexpected target site.  The number of possible
primer sets assessed is increased with the size of <code>numPrimers</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>The program OligoArrayAux
(<a href="http://www.unafold.org/Dinamelt/software/oligoarrayaux.php" class="external-link">http://www.unafold.org/Dinamelt/software/oligoarrayaux.php</a>) must be
installed in a location accessible by the system.  For example, the
following code should print the installed OligoArrayAux version when
executed from the R console:</p>
<p><code>system("hybrid-min -V")</code></p>
<p>To install OligoArrayAux from the downloaded source folder on Unix-like
platforms, open the shell (or Terminal on Mac OS) and type:</p>
<p><code>cd oligoarrayaux # change directory to the correct folder name</code></p>
<p><code>./configure</code></p>
<p><code>make</code></p>
<p><code>sudo make install</code></p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>ES Wright et al. (2013) "Exploiting Extension Bias in PCR to
Improve Primer Specificity in Ensembles of Nearly Identical DNA Templates."
Environmental Microbiology, doi:10.1111/1462-2920.12259.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="AmplifyDNA.html">AmplifyDNA</a></code>, <code><a href="CalculateEfficiencyPCR.html">CalculateEfficiencyPCR</a></code>,
<code><a href="DesignSignatures.html">DesignSignatures</a></code>, <code><a href="TileSeqs.html">TileSeqs</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.sqlite"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># not run (must have OligoArrayAux installed first):</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="va">tiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="TileSeqs.html">TileSeqs</a></span><span class="op">(</span><span class="va">db</span>, identifier<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Enterobacteriales"</span>,<span class="st">"Pseudomonadales"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu">DesignPrimers</span><span class="op">(</span><span class="va">tiles</span>, identifier<span class="op">=</span><span class="st">"Enterobacteriales"</span>, start<span class="op">=</span><span class="fl">280</span>, end<span class="op">=</span><span class="fl">420</span>,</span></span>
<span class="r-in"><span>           minProductSize<span class="op">=</span><span class="fl">50</span>, numPrimerSets<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

