<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Gene prediction consist of delimiting the boundaries of regions that
function as genes within a genome.  Class Genes provides objects and
functions for storing the boundaries of genes and associated information
resulting from gene prediction.
Gene prediction consist of delimiting the boundaries of regions that
function as genes within a genome.  Class Genes provides objects and
functions for storing the boundaries of genes and associated information
resulting from gene prediction."><title>Genes objects and accessors — Genes • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genes objects and accessors — Genes"><meta property="og:description" content="Gene prediction consist of delimiting the boundaries of regions that
function as genes within a genome.  Class Genes provides objects and
functions for storing the boundaries of genes and associated information
resulting from gene prediction.
Gene prediction consist of delimiting the boundaries of regions that
function as genes within a genome.  Class Genes provides objects and
functions for storing the boundaries of genes and associated information
resulting from gene prediction."><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Genes objects and accessors</h1>
      
      <div class="d-none name"><code>Genes.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Gene prediction consist of delimiting the boundaries of regions that
function as genes within a genome.  Class <code>Genes</code> provides objects and
functions for storing the boundaries of genes and associated information
resulting from gene prediction.</p>
<p>Gene prediction consist of delimiting the boundaries of regions that
function as genes within a genome.  Class <code>Genes</code> provides objects and
functions for storing the boundaries of genes and associated information
resulting from gene prediction.</p>
    </div>


    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>An object of class <code>Genes</code>.</p></dd>


<dt>xlim</dt>
<dd><p>Numeric vector of length 2 specifying the x-axis limits for
plotting.</p></dd>


<dt>ylim</dt>
<dd><p>Numeric vector of length 2 specifying the y-axis limits for
plotting.</p></dd>


<dt>interact</dt>
<dd><p>Logical determining whether the plot is interactive.  If
<code>TRUE</code>, clicking the plot on the right or left side will scroll one
frame in that direction.  To end interaction, either right-click, press the
escape key, or press the stop button depending on the graphics device in
use.</p></dd>


<dt>colorBy</dt>
<dd><p>Character string indicating the name of the column in
<code>x</code> that should be used for coloring.  Unambiguous abbreviations are
also permitted.</p></dd>


<dt>colorRamp</dt>
<dd><p>A function that will return <code>n</code> colors when given a
number <code>n</code>.  Examples are <code>rainbow</code>, <code>heat.colors</code>,
<code>terrain.colors</code>, <code>cm.colors</code>, or (the default)
<code>colorRampPalette</code>.</p></dd>


<dt>colorGenes</dt>
<dd><p>Character string specifying the color of genes, or
<code>NA</code> to color genes according to <code>colorBy</code>.</p></dd>


<dt>i</dt>
<dd><p>Numeric or character vector of row indices to extract from
<code>x</code>.</p></dd>


<dt>j</dt>
<dd><p>Numeric or character vector of column indices to extract from
<code>x</code>.  If <code>j</code> is missing, all columns are included and the returned
object will also belong to class <code>Genes</code>.</p></dd>


<dt>...</dt>
<dd><p>Other optional parameters.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Objects of class <code>Genes</code> are stored as numeric matrices containing
information pertaining to gene predictions.  The matrix columns include the
index (<code>"Index"</code>) of the corresponding sequence in the original genome,
the strand (<code>"Strand"</code>) where the gene is located (either <code>"+"</code>
(<code>0</code>) or <code>"-"</code> (<code>1</code>), the beginning (<code>"Begin"</code>) and
ending (<code>"End"</code>) positions of the gene, scores acquired during
prediction, and whether (<code>!= 0</code>) or not (<code>0</code>) the region was
predicted to be a gene.  Note that the start of the gene is at the beginning
position when the strand is <code>"+"</code> and end when the strand is
<code>"-"</code>.  By convention, rows with negative values in the <code>"Gene"</code>
column represent non-coding RNAs and rows with positive values represent
protein coding genes.</p>
<p>The <code>plot</code> method will show the total score of each prediction along
the genome.  This is most useful when displaying the result of setting
<code>allScores</code> to <code>TRUE</code> in <code>FindGenes</code>.  Here, possible genes
on either strand will be shown (by default), with the predicted genes
highlighted.  The beginning (solid) and ending (dashed) positions are
denoted by vertical lines.  Note that the x-axis is cumulative genome
position, and changes between genome sequences indices are demarcated by
dashed vertical lines.</p>
<p>Objects of class <code>Genes</code> are stored as numeric matrices containing
information pertaining to gene predictions.  The matrix columns include the
index (<code>"Index"</code>) of the corresponding sequence in the original genome,
the strand (<code>"Strand"</code>) where the gene is located (either <code>"+"</code>
(<code>0</code>) or <code>"-"</code> (<code>1</code>), the beginning (<code>"Begin"</code>) and
ending (<code>"End"</code>) positions of the gene, scores acquired during
prediction, and whether (<code>!= 0</code>) or not (<code>0</code>) the region was
predicted to be a gene.  Note that the start of the gene is at the beginning
position when the strand is <code>"+"</code> and end when the strand is
<code>"-"</code>.  By convention, rows with negative values in the <code>"Gene"</code>
column represent non-coding RNAs and rows with positive values represent
protein coding genes.</p>
<p>The <code>plot</code> method will show the total score of each prediction along
the genome.  This is most useful when displaying the result of setting
<code>allScores</code> to <code>TRUE</code> in <code>FindGenes</code>.  Here, possible genes
on either strand will be shown (by default), with the predicted genes
highlighted.  The beginning (solid) and ending (dashed) positions are
denoted by vertical lines.  Note that the x-axis is cumulative genome
position, and changes between genome sequences indices are demarcated by
dashed vertical lines.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="ExtractGenes.html">ExtractGenes</a></code>, <code><a href="FindGenes.html">FindGenes</a></code>,
<code><a href="WriteGenes.html">WriteGenes</a></code></p>
<p><code><a href="ExtractGenes.html">ExtractGenes</a></code>, <code><a href="FindGenes.html">FindGenes</a></code>,
<code><a href="WriteGenes.html">WriteGenes</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># import a test genome</span></span></span>
<span class="r-in"><span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"Chlamydia_trachomatis_NC_000117.fas.gz"</span>,</span></span>
<span class="r-in"><span>  package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="va">fas</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="FindGenes.html">FindGenes</a></span><span class="op">(</span><span class="va">genome</span>, allScores<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Iter  Models Start Motif  Init  Fold UpsNt  Term   RBS  Auto  Stop Genes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          1 18.40                 _                                 8861          1 18.40                 \                                 8861          1 18.40                 |                                 8861          1 18.40                 /                                 8861          1 18.40                 _                                 8861          1 18.40                 \                                 8861          1 18.40                 |                                 8861          1 18.40                 /                                 8861          1 18.40                 _                                 8862         12 18.40  0.48  0.90                                       8883         15 18.38  0.53  1.16  0.40  1.41                           8914         14 15.59  0.51  1.26  0.49  1.77  0.24                     8945         14 15.63  0.51  1.26  0.48  1.82  0.25  0.98  0.10  0.07   8976         14 15.62  0.51  1.24  0.50  1.91  0.25  1.04  0.10  0.07   8977         12 15.60  0.51  1.25  0.48  1.92  0.25  1.03  0.10  0.07   8968         15 15.59  0.51  1.25  0.51  1.94  0.25  1.06  0.10  0.07   8969         12 15.59  0.52  1.25  0.51  1.95  0.26  1.06  0.10  0.07   89510        12 15.58  0.52  1.25  0.52  1.95  0.26  1.09  0.10  0.07   89411        12 15.58  0.52  1.25  0.52  1.96  0.26  1.10  0.10  0.07   89412        11 15.57  0.52  1.25  0.50  1.97  0.26  1.11  0.10  0.07   89413        12 15.57  0.52  1.25  0.50  1.98  0.26  1.12  0.10  0.07   89414        11 15.57  0.52  1.25  0.50  1.98  0.26  1.13  0.10  0.07   89415        14 15.57  0.52  1.25  0.51  1.99  0.26  1.12  0.10  0.07   89416        14 15.57  0.52  1.25  0.52  1.99  0.26  1.14  0.10  0.07   894</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 14.13 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">x</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genes object of size 109,490 specifying:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 894 protein coding genes from 66 to 5,361 nucleotides.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 108,596 open reading frames from 60 to 5,370 nucleotides.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Index Strand Begin  End TotalScore ... Gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     1      0     1 1176      78.80 ...    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     1      0     2   82      -4.38 ...    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     1      0    16 1176      61.98 ...    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     1      0    17   82      -8.85 ...    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [ reached 'max' / getOption("max.print") -- omitted 2 rows ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... with 109,484 more rows.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">unclass</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="co"># the underlying structure</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Index Strand Begin  End TotalScore LengthScore CodingScore CodonModel</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     1      0     1 1176  78.796288   39.252787  40.3002513          1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CouplingScore  StartScore StopScore InitialCodonScore TerminationCodonScore</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     3.2383375   0.0000000  0.187984       -3.86299429            -0.2739789</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   RibosomeBindingSiteScore AutocorrelationScore UpstreamNucleotideScore</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  0.00000          -0.04609886                0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   UpstreamMotifScore FoldingScore FractionReps Gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           0.000000    0.0000000            1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [ reached getOption("max.print") -- omitted 5 rows ]</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co"># default coloring by "Strand"</span></span></span>
<span class="r-plt img"><img src="Genes-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># color by RBS score (blue is weak/low, red is strong/high)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, colorBy<span class="op">=</span><span class="st">"RibosomeBindingSiteScore"</span>, colorGenes<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="Genes-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># color by fraction of times a gene was chosen</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, colorBy<span class="op">=</span><span class="st">"FractionReps"</span>, colorGenes<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="Genes-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># color by which codon model was selected for each ORF</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, colorBy<span class="op">=</span><span class="st">"CodonModel"</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3e4</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="Genes-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># import a test genome</span></span></span>
<span class="r-in"><span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"Chlamydia_trachomatis_NC_000117.fas.gz"</span>,</span></span>
<span class="r-in"><span>  package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="fu">readDNAStringSet</span><span class="op">(</span><span class="va">fas</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="FindGenes.html">FindGenes</a></span><span class="op">(</span><span class="va">genome</span>, allScores<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Iter  Models Start Motif  Init  Fold UpsNt  Term   RBS  Auto  Stop Genes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          1 18.40                 _                                 8861          1 18.40                 \                                 8861          1 18.40                 |                                 8861          1 18.40                 /                                 8861          1 18.40                 _                                 8861          1 18.40                 \                                 8861          1 18.40                 |                                 8861          1 18.40                 /                                 8861          1 18.40                 _                                 8862         12 18.40  0.48  0.90                                       8883         15 18.38  0.53  1.16  0.42  1.41                           8904         14 15.60  0.51  1.25  0.47  1.75  0.24                     8945         12 15.62  0.51  1.25  0.49  1.82  0.25  0.96  0.10  0.07   8976         13 15.59  0.51  1.24  0.49  1.93  0.25  1.05  0.10  0.07   8977         13 15.58  0.51  1.25  0.51  1.93  0.25  1.07  0.10  0.07   8988         13 15.57  0.51  1.25  0.52  1.95  0.25  1.09  0.10  0.07   8979         13 15.56  0.51  1.26  0.51  1.96  0.25  1.08  0.10  0.07   89810        13 15.56  0.51  1.26  0.52  1.96  0.25  1.12  0.10  0.07   897</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 10.2 secs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">x</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genes object of size 109,490 specifying:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 897 protein coding genes from 66 to 5,361 nucleotides.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 108,593 open reading frames from 60 to 5,370 nucleotides.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Index Strand Begin  End TotalScore ... Gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     1      0     1 1176      78.65 ...    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     1      0     2   82      -4.63 ...    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     1      0    16 1176      62.03 ...    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     1      0    17   82      -8.67 ...    0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [ reached 'max' / getOption("max.print") -- omitted 2 rows ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... with 109,484 more rows.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">unclass</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="co"># the underlying structure</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Index Strand Begin  End TotalScore LengthScore CodingScore CodonModel</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     1      0     1 1176  78.649343   39.233990  40.8310925          1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   CouplingScore  StartScore StopScore InitialCodonScore TerminationCodonScore</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     2.9226209   0.0000000 0.1909784       -3.95004382            -0.3401371</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   RibosomeBindingSiteScore AutocorrelationScore UpstreamNucleotideScore</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                0.0000000          -0.23915805                0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   UpstreamMotifScore FoldingScore FractionReps Gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           0.000000    0.0000000            1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [ reached getOption("max.print") -- omitted 5 rows ]</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co"># default coloring by "Strand"</span></span></span>
<span class="r-plt img"><img src="Genes-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># color by RBS score (blue is weak/low, red is strong/high)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, colorBy<span class="op">=</span><span class="st">"RibosomeBindingSiteScore"</span>, colorGenes<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="Genes-6.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># color by fraction of times a gene was chosen</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, colorBy<span class="op">=</span><span class="st">"FractionReps"</span>, colorGenes<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="Genes-7.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># color by which codon model was selected for each ORF</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, colorBy<span class="op">=</span><span class="st">"CodonModel"</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3e4</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="Genes-8.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

