<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Assign Sequences a Taxonomic Classification — IdTaxa • DECIPHER</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Assign Sequences a Taxonomic Classification — IdTaxa"><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/DesignSpeciesSpecificPrimers.html">How to design species-specific primers</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Assign Sequences a Taxonomic Classification</h1>
      
      <div class="d-none name"><code>IdTaxa.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">IdTaxa</span><span class="op">(</span></span>
<span>  <span class="va">test</span>,</span>
<span>  <span class="va">trainingSet</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"extended"</span>,</span>
<span>  strand <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">60</span>,</span>
<span>  bootstraps <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  samples <span class="op">=</span> <span class="va">L</span><span class="op">^</span><span class="fl">0.47</span>,</span>
<span>  minDescend <span class="op">=</span> <span class="fl">0.98</span>,</span>
<span>  fullLength <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  processors <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>test</dt>
<dd><p>An <code>AAStringSet</code>, <code>DNAStringSet</code>, or
<code>RNAStringSet</code> of unaligned sequences.</p></dd>


<dt>trainingSet</dt>
<dd><p>An object of class <code>Taxa</code> and subclass Train
compatible with the class of <code>test</code>.</p></dd>


<dt>type</dt>
<dd><p>Character string indicating the type of output desired.  This
should be (an abbreviation of) one of <code>"extended"</code> or
<code>"collapsed"</code>.  (See value section below.)</p></dd>


<dt>strand</dt>
<dd><p>Character string indicating the orientation of the <code>test</code>
sequences relative to the <code>trainingSet</code>.  This should be (an
abbreviation of) one of <code>"both"</code>, <code>"top"</code>, or <code>"bottom"</code>.
The top strand is defined as the input <code>test</code> sequences being in the
same orientation as the <code>trainingSet</code>, and the bottom strand is its
reverse complement orientation.  The default of <code>"both"</code> will classify
using both orientations and choose the result with highest confidence.
Choosing the correct <code>strand</code> will make classification over 2-fold
faster, assuming that all of the reads are in the same orientation. Note
that <code>strand</code> is ignored when <code>test</code> is an <code>AAStringSet</code>.</p></dd>


<dt>threshold</dt>
<dd><p>Numeric specifying the confidence at which to truncate the
output taxonomic classifications.  Lower values of <code>threshold</code> will
classify deeper into the taxonomic tree at the expense of accuracy, and
vise-versa for higher values of <code>threshold</code>.</p></dd>


<dt>bootstraps</dt>
<dd><p>Integer giving the maximum number of bootstrap replicates
to perform for each sequence.  The number of bootstrap replicates is set
automatically such that (on average) 99\<!-- % of k-mers are sampled in each -->
<code>test</code> sequence.</p></dd>


<dt>samples</dt>
<dd><p>A function or call written as a function of <code>L', which will evaluate to a numeric vector the same length as </code>L'.  Typically of the form
``<code>A + B*L^C</code>'', where <code>A', </code>B', and `C' are constants.</p></dd>


<dt>minDescend</dt>
<dd><p>Numeric giving the minimum fraction of <code>bootstraps</code>
required to descend the tree during the initial tree descend phase of the
algorithm.  Higher values are less likely to descend the tree, causing
direct comparison against more sequences in the <code>trainingSet</code>.  Lower
values may increase classification speed at the expense of accuracy.
Suggested values are between <code>1.0</code> and <code>0.9</code>.</p></dd>


<dt>fullLength</dt>
<dd><p>Numeric specifying the fold-difference in sequence lengths
between sequences in <code>test</code> and <code>trainingSet</code> that is allowable,
or <code>0</code> (the default) to consider all sequences in <code>trainingSet</code>
regardless of length.  Can be specified as either a single numeric (&gt; 1), or
two numerics specifying the upper and lower fold-difference.  If
<code>fullLength</code> is between <code>0</code> and <code>1</code> (exclusive), the
fold-difference is inferred from the length variability among sequences
belonging to each class based on the <code>foldDifference</code> quantiles.  For
example, setting <code>fullLength</code> to <code>0.99</code> would use the 1st and 99th
percentile of intra-group length variability from the <code>trainingSet</code>.
In the case of full-length sequences, specifying <code>fullLength</code> can
improve both speed and accuracy by using sequence length as a pre-filter to
classification.  Note that <code>fullLength</code> should only be greater than
<code>0</code> when both the <code>test</code> and <code>trainingSet</code> consist of
full-length sequences.</p></dd>


<dt>processors</dt>
<dd><p>The number of processors to use, or <code>NULL</code> to
automatically detect and use all available processors.</p></dd>


<dt>verbose</dt>
<dd><p>Logical indicating whether to display progress.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>If <code>type</code> is <code>"extended"</code> (the default) then an object of
class <code>Taxa</code> and subclass Train is returned.  This is stored as a list
with elements corresponding to their respective sequence in <code>test</code>.
Each list element contains components:</p>
<dl><dt>taxon</dt>
<dd><p>A character vector
containing the taxa to which the sequence was assigned.</p></dd>
 <dt>confidence</dt>
<dd><p>A numeric vector giving the corresponding percent confidence for each taxon.</p></dd>
 <dt>rank</dt>
<dd><p>If the classifier was trained with a set of <code>rank</code>s, a
character vector containing the rank name of each taxon.</p></dd>


</dl><p>If <code>type</code> is <code>"collapsed"</code> then a character vector is returned
with the taxonomic assignment for each sequence.  This takes the repeating
form <code>Taxon name [rank, confidence\%]; ...'' if \code{rank}s were supplied during training, or </code>Taxon name confidence\<!-- %}; ...'' otherwise. --></p>


<p>Classifies sequences according to a training set by assigning a confidence
to taxonomic labels for each taxonomic level.</p>


<p>Sequences in <code>test</code> are each assigned a taxonomic classification based
on the <code>trainingSet</code> created with <code><a href="LearnTaxa.html">LearnTaxa</a></code>.  Each
taxonomic level is given a confidence between 0\<!-- % and 100\\%, and the -->
taxonomy is truncated where confidence drops below <code>threshold</code>.  If the
taxonomic classification was truncated, the last group is labeled with
``unclassified_'' followed by the final taxon's name.  Note that the
reported confidence is not a p-value but does directly relate to a given
classification's probability of being wrong.  The default <code>threshold</code>
of <code>60%</code> is intended to minimize the rate of incorrect classifications.
Lower values of <code>threshold</code> (e.g., <code>50%</code>) may be preferred to
increase the taxonomic depth of classifications.  Values of <code>60%</code> or
<code>50%</code> are recommended for nucleotide sequences and <code>50%</code> or
<code>40%</code> for amino acid sequences.</p>


<p></p>


<p>Murali, A., et al. (2018). IDTAXA: a novel approach for accurate
taxonomic classification of microbiome sequences. Microbiome, 6, 140.
https://doi.org/10.1186/s40168-018-0521-5Cooley, N. and Wright, E. (2021). Accurate annotation of protein coding
sequences with IDTAXA. NAR Genomics and Bioinformatics, <b>3(3)</b>.
https://doi.org/10.1093/nargab/lqab080</p>


<p><code><a href="LearnTaxa.html">LearnTaxa</a></code>, <code><a href="Taxa.html">Taxa-class</a></code></p>


<p>Erik Wright <a href="mailto:eswright@pitt.edu">eswright@pitt.edu</a></p>


    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

