<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DECIPHER">
<title>The Art of Multiple Sequence Alignment in R • DECIPHER</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="The Art of Multiple Sequence Alignment in R">
<meta property="og:description" content="DECIPHER">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>The Art of Multiple Sequence Alignment in R</h1>
                        <h4 data-toc-skip class="author">Erik S.
Wright</h4>
            
            <h4 data-toc-skip class="date">2023-08-24</h4>
      
      
      <div class="d-none name"><code>ArtOfAlignmentInR.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This document is intended to illustrate the art of multiple sequence
alignment in using <strong>DECIPHER</strong>. Even though its beauty is
often concealed, multiple sequence alignment is a form of art in more
ways than one. Take a look at Figure @ref(f1) for an illustration of
what is happening behind the scenes during multiple sequence alignment.
The practice of sequence alignment is one that requires a degree of
skill, and it is that art which this vignette intends to convey. It is
simply not enough to “plug” sequences into a multiple sequence aligner
and blindly trust the result. An appreciation for the art as well a
careful consideration of the results are required.</p>
<p>What really is multiple sequence alignment, and is there a single
correct alignment? Generally speaking, alignment seeks to perform the
act of taking multiple divergent biological sequences of the same “type”
and fitting them to a form that reflects some shared quality. That
quality may be how they look structurally, how they evolved from a
common ancestor, or optimization of a mathematical construct. As with
most multiple sequence aligners, <strong>DECIPHER</strong> is “trained”
to maximize scoring metrics in order to accomplish a combination of both
structural alignment and evolutionary alignment. The idea is to give the
alignment a biological basis even though the molecules that the
sequences represent will never meet each other and align under any
natural circumstance.</p>
<p>The workhorse for sequence alignment in <strong>DECIPHER</strong> is
<code>AlignProfiles</code>, which takes in two aligned sets of DNA, RNA,
or amino acid (AA) sequences and returns a merged alignment. For more
than two sequences, the function <code>AlignSeqs</code> can be used to
perform multiple sequence alignment in a progressive/iterative manner on
sequences of the same kind. In this case, multiple alignment works by
aligning two sequences, merging with another sequence, merging with
another set of sequences, and so-forth until all the sequences are
aligned. This process is iterated to further refine the alignment. There
are other functions that extend use of <code>AlignSeqs</code> for
different purposes:</p>
<ol style="list-style-type: decimal">
<li>The first is <code>AlignTranslation</code>, which will align DNA/RNA
sequences based on their amino acid translation and then reverse
translate them back to DNA/RNA. Aligning protein sequences is more
accurate since amino acids are more conserved than their corresponding
coding sequence.</li>
<li>The second function, <code>AlignDB</code>, enables generating
alignments from many more sequences than are possible to fit in memory.
Its main purpose is to merge sub-alignments where each alignment alone
is composed of many thousands of sequences. This is accomplished by
storing all of the aligned sequences in a database and only working with
“profiles” representing the alignment.</li>
<li>The function <code>AdjustAlignment</code> takes in an existing
alignment and shifts groups of gaps right and left to achieve a better
alignment. Its purpose is to eliminate artifacts that accumulate during
progressive alignment, and to replace the tedious &amp; subjective
process of manually correcting an alignment.</li>
<li>Finally, <code>StaggerAlignment</code> will create a “staggered”
alignment by separating potentially non-homologous positions into
separate columns. This function will help minimize false homologies when
building a phylogenetic tree, although the resulting alignment is not as
aesthetically pleasing.</li>
<li>The functions <code>FindSynteny</code> and <code>AlignSynteny</code>
can be used in combination to perform pairwise alignment of homologous
regions from multiple genomes or non-collinear sequences. These
functions interact with a sequence database containing the genomes,
which can each be comprised of multiple sequences (i.e., scaffolds,
contigs, or chromosomes).</li>
</ol>
</div>
<div class="section level2">
<h2 id="alignment-speed">Alignment Speed<a class="anchor" aria-label="anchor" href="#alignment-speed"></a>
</h2>
<p>The dynamic programming method used by <strong>DECIPHER</strong> for
aligning two profiles requires order <code>N*M</code> time and memory
space where <code>N</code> and <code>M</code> are the width of the
pattern and subject. Since multiple sequence alignment is an inherently
challenging problem for long sequences, heuristics are employed to
maximize speed while maintaining reasonable accuracy. In this regard,
the two control parameters available to the user are
<code>restrict</code> and <code>anchor</code>. The objective of the
<code>restrict</code> parameter is to convert the problem from one
taking quadratic time to linear time. The goal of the
<code>anchor</code> parameter is do the equivalent for memory space so
that very long sequences can be efficiently aligned.</p>
<p>The orange diagonal line in Figure @ref(f2) shows the optimal path
for aligning two sequence profiles. The blue segments to the left and
right of the optimal path give the constraint boundaries, which the user
controls with the <code>restrict</code> parameter. Areas above and below
the upper and lower (respectively) constraint boundaries are neglected
from further consideration. A higher (less negative) value of
<code>restrict[1]</code> will further constrain the possible “alignment
space,” which represents all possible alignments between two sequences.
Since the optimal path is not known till completion of the matrix, it is
risky to overly constrain the matrix. This is particularly true in
situations where the sequences are not mostly overlapping because the
optimal path will likely not be diagonal, causing the path to cross a
constraint boundary. In the non-overlapping case
<code>restrict[1]</code> could be set below the default to ensure that
the entire “alignment space” is available.</p>
<p>Neglecting the “corners” of the alignment space effectively converts
a quadratic time problem into a near-linear time problem. We can see
this by comparing <code>AlignProfiles</code> with and without
restricting the matrix at different sequence lengths. To extend our
comparison we can include the <strong>Biostrings</strong> function
<code>pairwiseAlignment</code>. In this simulation, two sequences with
90% identity are aligned and the elapsed time is recorded for a variety
of sequence lengths. As can be seen in Figure @ref(f3) below,
<em>without</em> restriction <code>AlignProfiles</code> takes quadratic
time in the same manner as <code>pairwiseAlignment</code>. However,
<em>with</em> restriction <code>AlignProfiles</code> takes linear time,
requiring far less than a microsecond per nucleotide.</p>
<p>The parameter <code>anchor</code> controls the fraction of sequences
that must share a common region to anchor the alignment space (Fig.
@ref(f2)). <code>AlignProfiles</code> will search for shared anchor
points between the two sequence sets being aligned, and if the fraction
shared is above <code>anchor</code> (70% by default) then that position
is fixed in the “alignment space.” Anchors are 15-mer (for DNA/RNA) or
7-mer (for AA) exact matches between two sequences that must occur in
the same order in both sequence profiles. Anchoring generally does not
affect accuracy, but can greatly diminish the amount of memory required
for alignment. In Fig. @ref(f2), the largest white box represents the
maximal memory space required with anchoring, while the entire alignment
space (grey plus white areas) would be required without anchoring. The
longest pair of sequence profiles that can be aligned without anchoring
is about 46 thousand nucleotides, as shown by the end of the red dotted
line in Figure @ref(f3). If regularly spaced anchor points are available
then the maximum sequence length is greatly extended. In the vast
majority of cases anchoring gives the same result as without anchoring,
but with less time and memory space required.</p>
</div>
<div class="section level2">
<h2 id="alignment-accuracy">Alignment Accuracy<a class="anchor" aria-label="anchor" href="#alignment-accuracy"></a>
</h2>
<p>Figure @ref(f4) compares the performance of <strong>DECIPHER</strong>
to other sequence alignment software on structural amino acid benchmarks
<span class="citation">[@Edgar:2010]</span>. All benchmarks have flaws,
some of which can easily be found by eye in highly similar sequence
sets, and therefore benchmark results should treated with care <span class="citation">[@Iantorno:2014]</span>. As can be seen in the figure,
the performance of <strong>DECIPHER</strong> is similar to that of other
popular alignment software such as MAFFT <span class="citation">[@Katoh:2002]</span> and MUSCLE <span class="citation">[@Edgar:2004]</span> for smaller benchmarks. However,
<strong>DECIPHER</strong> outperforms other programs on large sequence
sets (Fig. @ref(f5)), and its relative advantage continues increase as
more sequences are aligned <span class="citation">[@Wright:2015]</span>.
Importantly, this is because <strong>DECIPHER</strong> exhibits far less
fall-off in accuracy as additional sequences are added.</p>
<p>The accuracy of protein alignment begins to drop-off when sequences
in the reference alignment have less than 40% average pairwise identity
(Fig. @ref(f4)). A similar decline in performance is observed with
DNA/RNA sequences, but the drop-off occurs much earlier at around 60%
sequence identity. Therefore, it is generally preferable to align coding
sequences by their translation using <code>AlignTranslation</code>. This
function first translates the input DNA/RNA sequences, then aligns the
translation, and finally (conceptually) reverse translates the amino
acid sequences to obtain aligned DNA/RNA sequences. Nevertheless, even
protein alignment cannot be considered reliable when the sequences being
aligned differ by more than 70%.</p>
</div>
<div class="section level2">
<h2 id="recommendations-for-optimal-performance">Recommendations for optimal performance<a class="anchor" aria-label="anchor" href="#recommendations-for-optimal-performance"></a>
</h2>
<p><strong>DECIPHER</strong> has a number of alignment functions and
associated parameters. The flow-chart in Figure @ref(f6) is intended to
simplify this process for the most frequently encountered multiple
sequence alignment tasks. For more information on any of these
suggestions, refer to the examples in the following sections of this
vignette.</p>
</div>
<div class="section level2">
<h2 id="single-gene-alignment">Single Gene Alignment<a class="anchor" aria-label="anchor" href="#single-gene-alignment"></a>
</h2>
<div class="section level3">
<h3 id="example-protein-coding-sequences">Example: Protein coding sequences<a class="anchor" aria-label="anchor" href="#example-protein-coding-sequences"></a>
</h3>
<p>For this example we are going to align the <em>rplB</em> coding
sequence from many different Bacteria. The <em>rplB</em> gene encodes
one of the primary ribosomal RNA binding proteins: the 50S ribosomal
protein L2. We begin by loading the library and importing the sequences
from a FASTA file. Be sure to change the path names to those on your
system by replacing all of the text inside quotes labeled “<span class="math inline">\(&lt;&lt;\)</span>path to …<span class="math inline">\(&gt;&gt;\)</span>” with the actual path on your
system.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DECIPHER</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># specify the path to your sequence file:</span></span>
<span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="st">"&lt;&lt;path to FASTA file&gt;&gt;"</span></span>
<span><span class="co"># OR find the example sequence file used in this tutorial:</span></span>
<span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"50S_ribosomal_protein_L2.fas"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">fas</span><span class="op">)</span></span>
<span><span class="va">dna</span> <span class="co"># the unaligned sequences</span></span></code></pre></div>
<pre><code><span><span class="co">## DNAStringSet object of length 317:</span></span>
<span><span class="co">##       width seq                                             names               </span></span>
<span><span class="co">##   [1]   819 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">AAAAAA</span><span style="color: #000000; background-color: #FFD7AF;">TTTT</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAAAAAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span> Rickettsia prowaz...</span></span>
<span><span class="co">##   [2]   822 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">AAAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span> Porphyromonas gin...</span></span>
<span><span class="co">##   [3]   822 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">AAAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span> Porphyromonas gin...</span></span>
<span><span class="co">##   [4]   822 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">AAAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span> Porphyromonas gin...</span></span>
<span><span class="co">##   [5]   819 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span> Pasteurella multo...</span></span>
<span><span class="co">##   ...   ... ...</span></span>
<span><span class="co">## [313]   819 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span> Pectobacterium at...</span></span>
<span><span class="co">## [314]   822 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAAAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span> Acinetobacter sp....</span></span>
<span><span class="co">## [315]   864 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span> Thermosynechococc...</span></span>
<span><span class="co">## [316]   831 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span> Bradyrhizobium ja...</span></span>
<span><span class="co">## [317]   840 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span> Gloeobacter viola...</span></span></code></pre>
<p>We can align the DNA by either aligning the coding sequences
directly, or their translations (amino acid sequences). Both methods
result in an aligned set of DNA sequences, unless the argument
<code>type</code> is <code>"AAStringSet"</code> in
<code>AlignTranslation</code>. A quick inspection reveals that the
method of translating before alignment yields a more appealing result.
In particular, the reading frame is maintained when aligning the
translations. However, if the dna did not code for a protein then the
only option would be to use <code>AlignSeqs</code> because the
translation would be meaningless.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">AA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignTranslation.html">AlignTranslation</a></span><span class="op">(</span><span class="va">dna</span>, type<span class="op">=</span><span class="st">"AAStringSet"</span><span class="op">)</span> <span class="co"># align the translation</span></span>
<span><span class="fu"><a href="../reference/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">AA</span>, highlight<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="co"># view the alignment</span></span>
<span></span>
<span><span class="va">DNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignSeqs.html">AlignSeqs</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span> <span class="co"># align the sequences directly without translation</span></span>
<span><span class="va">DNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignTranslation.html">AlignTranslation</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span> <span class="co"># align the translation then reverse translate</span></span>
<span></span>
<span><span class="co"># write the aligned sequences to a FASTA file</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">writeXStringSet</a></span><span class="op">(</span><span class="va">DNA</span>, file<span class="op">=</span><span class="st">"&lt;&lt;path to output file&gt;&gt;"</span><span class="op">)</span></span></code></pre></div>
<p>Note that frameshift errors can greatly disrupt the alignment of
protein coding sequences. Frameshifts can be corrected by first using
<code>CorrectFrameshifts</code> on the nucleotide sequences, and then
using the corrected sequences as input to <code>AlignTranslation</code>
with the argument <code>readingFrame</code> equal to <code>1</code>.</p>
<p>If the input sequences include exact replicates, then alignment can
be accelerated by de-replicating the sequences before alignment. The
sequences can then be re-replicated after alignment to create a larger
alignment of all the original sequences. <code>AlignSeqs</code> does not
automatically handle redundancy in the input sequences, but doing so is
fairly straightforward. In this case there aren’t any exact duplicates
in the example <code>dna</code> sequences. Nonetheless, the process to
de-replicate before alignment and re-replicate after alignment would
look like:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">u_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span> <span class="co"># the unique input sequences</span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">dna</span>, <span class="va">u_dna</span><span class="op">)</span> <span class="co"># de-replication index</span></span>
<span></span>
<span><span class="va">U_DNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignSeqs.html">AlignSeqs</a></span><span class="op">(</span><span class="va">u_dna</span><span class="op">)</span> <span class="co"># align the sequences directly without translation</span></span>
<span><span class="va">DNA</span> <span class="op">&lt;-</span> <span class="va">U_DNA</span><span class="op">[</span><span class="va">index</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">DNA</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span> <span class="co"># the re-replicated alignment</span></span></code></pre></div>
<p>Also, when aligning nucleotide sequences (or their translations), it
may be the case that the sequences are in different orientations. If so,
consider reorienting the sequences so that they all have the same
directionality and complementarity by using
<code>OrientNucleotides</code> prior to alignment.</p>
</div>
<div class="section level3">
<h3 id="example-non-coding-rna-sequences">Example: Non-coding RNA sequences<a class="anchor" aria-label="anchor" href="#example-non-coding-rna-sequences"></a>
</h3>
<p>Much like proteins, non-coding RNAs often have a conserved secondary
structure that can be used to improve their alignment. The
<code>PredictDBN</code> function will predict base pairings from a
sequence alignment by calculating the mutual information between pairs
of positions. If RNA sequences are given as input,
<code>AlignSeqs</code> will automatically use the output of
<code>PredictDBN</code> to iteratively improve the alignment. Providing
an <em>RNAStringSet</em> also causes single-base and double-base
substitution matrices to be used, and is preferable to providing a
<em>DNAStringSet</em> when the sequences are non-coding RNA. The type of
the input sequences can easily be converted to RNA, as shown below.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># database containing 16S ribosomal RNA sequences</span></span>
<span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.sqlite"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">db</span>, remove<span class="op">=</span><span class="st">"all"</span>, type<span class="op">=</span><span class="st">"RNAStringSet"</span><span class="op">)</span></span>
<span><span class="co"># or if starting with DNA sequences, convert to RNA with:</span></span>
<span><span class="co"># rna &lt;- RNAStringSet(dna)</span></span>
<span><span class="co"># or import RNA sequences directly using:</span></span>
<span><span class="co"># rna &lt;- readRNAStringSet("&lt;&lt;path to FASTA file&gt;&gt;")</span></span>
<span></span>
<span><span class="va">alignedRNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignSeqs.html">AlignSeqs</a></span><span class="op">(</span><span class="va">rna</span><span class="op">)</span> <span class="co"># align with RNA secondary structure</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="example-aligning-two-aligned-sequence-sets">Example: Aligning two aligned sequence sets<a class="anchor" aria-label="anchor" href="#example-aligning-two-aligned-sequence-sets"></a>
</h3>
<p>It is sometimes useful to align two or more previously-aligned sets
of sequences. Here we can use the function <code>AlignProfiles</code> to
directly align profiles of the two sequence sets:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">half</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">dna1</span> <span class="op">&lt;-</span> <span class="va">dna</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">half</span><span class="op">]</span> <span class="co"># first half</span></span>
<span><span class="va">dna2</span> <span class="op">&lt;-</span> <span class="va">dna</span><span class="op">[</span><span class="op">(</span><span class="va">half</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span><span class="op">]</span> <span class="co"># second half</span></span>
<span></span>
<span><span class="va">AA1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignTranslation.html">AlignTranslation</a></span><span class="op">(</span><span class="va">dna1</span>, type<span class="op">=</span><span class="st">"AAStringSet"</span><span class="op">)</span></span>
<span><span class="va">AA2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignTranslation.html">AlignTranslation</a></span><span class="op">(</span><span class="va">dna2</span>, type<span class="op">=</span><span class="st">"AAStringSet"</span><span class="op">)</span></span>
<span><span class="va">AA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignProfiles.html">AlignProfiles</a></span><span class="op">(</span><span class="va">AA1</span>, <span class="va">AA2</span><span class="op">)</span> <span class="co"># align two alignments</span></span></code></pre></div>
<p>When the two sequence sets are very large it may be impossible to fit
both sets of input sequences and the output alignment into memory at
once. The function <code>AlignDB</code> can align the sequences in two
database tables, or two sets of sequences corresponding to separate
identifiers in the same table. <code>AlignDB</code> takes as input two
tblNames and/or identifiers, and iteratively builds a profile for each
of those respective sequence alignments in the database. These profiles
are aligned, and the insertions are iteratively applied to each of the
input sequences until the completed alignment has been stored in
add2tbl.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Align DNA sequences stored in separate tables:</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">":memory:"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">AA1</span>, <span class="st">"DNAStringSet"</span>, <span class="va">dbConn</span>, <span class="st">"AA1"</span>, tblName<span class="op">=</span><span class="st">"AA1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">AA2</span>, <span class="st">"DNAStringSet"</span>, <span class="va">dbConn</span>, <span class="st">"AA2"</span>, tblName<span class="op">=</span><span class="st">"AA2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/AlignDB.html">AlignDB</a></span><span class="op">(</span><span class="va">dbConn</span>, tblName<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AA1"</span>, <span class="st">"AA2"</span><span class="op">)</span>, add2tbl<span class="op">=</span><span class="st">"AA"</span>,</span>
<span>        type<span class="op">=</span><span class="st">"AAStringSet"</span><span class="op">)</span></span>
<span><span class="va">AA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span>, tblName<span class="op">=</span><span class="st">"AA"</span>, type<span class="op">=</span><span class="st">"AAStringSet"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span>, tblName<span class="op">=</span><span class="st">"AA"</span><span class="op">)</span></span>
<span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span></code></pre></div>
<p>The number of sequences required to fit into memory when aligning two
sequence sets with <code>AlignDB</code> is controlled by the batchSize
parameter. In this way <code>AlignDB</code> can be used to align large
sequence alignments with only minimal memory required.</p>
</div>
</div>
<div class="section level2">
<h2 id="advanced-options-features">Advanced Options &amp; Features<a class="anchor" aria-label="anchor" href="#advanced-options-features"></a>
</h2>
<div class="section level3">
<h3 id="example-building-a-guide-tree">Example: Building a Guide Tree<a class="anchor" aria-label="anchor" href="#example-building-a-guide-tree"></a>
</h3>
<p>The <code>AlignSeqs</code> function uses a guide tree to decide the
order in which to align pairs of sequence profiles. The
<code>guideTree</code> input is a <em>dendrogram</em> (tree) object with
one leaf per input sequence. By default this guide tree is generated
directly from the input sequences using the order of shared k-mers
(i.e., when the argument <code>guideTree</code> is <code>NULL</code>).
This default guide tree performs very well but requires <span class="math inline">\(\mathcal{O}(n^2)\)</span> time and memory space to
construct. Therefore, it may be useful to rely on a chained guide tree
when aligning hundreds of thousands of unique sequences.</p>
<p>It has been shown that reasonably accurate alignments of tens of
thousands of sequences can be obtained by using a chain guide tree <span class="citation">[@Boyce:2014]</span>. With a chained guide tree,
sequences are added one-by-one to a growing profile representing all of
the aligned sequences. Figure @ref(f7) shows the result of using
<strong>DECIPHER</strong> to align increasing numbers of Cytochrome P450
sequences (in accordance with the method in reference <span class="citation">[@Boyce:2014]</span>), using either a chained guide
tree or the default guide tree. A chained guide tree can be easily
generated, as shown below.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># form a chained guide tree</span></span>
<span><span class="va">gT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">dna</span><span class="op">)</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"height"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"label"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span><span class="op">[</span><span class="va">x</span><span class="op">]</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"members"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1L</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"leaf"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span>        <span class="va">x</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">gT</span>, <span class="st">"height"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">gT</span>, <span class="st">"members"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">gT</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"dendrogram"</span></span>
<span></span>
<span><span class="co"># use the guide tree as input for alignment</span></span>
<span><span class="va">DNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignTranslation.html">AlignTranslation</a></span><span class="op">(</span><span class="va">dna</span>,</span>
<span>    guideTree<span class="op">=</span><span class="va">gT</span>,</span>
<span>    iterations<span class="op">=</span><span class="fl">0</span>,</span>
<span>    refinements<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>It is also possible to read a Newick formatted tree into using the
function <code>ReadDendrogram</code>, and specify this object as the
input <code>guideTree</code>.</p>
</div>
<div class="section level3">
<h3 id="example-post-processing-an-existing-multiple-alignment">Example: Post-processing an existing multiple alignment<a class="anchor" aria-label="anchor" href="#example-post-processing-an-existing-multiple-alignment"></a>
</h3>
<p>There are several steps that can be taken after alignment to verify
or improve the alignment. The most important step is to look at the
result to ensure that it meets expectations. Spurious (unalignable)
sequences can then be removed and the alignment process repeated as
desired. The simplest way to view sequences with
<strong>DECIPHER</strong> is by using the function
<code>BrowseSeqs</code>. The <code>highlight</code> parameter controls
which sequence, if any, is in focus (highlighted). A value of zero
highlights the consensus sequence as shown below.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">DNA</span>, highlight<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>All <strong>DECIPHER</strong> multiple sequence alignments are
optimized using <code>AdjustAlignment</code> (unless the input argument
<code>FUN</code> is changed), with the goal of removing artifacts of the
progressive alignment process. This function will efficiently correct
most obvious inaccuracies that could be found by-eye. Therefore, making
manual corrections is not recommended unless additional expert knowledge
of the sequences is available. The advantage of using
<code>AdjustAlignment</code> is that it is a repeatable process that is
not subjective, unlike most manual adjustments. In order to further
refine an existing alignment, <code>AdjustAlignment</code> can be called
directly.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DNA_adjusted</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AdjustAlignment.html">AdjustAlignment</a></span><span class="op">(</span><span class="va">DNA</span><span class="op">)</span></span></code></pre></div>
<p>It is common to use alignment as a preliminary step before the
creation of a phylogenetic tree. <strong>DECIPHER</strong>, like the
majority of alignment programs, attempts to maximize homologous
positions between the sequences being aligned. Such an alignment is
particularly useful when investigating which residues are in the same
structural position of a protein. However, disparate sequence regions
tend to be concentrated into the same “gappy” areas of the alignment.
When viewed from a phylogenetic perspective these homologies have highly
implausible insertion/deletion scenarios.</p>
<p>To mitigate the problem of false homologies,
<code>StaggerAlignment</code> will automatically generate a staggered
version of an existing alignment. Staggered alignments separate
potentially non-homologous regions into separate columns of the
alignment. The result is an alignment that is less visually appealing,
but likely more accurate from a phylogenetic perspective.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DNA_staggered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/StaggerAlignment.html">StaggerAlignment</a></span><span class="op">(</span><span class="va">DNA</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="aligning-homologous-regions-of-multiple-genomes">Aligning Homologous Regions of Multiple Genomes<a class="anchor" aria-label="anchor" href="#aligning-homologous-regions-of-multiple-genomes"></a>
</h2>
<p>The functions described so far have all required collinear sequences
as input. This requirement is frequently broken by genomes, which may
include many sequence rearrangements such as inversion, duplication, and
reordering. <code>FindSynteny</code> will find homologous regions
between pairs of genomes, which can then be aligned using
<code>AlignSynteny</code>. A database of sequences identified by their
genome name is used as input to both functions. This enables the
alignment of genomes that are composed of many contigs, so long as they
all share the same <code>identifier</code> in the database. The example
below uses a database containing five <em>Influenza virus A</em>
genomes, which are each composed of eight separate segments.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Influenza.sqlite"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span>
<span><span class="va">synteny</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindSynteny.html">FindSynteny</a></span><span class="op">(</span><span class="va">db</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">synteny</span> <span class="co"># an object of class `Synteny`</span></span></code></pre></div>
<pre><code><span><span class="co">##          H9N2     H5N1     H2N2     H7N9     H1N1</span></span>
<span><span class="co">## H9N2   8 seqs 74% hits 78% hits 76% hits 73% hits</span></span>
<span><span class="co">## H5N1 8 blocks   8 seqs 70% hits 74% hits 84% hits</span></span>
<span><span class="co">## H2N2 8 blocks 8 blocks   8 seqs 73% hits 71% hits</span></span>
<span><span class="co">## H7N9 8 blocks 8 blocks 8 blocks   8 seqs 74% hits</span></span>
<span><span class="co">## H1N1 8 blocks 8 blocks 8 blocks 8 blocks   8 seqs</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">InfluenzaA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AlignSynteny.html">AlignSynteny</a></span><span class="op">(</span><span class="va">synteny</span>, <span class="va">db</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">InfluenzaA</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DNAStringSet object of length 16:</span></span>
<span><span class="co">##      width seq                                              names               </span></span>
<span><span class="co">##  [1]  2328 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AAAAAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span> 1.H9N2</span></span>
<span><span class="co">##  [2]  2328 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AAAAAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span> 1.H5N1</span></span>
<span><span class="co">##  [3]  2339 <span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span>...<span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAAAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span> 2.H9N2</span></span>
<span><span class="co">##  [4]  2339 <span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span>...<span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAAAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span> 2.H5N1</span></span>
<span><span class="co">##  [5]  2225 <span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span>...<span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AAAAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFAF;">C</span> 3.H9N2</span></span>
<span><span class="co">##  ...   ... ...</span></span>
<span><span class="co">## [12]   865 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">CC</span>...<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AAAAAA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span> 6.H5N1</span></span>
<span><span class="co">## [13]  1453 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span>...<span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span> 7.H9N2</span></span>
<span><span class="co">## [14]  1453 <span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span>...<span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span> 7.H5N1</span></span>
<span><span class="co">## [15]  1317 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AAAA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span>...<span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">A</span> 8.H9N2</span></span>
<span><span class="co">## [16]  1317 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span>...<span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #AFFFAF;">CCC</span> 8.H5N1</span></span></code></pre>
<p>The output is a list, with each list component containing a
<code>DNAStringSetList</code> of pairwise alignments between two
genomes. Names of the output correspond to their sequence’s
<code>identifier</code> in the database, and the index of the syntenic
block.</p>
<p>It is also possible to display the blocks of synteny between all
pairs of genomes. Figure @ref(f8) shows the different genome segments
(i.e., sequences) separated by thin horizontal and vertical lines. The
syntenic blocks are diagonal lines that are composed of many homologous
“hits” between the genomes.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/pairs.html" class="external-link">pairs</a></span><span class="op">(</span><span class="va">synteny</span>, boxBlocks<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co"># scatterplot matrix</span></span></code></pre></div>
<p><img src="ArtOfAlignmentInR_files/figure-html/expr12-1.png" width="700"></p>
<p>{#f8} Dot plots showing the homologous regions among five
<em>Influenza virus A</em> genomes.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>All of the output in this vignette was produced under the following
conditions:</p>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=zh_CN.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=zh_CN.UTF-8        LC_COLLATE=zh_CN.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=zh_CN.UTF-8    LC_MESSAGES=zh_CN.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=zh_CN.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=zh_CN.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Asia/Shanghai</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] DECIPHER_2.29.1     RSQLite_2.3.1       Biostrings_2.68.1  </span></span>
<span><span class="co">## [4] GenomeInfoDb_1.36.1 XVector_0.40.0      IRanges_2.34.1     </span></span>
<span><span class="co">## [7] S4Vectors_0.38.1    BiocGenerics_0.46.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] bit_4.0.5               jsonlite_1.8.7          highr_0.10             </span></span>
<span><span class="co">##  [4] crayon_1.5.2            compiler_4.3.1          blob_1.2.4             </span></span>
<span><span class="co">##  [7] stringr_1.5.0           bitops_1.0-7            jquerylib_0.1.4        </span></span>
<span><span class="co">## [10] systemfonts_1.0.4       textshaping_0.3.6       yaml_2.3.7             </span></span>
<span><span class="co">## [13] fastmap_1.1.1           R6_2.5.1                knitr_1.43             </span></span>
<span><span class="co">## [16] desc_1.4.2              rprojroot_2.0.3         DBI_1.1.3              </span></span>
<span><span class="co">## [19] GenomeInfoDbData_1.2.10 bslib_0.5.1             rlang_1.1.1            </span></span>
<span><span class="co">## [22] cachem_1.0.8            stringi_1.7.12          xfun_0.40              </span></span>
<span><span class="co">## [25] fs_1.6.3                sass_0.4.7              bit64_4.0.5            </span></span>
<span><span class="co">## [28] memoise_2.0.1           cli_3.6.1               pkgdown_2.0.7          </span></span>
<span><span class="co">## [31] magrittr_2.0.3          zlibbioc_1.46.0         digest_0.6.33          </span></span>
<span><span class="co">## [34] rstudioapi_0.15.0       lifecycle_1.0.3         vctrs_0.6.3            </span></span>
<span><span class="co">## [37] evaluate_0.21           glue_1.6.2              ragg_1.2.5             </span></span>
<span><span class="co">## [40] RCurl_1.98-1.12         rmarkdown_2.24          purrr_1.0.2            </span></span>
<span><span class="co">## [43] pkgconfig_2.0.3         tools_4.3.1             htmltools_0.5.6</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
