<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DECIPHER">
<title>Design Primers That Yield Group-Specific Signatures • DECIPHER</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Design Primers That Yield Group-Specific Signatures">
<meta property="og:description" content="DECIPHER">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Design Primers That Yield Group-Specific Signatures</h1>
                        <h4 data-toc-skip class="author">Erik S.
Wright</h4>
            
            <h4 data-toc-skip class="date">2023-08-24</h4>
      
      
      <div class="d-none name"><code>DesignSignatures.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <strong>DECIPHER</strong> function <code>DesignSignatures</code>
was created to perform the most common primer design task: to
efficiently amplify a shared region of DNA that will produce maximal
amplicon diversity. This goal is independent of the method being used to
differentiate amplicons, although some of the design details may depend
on the technique being employed. <code>DesignSignatures</code> supports
three common methods for distinguishing amplicons: sequencing, High
Resolution Melting (HRM) analysis, and Fragment Length Polymorphism
(FLP) analysis. As a case study, this tutorial focuses on designing
primers to differentiate SNP alleles using HRM analysis. However, the
methods described here can easily be adjusted to the other experimental
techniques.</p>
<p>As input, we will use known sequence variants of the Isocitrate
Dehydrogenase 2 ( gene that have been implicated in human gliomas.
First, a database is constructed where the variants are identified by
their unique name. Next, this database is used to design primers that
will yield maximal amplicon diversity. The top scoring primers are
further investigated to determine whether they can easily be
distinguished experimentally. Finally, the process is repeated using a
restriction enzyme to further diversify the amplicons’ signatures. Note
that much of the functionality described here is accessible via a web
tool at <a href="http://DECIPHER.codes" class="external-link uri">http://DECIPHER.codes</a>.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="startup">Startup<a class="anchor" aria-label="anchor" href="#startup"></a>
</h3>
<p>To get started we need to load the <strong>DECIPHER</strong> package,
which automatically loads several other required packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DECIPHER</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: Biostrings</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     findMatches</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: XVector</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biostrings'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     strsplit</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: RSQLite</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: parallel</span></span></code></pre>
<p>Help for the <code>DesignSignatures</code> function can be accessed
through:</p>
<pre><code>&gt; ? DesignSignatures</code></pre>
<p>If <strong>DECIPHER</strong> is installed on your system, the code in
each example can be obtained via:</p>
<pre><code>&gt; browseVignettes("DECIPHER")</code></pre>
</div>
<div class="section level3">
<h3 id="creating-a-sequence-database">Creating a Sequence Database<a class="anchor" aria-label="anchor" href="#creating-a-sequence-database"></a>
</h3>
<p>We begin with a set of <em>IDH2</em> sequences in a FASTA file that
is included as part of the <strong>DECIPHER</strong> package. The file
contains the common IDH2 allele, and three alleles with single
nucleotide polymorphisms (SNPs). Our goal is to design primers that can
distinguish all four variants of the IDH2 gene. If you wish to follow
along with your own FASTA file of unaligned sequences, be sure to change
the path names to those on your system by replacing all of the text
inside quotes labeled “<span class="math inline">\(&lt;&lt;\)</span>path
to …<span class="math inline">\(&gt;&gt;\)</span>” with the actual path
on your system.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify the path to your sequence file:</span></span>
<span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="st">"&lt;&lt;path to FASTA file&gt;&gt;"</span></span>
<span><span class="co"># OR find the example sequence file used in this tutorial:</span></span>
<span><span class="va">fas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"IDH2.fas"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></code></pre></div>
<p>Next, there are two options for importing the sequences into a
database: either save a database file or maintain the database in
memory. Here we will build the database in memory because it is a small
set of sequences and we do not intend to use the database later:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify a path for where to write the sequence database</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="st">"&lt;&lt;path to write sequence database&gt;&gt;"</span></span>
<span><span class="co"># OR create the sequence database in memory</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">":memory:"</span><span class="op">)</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">fas</span>, <span class="st">"FASTA"</span>, <span class="va">dbConn</span>, <span class="st">""</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
Reading FASTA file chunk 1
## 
## 4 total sequences in table Seqs.
## Time difference of 0.04 secs</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="co"># number of sequences in the database</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="defining-groups">Defining Groups<a class="anchor" aria-label="anchor" href="#defining-groups"></a>
</h3>
<p>At this point it is necessary to define groups of related sequences
in the database we have just created. Ideally, groups should designate
the different variants (i.e., strains or alleles) that we hope to
distinguish.</p>
<p>In this example, we will define groups based on the allele’s name,
which was included in the description of each FASTA record that we
imported. This is the most common case, and simply uses each sequence’s
name as its group name.</p>
<p>To assign group names directly from the sequence names (FASTA
identifiers), simply run the code shown below. In more complex cases, we
could use the functions <code>IdentifyByRank</code>,
<code>FormGroups</code>, <code>Clusterize</code>, or
<code>TreeLine</code> to define groups.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># if each sequence belongs to its own group,</span></span>
<span><span class="co"># then identify the sequences with a number:</span></span>
<span><span class="va">desc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">N</span><span class="op">)</span><span class="op">)</span> <span class="co"># N is the number of sequences</span></span>
<span><span class="co"># OR get the FASTA record description:</span></span>
<span><span class="va">desc</span> <span class="op">&lt;-</span> <span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">dbConn</span>, <span class="st">"select description from Seqs"</span><span class="op">)</span><span class="op">$</span><span class="va">description</span></span>
<span><span class="co"># show the unique descriptors:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">desc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "IDH2"  "R172K" "R172G" "R172M"</span></span></code></pre>
<p>Now that we have unique names for our four variants, we must add them
to the database as the identifier of each sequence.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Add2DB.html">Add2DB</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>identifier<span class="op">=</span><span class="va">desc</span>, stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>, <span class="va">dbConn</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Expression:</span></span>
<span><span class="co">## update Seqs set identifier = :identifier where row_names = :row_names</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Added to table Seqs:  "identifier".</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 0 secs</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="designing-primers-with-diverse-signatures">Designing primers with diverse signatures<a class="anchor" aria-label="anchor" href="#designing-primers-with-diverse-signatures"></a>
</h2>
<div class="section level3">
<h3 id="adjusting-input-parameters">Adjusting Input Parameters<a class="anchor" aria-label="anchor" href="#adjusting-input-parameters"></a>
</h3>
<p>Before using our database to design primers, we must carefully
consider the inputs that will be used. These inputs can have a large
impact on the outcome, and therefore should be tailored to the
experimental goal. Below are recommended inputs for sequencing, FLP, and
HRM:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Designing primers for sequencing experiments:</span></span>
<span><span class="va">TYPE</span> <span class="op">&lt;-</span> <span class="st">"sequence"</span></span>
<span><span class="va">MIN_SIZE</span> <span class="op">&lt;-</span> <span class="fl">300</span> <span class="co"># base pairs</span></span>
<span><span class="va">MAX_SIZE</span> <span class="op">&lt;-</span> <span class="fl">700</span></span>
<span><span class="va">RESOLUTION</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co"># k-mer signature</span></span>
<span><span class="va">LEVELS</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co"># max number of each k-mer</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Designing primers for community fingerprinting (FLP):</span></span>
<span><span class="va">TYPE</span> <span class="op">&lt;-</span> <span class="st">"length"</span></span>
<span><span class="co"># it is important to have a width range of lengths</span></span>
<span><span class="va">MIN_SIZE</span> <span class="op">&lt;-</span> <span class="fl">200</span> <span class="co"># base pairs</span></span>
<span><span class="va">MAX_SIZE</span> <span class="op">&lt;-</span> <span class="fl">1400</span></span>
<span><span class="co"># define bin boundaries for distinguishing length,</span></span>
<span><span class="co"># the values below require high-resolution, but</span></span>
<span><span class="co"># the bin boundaries can be redefined for lower</span></span>
<span><span class="co"># resolution experiments such as gel runs</span></span>
<span><span class="va">RESOLUTION</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">700</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">705</span>, <span class="fl">1000</span>, <span class="fl">5</span><span class="op">)</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1010</span>, <span class="fl">1400</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">LEVELS</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="co"># presence/absence of the length</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Designing primers for high resolution melting (HRM):</span></span>
<span><span class="va">TYPE</span> <span class="op">&lt;-</span> <span class="st">"melt"</span></span>
<span><span class="va">MIN_SIZE</span> <span class="op">&lt;-</span> <span class="fl">55</span> <span class="co"># base pairs</span></span>
<span><span class="va">MAX_SIZE</span> <span class="op">&lt;-</span> <span class="fl">400</span></span>
<span><span class="co"># the recommended values for resolution</span></span>
<span><span class="va">RESOLUTION</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">100</span>, <span class="fl">0.25</span><span class="op">)</span> <span class="co"># degrees Celsius</span></span>
<span><span class="va">LEVELS</span> <span class="op">&lt;-</span> <span class="fl">10</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="select-restriction-enzymes">Select Restriction Enzymes<a class="anchor" aria-label="anchor" href="#select-restriction-enzymes"></a>
</h3>
<p>The following step is not applicable to sequencing experiments (i.e.,
the <code>TYPE</code> above is <code>"sequence"</code>), and is optional
for other types of experiments. If using HRM or FLP analysis to
differentiate amplicon signatures, a digestion step following PCR
amplification may generate fragments with widely different signatures.
This is especially useful when trying to distinguish a large number of
different groups based on small variations. Potential enzymes can be
selected from the built-in <code>RESTRICTION_ENZYMES</code> dataset,
which includes all of the standard enzymes sold by </p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ENZYMES</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> <span class="co"># required for sequencing</span></span>
<span><span class="co"># OR select restriction enzymes to consider</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">RESTRICTION_ENZYMES</span><span class="op">)</span> <span class="co"># load available enzymes</span></span>
<span><span class="co"># for this tutorial we will use the enzyme MslI</span></span>
<span><span class="va">ENZYMES</span> <span class="op">&lt;-</span> <span class="va">RESTRICTION_ENZYMES</span><span class="op">[</span><span class="st">"MslI"</span><span class="op">]</span></span>
<span><span class="va">ENZYMES</span></span></code></pre></div>
<pre><code><span><span class="co">##          MslI </span></span>
<span><span class="co">## "CAYNN/NNRTG"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="designing-primers">Designing Primers<a class="anchor" aria-label="anchor" href="#designing-primers"></a>
</h3>
<p>Now we can design primers that target all of the groups and result in
maximal amplicon diversity:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignSignatures.html">DesignSignatures</a></span><span class="op">(</span><span class="va">dbConn</span>,</span>
<span>                            type<span class="op">=</span><span class="va">TYPE</span>,</span>
<span>                            minProductSize<span class="op">=</span><span class="va">MIN_SIZE</span>,</span>
<span>                            maxProductSize<span class="op">=</span><span class="va">MAX_SIZE</span>,</span>
<span>                            resolution<span class="op">=</span><span class="va">RESOLUTION</span>,</span>
<span>                            levels<span class="op">=</span><span class="va">LEVELS</span>,</span>
<span>                            enzymes<span class="op">=</span><span class="va">ENZYMES</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Tallying 8-mers for 4 groups:</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 0.11 secs</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Designing primer sequences based on the group 'IDH2':</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 24.93 secs</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Selecting the most common primer sequences:</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 3.23 secs</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Determining PCR products from each group:</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 83.54 secs</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Scoring primer pair combinations:</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 0.34 secs</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Choosing optimal forward and reverse pairs:</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 0.51 secs</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Finding the best restriction enzyme:</span></span>
<span><span class="co">## ================================================================================</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Time difference of 2.51 secs</span></span></code></pre>
<p>The output <code>data.frame</code> is sorted by the sum of
<code>"score"</code> and <code>"digest_score"</code>. However, we can
separately look at the top scoring primers with and without using
restriction enzymes.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">primers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">score</span><span class="op">)</span>,<span class="op">]</span> <span class="co"># best primers without digestion</span></span></code></pre></div>
<pre><code><span><span class="co">##       forward_primer       reverse_primer        score coverage products</span></span>
<span><span class="co">## 48 TGGCTGGACCAAGCCCA CCTTGTACTGGTCGCCATGG 0.041666....        1        4</span></span>
<span><span class="co">##    similar_signatures missing_signatures enzyme digest_score fragments</span></span>
<span><span class="co">## 48                                         MslI            0         2</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">primers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">digest_score</span><span class="op">)</span>,<span class="op">]</span> <span class="co"># best primers with digestion</span></span></code></pre></div>
<pre><code><span><span class="co">##      forward_primer       reverse_primer        score coverage products</span></span>
<span><span class="co">## 1 TGGCTGGACCAAGCCCA CTGTGGCCTTGTACTGGTCG 0.035555....        1        4</span></span>
<span><span class="co">##   similar_signatures missing_signatures enzyme digest_score fragments</span></span>
<span><span class="co">## 1                                         MslI    0.1805556        13</span></span></code></pre>
<p>The PCR products digested with <em>MslI</em> are substantially higher
scoring (<code>digest_score</code>) than the undigested primers
(<code>score</code>). Note that the columns
<code>"similar_signatures"</code> and <code>"missing_signatures"</code>
are both empty for this primer set, indicating that all of the groups
have distinct signatures after digestion with the restriction
enzyme.</p>
</div>
</div>
<div class="section level2">
<h2 id="assessing-the-results">Assessing the Results<a class="anchor" aria-label="anchor" href="#assessing-the-results"></a>
</h2>
<div class="section level3">
<h3 id="view-the-target-sites">View the Target Sites<a class="anchor" aria-label="anchor" href="#view-the-target-sites"></a>
</h3>
<p>We can now look at the target sites of the top scoring primers and
restriction enzyme.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PSET</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># examine the top scoring primer set overall</span></span>
<span></span>
<span><span class="co"># select the first sequence from each group</span></span>
<span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span>,</span>
<span>                remove<span class="op">=</span><span class="st">"all"</span>,</span>
<span>                nameBy<span class="op">=</span><span class="st">"identifier"</span>,</span>
<span>                clause<span class="op">=</span><span class="st">"row_names =</span></span>
<span><span class="st">                        (select min(row_names) from Seqs as S</span></span>
<span><span class="st">                         where S.identifier = Seqs.identifier)"</span>,</span>
<span>                verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">f_primer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">forward_primer</span><span class="op">[</span><span class="va">PSET</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">r_primer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">reverse_primer</span><span class="op">[</span><span class="va">PSET</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">patterns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">f_primer</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/reverseComplement.html" class="external-link">reverseComplement</a></span><span class="op">(</span><span class="va">r_primer</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"[^A-Z]"</span>, <span class="st">""</span>, <span class="va">ENZYMES</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">dna</span>,</span>
<span>           patterns<span class="op">=</span><span class="va">patterns</span><span class="op">)</span></span></code></pre></div>
<p>We can see in Figure @ref(f1) that the forward and reverse primers
target sites on both sides of the SNP, which is located approximately in
the center of the amplicon. There are restriction sites next to the SNP,
which will shorten the amplicons and further spread their melting
curves. Furthermore, the final allele has two adjacent restriction
sites, whereas the others all have one.</p>
</div>
<div class="section level3">
<h3 id="plot-the-signatures">Plot the Signatures<a class="anchor" aria-label="anchor" href="#plot-the-signatures"></a>
</h3>
<p>Next we will compare the signatures of the top scoring primers with
and without digestion. The simplest way to accomplish this is to use the
<strong>Biostrings</strong> function <code>matchLRPatterns</code> to
find the amplicons. Refer to section~@ref(sec:startup) above for how to
easily obtain the code shown below:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PSET</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">score</span><span class="op">)</span> <span class="co"># top scoring without digestion</span></span>
<span></span>
<span><span class="va">f_primer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html" class="external-link">DNAString</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">forward_primer</span><span class="op">[</span><span class="va">PSET</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">r_primer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html" class="external-link">DNAString</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">reverse_primer</span><span class="op">[</span><span class="va">PSET</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">r_primer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/reverseComplement.html" class="external-link">reverseComplement</a></span><span class="op">(</span><span class="va">r_primer</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">dbConn</span>, <span class="st">"select distinct identifier from Seqs"</span><span class="op">)</span></span>
<span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="va">ids</span><span class="op">$</span><span class="va">identifier</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">TYPE</span><span class="op">==</span><span class="st">"sequence"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow<span class="op">=</span><span class="fl">4</span><span class="op">^</span><span class="va">RESOLUTION</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">TYPE</span><span class="op">==</span><span class="st">"melt"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">RESOLUTION</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span> <span class="co"># TYPE=="length"</span></span>
<span>    <span class="va">signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">RESOLUTION</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">signatures</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/abbreviate.html" class="external-link">abbreviate</a></span><span class="op">(</span><span class="va">ids</span>, <span class="fl">15</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span>, identifier<span class="op">=</span><span class="va">ids</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, remove<span class="op">=</span><span class="st">"all"</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="va">amplicons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/matchLRPatterns.html" class="external-link">matchLRPatterns</a></span><span class="op">(</span><span class="va">f_primer</span>, <span class="va">r_primer</span>,</span>
<span>                                 <span class="va">MAX_SIZE</span>, <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span>,</span>
<span>                                 max.Lmismatch<span class="op">=</span><span class="fl">2</span>, max.Rmismatch<span class="op">=</span><span class="fl">2</span>,</span>
<span>                                 Lfixed<span class="op">=</span><span class="st">"subject"</span>, Rfixed<span class="op">=</span><span class="st">"subject"</span><span class="op">)</span></span>
<span>    <span class="va">amplicons</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">amplicons</span>, <span class="st">"DNAStringSet"</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">amplicons</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span></span>
<span>        <span class="kw">next</span></span>
<span>    </span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">TYPE</span><span class="op">==</span><span class="st">"sequence"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/nucleotideFrequency.html" class="external-link">oligonucleotideFrequency</a></span><span class="op">(</span><span class="va">amplicons</span>, <span class="va">RESOLUTION</span><span class="op">)</span></span>
<span>        <span class="va">signatures</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">TYPE</span><span class="op">==</span><span class="st">"melt"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MeltDNA.html">MeltDNA</a></span><span class="op">(</span><span class="va">amplicons</span>, <span class="st">"melt curves"</span>, <span class="va">RESOLUTION</span><span class="op">)</span></span>
<span>        <span class="co"># weight melting curves by their amlicon's width</span></span>
<span>        <span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span><span class="op">*</span><span class="fu">width</span><span class="op">(</span><span class="va">amplicons</span><span class="op">)</span></span>
<span>        <span class="va">signatures</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">amplicons</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span> <span class="co"># TYPE=="length"</span></span>
<span>        <span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/bincode.html" class="external-link">.bincode</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">amplicons</span><span class="op">)</span>, <span class="va">RESOLUTION</span><span class="op">)</span></span>
<span>        <span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="va">signature</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>            <span class="va">signatures</span><span class="op">[</span><span class="va">j</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">signatures</span><span class="op">[</span><span class="va">j</span>, <span class="va">i</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">TYPE</span><span class="op">==</span><span class="st">"sequence"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">signatures</span><span class="op">)</span>, <span class="st">"minkowski"</span>, p<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="co"># L1-Norm</span></span>
<span>    <span class="fu"><a href="../reference/TreeLine.html">TreeLine</a></span><span class="op">(</span>myDistMatrix<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"UPGMA"</span>, showPlot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/mtext.html" class="external-link">mtext</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">RESOLUTION</span>, <span class="st">"-mer Profile Distance"</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span>,</span>
<span>        side<span class="op">=</span><span class="fl">2</span>, padj<span class="op">=</span><span class="op">-</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">TYPE</span><span class="op">==</span><span class="st">"melt"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/matplot.html" class="external-link">matplot</a></span><span class="op">(</span><span class="va">RESOLUTION</span>, <span class="va">signatures</span>, type<span class="op">=</span><span class="st">"l"</span>,</span>
<span>        xlab<span class="op">=</span><span class="st">"Temperature (degrees Celsius)"</span>, ylab<span class="op">=</span><span class="st">"Average Helicity"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span> <span class="co"># TYPE=="length"</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">20</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="cn">NA</span>,</span>
<span>            xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">RESOLUTION</span><span class="op">)</span>,</span>
<span>            xlab<span class="op">=</span><span class="st">"Group Index"</span>, ylab<span class="op">=</span><span class="st">"Amplicon Length"</span>,</span>
<span>            yaxs<span class="op">=</span><span class="st">"i"</span>, xaxs<span class="op">=</span><span class="st">"i"</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">1</span>, at<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span>, labels<span class="op">=</span><span class="cn">FALSE</span>, tck<span class="op">=</span><span class="op">-</span><span class="fl">0.01</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="cn">NA</span>,</span>
<span>            xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">RESOLUTION</span><span class="op">)</span>,</span>
<span>            xlab<span class="op">=</span><span class="st">""</span>, ylab<span class="op">=</span><span class="st">"Amplicon Length"</span>,</span>
<span>            yaxs<span class="op">=</span><span class="st">"i"</span>, xaxs<span class="op">=</span><span class="st">"i"</span>, xaxt<span class="op">=</span><span class="st">"n"</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">1</span>, at<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span>, labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/abbreviate.html" class="external-link">abbreviate</a></span><span class="op">(</span><span class="va">ids</span>, <span class="fl">7</span><span class="op">)</span>, las<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">xaxs</span> <span class="op">&lt;-</span> <span class="va">RESOLUTION</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">RESOLUTION</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span> <span class="co"># average lengths</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">signatures</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">w</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/graphics/segments.html" class="external-link">segments</a></span><span class="op">(</span><span class="va">i</span> <span class="op">-</span> <span class="fl">0.45</span>, <span class="va">xaxs</span><span class="op">[</span><span class="va">w</span><span class="op">]</span>, <span class="va">i</span> <span class="op">+</span> <span class="fl">0.45</span>, <span class="va">xaxs</span><span class="op">[</span><span class="va">w</span><span class="op">]</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="DesignSignatures_files/figure-html/expr10-1.png" width="700"></p>
<p>Figure @ref(f2) shows that two alleles will be indistinguishable when
using these primers because their melting curves are nearly overlapping.
We will now compare the result after using a restriction enzyme. Note
that this is only possible because we previously specified a restriction
enzyme in the input to <code>DesignSignatures</code>.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PSET</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">digest_score</span><span class="op">)</span> <span class="co"># top scoring with digestion</span></span>
<span></span>
<span><span class="va">f_primer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html" class="external-link">DNAString</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">forward_primer</span><span class="op">[</span><span class="va">PSET</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">r_primer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html" class="external-link">DNAString</a></span><span class="op">(</span><span class="va">primers</span><span class="op">$</span><span class="va">reverse_primer</span><span class="op">[</span><span class="va">PSET</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">r_primer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/reverseComplement.html" class="external-link">reverseComplement</a></span><span class="op">(</span><span class="va">r_primer</span><span class="op">)</span></span>
<span><span class="va">enzyme</span> <span class="op">&lt;-</span> <span class="va">RESTRICTION_ENZYMES</span><span class="op">[</span><span class="va">primers</span><span class="op">$</span><span class="va">enzyme</span><span class="op">[</span><span class="va">PSET</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">signatures</span><span class="op">[</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span> <span class="co"># initialize the results matrix used previously</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span>, identifier<span class="op">=</span><span class="va">ids</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, remove<span class="op">=</span><span class="st">"all"</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="va">amplicons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/matchLRPatterns.html" class="external-link">matchLRPatterns</a></span><span class="op">(</span><span class="va">f_primer</span>, <span class="va">r_primer</span>,</span>
<span>                                 <span class="va">MAX_SIZE</span>, <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span>,</span>
<span>                                 max.Lmismatch<span class="op">=</span><span class="fl">2</span>, max.Rmismatch<span class="op">=</span><span class="fl">2</span>,</span>
<span>                                 Lfixed<span class="op">=</span><span class="st">"subject"</span>, Rfixed<span class="op">=</span><span class="st">"subject"</span><span class="op">)</span></span>
<span>    <span class="va">amplicons</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">amplicons</span>, <span class="st">"DNAStringSet"</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">amplicons</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span></span>
<span>        <span class="kw">next</span></span>
<span>    <span class="va">digested</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DigestDNA.html">DigestDNA</a></span><span class="op">(</span><span class="va">enzyme</span>, <span class="va">amplicons</span>, strand<span class="op">=</span><span class="st">"top"</span><span class="op">)</span></span>
<span>    <span class="va">digested</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">digested</span><span class="op">)</span> <span class="co"># convert to DNAStringSet</span></span>
<span>    </span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">TYPE</span><span class="op">==</span><span class="st">"melt"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MeltDNA.html">MeltDNA</a></span><span class="op">(</span><span class="va">digested</span>, <span class="st">"melt curves"</span>, <span class="va">RESOLUTION</span><span class="op">)</span></span>
<span>        <span class="co"># weight melting curves by their fragment's width</span></span>
<span>        <span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span><span class="op">*</span><span class="fu">width</span><span class="op">(</span><span class="va">digested</span><span class="op">)</span></span>
<span>        <span class="va">signatures</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">digested</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span> <span class="co"># TYPE=="length"</span></span>
<span>        <span class="va">signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/bincode.html" class="external-link">.bincode</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">digested</span><span class="op">)</span>, <span class="va">RESOLUTION</span><span class="op">)</span></span>
<span>        <span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="va">signature</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>            <span class="va">signatures</span><span class="op">[</span><span class="va">j</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">signatures</span><span class="op">[</span><span class="va">j</span>, <span class="va">i</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">signature</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">TYPE</span><span class="op">==</span><span class="st">"melt"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/matplot.html" class="external-link">matplot</a></span><span class="op">(</span><span class="va">RESOLUTION</span>, <span class="va">signatures</span>, type<span class="op">=</span><span class="st">"l"</span>,</span>
<span>        xlab<span class="op">=</span><span class="st">"Temperature (degrees Celsius)"</span>, ylab<span class="op">=</span><span class="st">"Average Helicity"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span> <span class="co"># TYPE=="length"</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">20</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="cn">NA</span>,</span>
<span>            xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">RESOLUTION</span><span class="op">)</span>,</span>
<span>            xlab<span class="op">=</span><span class="st">"Group Index"</span>, ylab<span class="op">=</span><span class="st">"Amplicon Length"</span>,</span>
<span>            yaxs<span class="op">=</span><span class="st">"i"</span>, xaxs<span class="op">=</span><span class="st">"i"</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">1</span>, at<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span>, labels<span class="op">=</span><span class="cn">FALSE</span>, tck<span class="op">=</span><span class="op">-</span><span class="fl">0.01</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="cn">NA</span>,</span>
<span>            xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">RESOLUTION</span><span class="op">)</span>,</span>
<span>            xlab<span class="op">=</span><span class="st">""</span>, ylab<span class="op">=</span><span class="st">"Amplicon Length"</span>,</span>
<span>            yaxs<span class="op">=</span><span class="st">"i"</span>, xaxs<span class="op">=</span><span class="st">"i"</span>, xaxt<span class="op">=</span><span class="st">"n"</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">1</span>, at<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span>, labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/abbreviate.html" class="external-link">abbreviate</a></span><span class="op">(</span><span class="va">ids</span>, <span class="fl">7</span><span class="op">)</span>, las<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">xaxs</span> <span class="op">&lt;-</span> <span class="va">RESOLUTION</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">RESOLUTION</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span> <span class="co"># average lengths</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">signatures</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">w</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/graphics/segments.html" class="external-link">segments</a></span><span class="op">(</span><span class="va">i</span> <span class="op">-</span> <span class="fl">0.45</span>, <span class="va">xaxs</span><span class="op">[</span><span class="va">w</span><span class="op">]</span>, <span class="va">i</span> <span class="op">+</span> <span class="fl">0.45</span>, <span class="va">xaxs</span><span class="op">[</span><span class="va">w</span><span class="op">]</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="DesignSignatures_files/figure-html/expr11-1.png" width="700"></p>
<p>Figure @ref(f3) shows that the two amplicons which had nearly
identical melt curves can be easily distinguished after restriction
digest.</p>
</div>
<div class="section level3">
<h3 id="finishing-up">Finishing Up<a class="anchor" aria-label="anchor" href="#finishing-up"></a>
</h3>
<p>Finally, we can order the top scoring forward and reverse primers for
synthesis and try them out in a PCR reaction! The primers should be
synthesized in the same orientation given in the output
(<code>primers</code>).</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
