<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DECIPHER">
<title>Design Group-Specific FISH Probes • DECIPHER</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Design Group-Specific FISH Probes">
<meta property="og:description" content="DECIPHER">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Design Group-Specific FISH Probes</h1>
                        <h4 data-toc-skip class="author">Erik S.
Wright</h4>
            
            <h4 data-toc-skip class="date">2023-08-24</h4>
      
      
      <div class="d-none name"><code>DesignProbes.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This document describes how to design group-specific FISH probes
using the <strong>DECIPHER</strong> package through the use of the
<code>DesignProbes</code> function. Fluorescent In-Situ Hybridization
(FISH) is an laboratory technique that enables visualizing specific
microorganisms in their natural environment using fluorescent
microscopy. As a case study, this tutorial focuses on the small-subunit
ribosomal RNA (SSU rRNA) from a collection of bacteria found in drinking
water. Here we describe how to design probes targeting a group of
interest that will not cross-hybridize with other bacteria from the same
sample. A similar strategy could be used with any set of aligned
sequences that are separated into groups. For example, genus-specific
probes designed with this program that target all named bacteria and
archaea genera are provided online at <a href="http://DECIPHER.codes/16SOligos.html" class="external-link uri">http://DECIPHER.codes/16SOligos.html</a>.</p>
<p>A database of aligned DNA sequences separated into groups is used as
input to the program. First the function <code>TileSeqs</code> is used
to pre-process the sequences into overlapping tiles, which will serve as
the template DNA for probe design. Second, the <code>DesignProbes</code>
function determines the set of all possible probes that meet certain
design constraints, such as the ability to hybridize with the group of
interest under specified experimental conditions. Next, the complete set
of probes is scored by its potential to cross-hybridize with sequences
from other groups including a comprehensive database of non-target
sequences. Finally, the optimal set of dual probes is chosen that could
be used in a FISH experiment to identify the microbes of interest.</p>
</div>
<div class="section level2">
<h2 id="the-objective-of-fish-probe-design">The Objective of FISH Probe Design<a class="anchor" aria-label="anchor" href="#the-objective-of-fish-probe-design"></a>
</h2>
<p>The objective of FISH probe design is to balance sensitivity of the
probe to the target group of sequences, while simultaneously maximizing
specificity to prevent cross-hybridizations with other non-target
groups. Here there is a fundamental trade-off, because if a probe is too
weak it will not bind to its target, but if too strong it will also bind
to non-targets. This balancing act is complicated by the difficulty of
accurately predicting binding strength of the probe with a variety of
potential target and non-target sequences. To confront this challenge,
<code>DesignProbes</code> uses a state of the art model of probe
hybridization in the presence of the denaturant formamide. This model,
implemented in the function <code>CalculateEfficiencyFISH</code>, is
used in conjunction with reasonable safety factors to minimize the need
for experimental optimization.</p>
<p>In cases where the desired level of specificity cannot be achieved,
<code>DesignProbes</code> predicts the degree of potential
cross-hybridization with non-target groups. When a single probe is
insufficient, dual probes with distinct colors can be employed to
further increase specificity to the target group.
<code>DesignProbes</code> will choose the optimal combination of dual
probes that will minimize cross-hybridization overlap. During the
experiment, the combination of the individual probe colors provides
additional assurance of a positive identification. For example, if one
probe is labeled green (e.g., fluorescein) and the other is labeled red
(e.g., Cy3) then the hybridization of both probes at the same point
would appear yellow.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="installing-oligoarrayaux">Installing OligoArrayAux<a class="anchor" aria-label="anchor" href="#installing-oligoarrayaux"></a>
</h3>
<p>The program OligoArrayAux (<a href="http://www.unafold.org/Dinamelt/software/oligoarrayaux.php" class="external-link uri">http://www.unafold.org/Dinamelt/software/oligoarrayaux.php</a>)
is used to predict hybridization efficiency and must be installed in a
location accessible by the system. For example, the following code
should print the installed OligoArrayAux version when executed from the
R console:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"hybrid-min -V"</span><span class="op">)</span></span></code></pre></div>
<pre><code>hybrid-min (OligoArrayAux) 3.8
By Nicholas R. Markham and Michael Zuker
Copyright (C) 2006
Rensselaer Polytechnic Institute
Troy, NY 12810-3590 USA</code></pre>
</div>
<div class="section level3">
<h3 id="startup">Startup<a class="anchor" aria-label="anchor" href="#startup"></a>
</h3>
<p>{#sec:startup}</p>
<p>To get started we need to load the <strong>DECIPHER</strong> package,
which automatically loads several other required packages.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DECIPHER</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: Biostrings</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     findMatches</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: XVector</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biostrings'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     strsplit</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: RSQLite</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: parallel</span></span></code></pre>
<p>Help for the <code>DesignProbes</code> function can be accessed
through:</p>
<pre><code>&gt; ? DesignProbes</code></pre>
<p>If <strong>DECIPHER</strong> is installed on your system, the code in
each example can be obtained via:</p>
<pre><code>&gt; browseVignettes("DECIPHER")</code></pre>
</div>
<div class="section level3">
<h3 id="creating-a-sequence-database">Creating a Sequence Database<a class="anchor" aria-label="anchor" href="#creating-a-sequence-database"></a>
</h3>
<p>We begin with a set of aligned 16S sequences representing a variety
of bacteria collected from drinking water samples. We wish to design
probes targeting only the genus <em>Sphingopyxis</em>, which is a common
organism detected in drinking water distribution systems. This example
uses a GenBank sequence file included as part of the
<strong>DECIPHER</strong> package, but you could follow along with your
own GenBank, FASTA, or FASTQ file of aligned sequences. Be sure to
change the path names to those on your system by replacing all of the
text inside quotes labeled “<span class="math inline">\(&lt;&lt;\)</span>path to …<span class="math inline">\(&gt;&gt;\)</span>” with the actual path on your
system.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify the path to your sequence file:</span></span>
<span><span class="va">gb</span> <span class="op">&lt;-</span> <span class="st">"&lt;&lt;path to GenBank file&gt;&gt;"</span></span>
<span><span class="co"># OR find the example sequence file used in this tutorial:</span></span>
<span><span class="va">gb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bacteria_175seqs.gen"</span>, package<span class="op">=</span><span class="st">"DECIPHER"</span><span class="op">)</span></span></code></pre></div>
<p>Next, there are two options for importing the sequences into a
database: either save a database file or maintain the database in
memory. Here we will build the database in memory because it is a small
set of sequences and we do not intend to use the database later:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify a path for where to write the sequence database</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="st">"&lt;&lt;path to write sequence database&gt;&gt;"</span></span>
<span><span class="co"># OR create the sequence database in memory</span></span>
<span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">":memory:"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">gb</span>, <span class="st">"GenBank"</span>, <span class="va">dbConn</span>, <span class="st">"Bacteria"</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
Reading GenBank file chunk 1
## 
## 175 total sequences in table Seqs.
## Time difference of 0.08 secs</code></pre>
</div>
<div class="section level3">
<h3 id="defining-groups">Defining Groups<a class="anchor" aria-label="anchor" href="#defining-groups"></a>
</h3>
<p>At this point we need to define groups of related sequences in the
database we just created. In this case we wish to define phylogenetic
groups based on the taxonomic information included in the original
GenBank file. The advantage of using a GenBank file is that the
taxonomic information is automatically imported from the ORGANISM field
into a “rank” column in the sequence database. When importing from a
FASTA or FASTQ file the taxonomic information must come from another
source, such as parsing the sequence identifiers or by building a tree
with the function <code>TreeLine</code>. In the case of a GenBank file,
the “rank” column can be used by the function
<code>IdentifyByRank</code> to identify the groups at any taxonomic
level (e.g., genus).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IdentifyByRank.html">IdentifyByRank</a></span><span class="op">(</span><span class="va">dbConn</span>, level<span class="op">=</span><span class="cn">Inf</span>, add2tbl<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Updating column: "identifier"...</span></span>
<span><span class="co">## Formed 72 distinct groups.</span></span>
<span><span class="co">## Added to table Seqs: "identifier".</span></span>
<span><span class="co">## Time difference of 0.01 secs</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="probe-design-steps">Probe Design Steps<a class="anchor" aria-label="anchor" href="#probe-design-steps"></a>
</h2>
<div class="section level3">
<h3 id="tiling-sequences">Tiling Sequences<a class="anchor" aria-label="anchor" href="#tiling-sequences"></a>
</h3>
<p>We continue by creating a set of k-mers, known as “tiles”, that
represent the sequences in each group. Here we must make several
decisions that will affect probe design in the future. The defaults are
to create tiles of length 26-27 nucleotides with up to 10 permutations
that represent at least 90% of the permutations present in each target
site. These parameters will generally fit most probe designs, but it is
recommended to read the help file for the <code>TileSeqs</code> function
to make sure that it is doing what is desired. We will save the
resulting tiles back to the database as a new table named “Tiles” in
case we wish to access them in the future.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TileSeqs.html">TileSeqs</a></span><span class="op">(</span><span class="va">dbConn</span>, add2tbl<span class="op">=</span><span class="st">"Tiles"</span><span class="op">)</span></span></code></pre></div>
<pre><code>  |===========================================================================| 100%

Time difference of 572.06 secs</code></pre>
</div>
<div class="section level3">
<h3 id="designing-all-possible-probes">Designing All Possible Probes<a class="anchor" aria-label="anchor" href="#designing-all-possible-probes"></a>
</h3>
<p>Next we wish to design probes targeting the 5 sequences identified as
belonging to the genus <em>Sphingopyxis</em>. As an example, we will
begin by designing probes for every possible target site that meet
certain design criteria. By default, we allow up to 4 probe permutations
as long as they cover 90% of the permutations present in their target
site. Note that the input parameters to <code>DesignProbes</code> should
be carefully considered in order adequately represent the experimental
conditions that will be used. For example, here we limit the target
sites to those between alignment positions 120 and 1,450 because this is
the region shared by most of the sequences.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">probes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignProbes.html">DesignProbes</a></span><span class="op">(</span><span class="va">tiles</span>, identifier<span class="op">=</span><span class="st">"Sphingopyxis"</span>,</span>
<span>    start<span class="op">=</span><span class="fl">120</span>, end<span class="op">=</span><span class="fl">1450</span><span class="op">)</span></span></code></pre></div>
<pre><code>Sphingopyxis (678):
  |===========================================================================| 100%

Time difference of 250 secs</code></pre>
<p>We can now examine the top-scoring probe with maximal sensitivity and
specificity to the target group.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">probes</span><span class="op">$</span><span class="va">score</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">probes</span><span class="op">[</span><span class="va">o</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="op">]</span></span></code></pre></div>
<pre><code>      identifier start start_aligned permutations
826 Sphingopyxis   859          1019            3
           score                 probe.1
826 -1.43876.... CCGCGATTAGGATGTCAAACGCT
                    probe.2                 probe.3 probe.4
826 CCGCGATCAGGATGTCAAATGCT CCGCGATCAGGATGTCAAACGCT    &lt;NA&gt;
    efficiency.1 efficiency.2 efficiency.3 efficiency.4
826    0.5014489    0.5506237    0.6315190           NA
       FAm.1    FAm.2    FAm.3    FAm.4 coverage.1
826 35.04488 36.57339 39.17166       NA        0.4
    coverage.2 coverage.3 coverage.4
826        0.2        0.4         NA
mismatches
826 unclassified_Sphingomonadaceae (63.2
4.1
unclassified_Sphingomonadales (9.1
-17.8</code></pre>
<p>The best probe out of 678 candidate target sites has three
permutations, with formamide melt points (FAm) between 35% and 39% [FA]
(v/v). These three permutations can be represented with a single
consensus probe:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ConsensusSequence.html">ConsensusSequence</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="va">probes</span><span class="op">[</span><span class="va">o</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"probe"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>  A DNAStringSet instance of length 1
    width seq
[1]    23 CCGCGATYAGGATGTCAAAYGCT</code></pre>
<p>Based on the column <em>mismatches</em>, this probe has two predicted
cross-hybridizations with other groups in the sequence set. These
groups, unclassified Sphingomonadaceae and unclassified
Sphingomonadales, are predicted to have 63.2% and 9.1% hybridization
efficiency at equilibrium in the hybridization buffer with 35% [FA]
(v/v). The first non-target has exactly the same Gibb’s free energy
because it is a perfect match to the probe. Note that its formamide melt
point (FAm) is 4.1% greater than the probe permutation with the lowest
melt point, which is the limiting factor in the experiment. The second
probe has 6.2 kcal/mol greater Gibb’s free energy and a formamide melt
point 17.8% [FA] less than the probe with the target. We can examine the
predicted melt curve of the probes with the target and non-target:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>FA_range <span class="ot">&lt;-</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">70</span> <span class="co"># [FA] (</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>probe <span class="ot">&lt;-</span> probes<span class="sc">$</span>probe[o[<span class="dv">1</span>], <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>targets <span class="ot">&lt;-</span> <span class="fu">reverseComplement</span>(<span class="fu">DNAStringSet</span>(probe))</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>f <span class="ot">&lt;-</span> <span class="cf">function</span>(FA)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">CalculateEfficiencyFISH</span>(probe, targets,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">temp=</span><span class="dv">46</span>, <span class="at">P=</span><span class="fl">250e-9</span>, <span class="at">ions=</span><span class="dv">1</span>, FA)[, <span class="st">"HybEff"</span>]</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>efficiency <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">unlist</span>(<span class="fu">lapply</span>(FA_range, f)), <span class="at">ncol=</span><span class="dv">3</span>, <span class="at">byrow=</span><span class="cn">TRUE</span>)</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(FA_range, efficiency, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="at">ylab=</span><span class="st">"Hybridization Efficiency"</span>,</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="fu">expression</span>(<span class="fu">paste</span>(<span class="st">"[Formamide] (</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a><span class="st">    type="</span>l<span class="st">", lwd=2, col="</span>Blue<span class="st">", main="</span>Formamide Curve<span class="st">", lty=1)</span></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a><span class="st">nontargets &lt;- DNAStringSet(c("</span>AGCGTTTGACATCCTGATCGCGG<span class="st">",</span></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a><span class="st">    "</span>AGCTTTTGACATCCCGGTCGCGG<span class="st">"))</span></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a><span class="st">f &lt;- function(FA)</span></span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a><span class="st">    CalculateEfficiencyFISH(probe[3:2], nontargets,</span></span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a><span class="st">        temp=46, P=250e-9, ions=1, FA)[, "</span>HybEff<span class="st">"]</span></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a><span class="st">efficiency &lt;- matrix(unlist(lapply(FA_range, f)), ncol=2, byrow=TRUE)</span></span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a><span class="st">matlines(FA_range, efficiency, col="</span>Red<span class="st">", lwd=2, lty=3)</span></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a><span class="st">abline(h=0.5, lty=2, lwd=2, col="</span>Orange<span class="st">")</span></span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a><span class="st">abline(v=35, lty=2, lwd=2, col="</span>Green<span class="st">")</span></span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a><span class="st">legend("</span>topright<span class="st">", legend=c("</span>Targets<span class="st">", "</span>Non<span class="sc">-</span>Targets<span class="st">", "</span><span class="dv">50</span></span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Experimental [FA]"</span>), <span class="at">col=</span><span class="fu">c</span>(<span class="st">"Blue"</span>, <span class="st">"Red"</span>, <span class="st">"Orange"</span>, <span class="st">"Green"</span>),</span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">lwd=</span><span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>), <span class="at">lty=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">2</span>))</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="finding-non-targets-in-a-comprehensive-reference-database">Finding Non-targets in a Comprehensive Reference Database<a class="anchor" aria-label="anchor" href="#finding-non-targets-in-a-comprehensive-reference-database"></a>
</h3>
<p>It is now important to find potential non-targets not included in the
sequence set used for design. For this purpose we can either build a new
comprehensive reference database, or download one online from <a href="http://DECIPHER.cee.wisc.edu/Download.html" class="external-link uri">http://DECIPHER.cee.wisc.edu/Download.html</a>. The
reference database must be unzipped after download. Each database
includes a pre-processed set of tiles, generated in the same manner as
shown above, that can be found in the table named “Tiles”. Be sure to
change the path names to those on your system by replacing all of the
text inside quotes labeled “<span class="math inline">\(&lt;&lt;\)</span>path to …<span class="math inline">\(&gt;&gt;\)</span>” with the actual path on your
system.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dbConn_ref</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"&lt;&lt;path to reference database&gt;&gt;"</span><span class="op">)</span></span>
<span><span class="co"># select the most common k-mers</span></span>
<span><span class="va">ref_tiles</span> <span class="op">&lt;-</span> <span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">dbConn_ref</span>,</span>
<span>    <span class="st">"select * from Tiles where groupCoverage &gt; 0.2 and coverage &gt; 0.01"</span><span class="op">)</span></span>
<span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">dbConn_ref</span><span class="op">)</span></span>
<span><span class="va">ref_tiles</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"ref"</span>, <span class="va">ref_tiles</span><span class="op">$</span><span class="va">id</span>, sep<span class="op">=</span><span class="st">"_"</span><span class="op">)</span></span></code></pre></div>
<p>Next we can use these tiles to find hits in the comprehensive
database of non-targets that may not have been included in the sequence
set used for design. This step is unnecessary if a comprehensive
sequence set was used during the design steps above. Here we used a
reduced set of non-targets during design, so we must search the
comprehensive reference database for other non-targets.</p>
<p>This step requires searching for a substantial number of probes in a
very large number of non-targets. To accomplish this feat in a
reasonable time we will leverage <strong>Biostrings</strong>’ dictionary
matching functionality. Nevertheless this step still requires several
minutes to complete. Note that this process can be accelerated by
eliminating the lowest scoring probes (i.e.,
<code>probes &lt;- probes[o[1:20],]</code>), but this should only be
used in cases where the number of non-targets included in the design
sequence set was sufficient to narrow down the list of potential probes
to reasonable candidates.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="va">ref_tiles</span><span class="op">$</span><span class="va">target_site</span><span class="op">)</span></span>
<span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">probes</span><span class="op">$</span><span class="va">probe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">probes_rc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/reverseComplement.html" class="external-link">reverseComplement</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">DNAStringSet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">probes</span><span class="op">$</span><span class="va">probe</span><span class="op">)</span><span class="op">[</span><span class="va">w</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/PDict-class.html" class="external-link">PDict</a></span><span class="op">(</span><span class="va">probes_rc</span>, tb.start<span class="op">=</span><span class="fl">1</span>, tb.width<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">hits1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/matchPDict-exact.html" class="external-link">vwhichPDict</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">seqs</span>, max.mismatch<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">hits1</span>, <span class="va">length</span>, <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">integer</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hits1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">hits1</span>, use.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">hits1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span>, <span class="va">l</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/PDict-class.html" class="external-link">PDict</a></span><span class="op">(</span><span class="va">probes_rc</span>, tb.end<span class="op">=</span><span class="op">-</span><span class="fl">1</span>, tb.width<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">hits2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/matchPDict-exact.html" class="external-link">vwhichPDict</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">seqs</span>, max.mismatch<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">hits2</span>, <span class="va">length</span>, <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">integer</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hits2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">hits2</span>, use.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">hits2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span>, <span class="va">l</span><span class="op">)</span></span>
<span><span class="va">hits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">hits1</span>, <span class="va">hits2</span><span class="op">)</span></span></code></pre></div>
<p>Now we can use the thermodynamic model to predict the degree of
cross-hybridization between each probe and non-target. This information
is used to update the score for every probe, and is stored in the column
<em>mismatches</em> alongside previous non-targets identified during
design. Refer to section~@ref(sec:startup) above for how to easily
obtain the code shown below:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>Hyb_FA <span class="ot">&lt;-</span> <span class="dv">35</span> <span class="co"># the default Hybridization [FA] (</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>count <span class="ot">&lt;-</span> 0L</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>pBar <span class="ot">&lt;-</span> <span class="fu">txtProgressBar</span>(<span class="at">style=</span><span class="dv">3</span>)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(probes)[<span class="dv">1</span>]) {</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># for each hit calculate the degree of cross-hybridization</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>    w <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(probes[i, <span class="st">"probe"</span>]))</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>    results <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(w)) {</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>        count <span class="ot">&lt;-</span> count <span class="sc">+</span> 1L</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>        w <span class="ot">&lt;-</span> <span class="fu">which</span>(hits<span class="sc">==</span>count)</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">length</span>(w) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>            ns <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(<span class="fu">unique</span>(<span class="fu">names</span>(hits[w])))</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>            eff <span class="ot">&lt;-</span> <span class="fu">CalculateEfficiencyFISH</span>(<span class="fu">rep</span>(probes<span class="sc">$</span>probe[i, j],</span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>                    <span class="fu">length</span>(ns)),</span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>                ref_tiles<span class="sc">$</span>target_site[ns],</span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>                <span class="dv">46</span>, <span class="co"># temperature</span></span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a>                <span class="fl">250e-9</span>, <span class="co"># [Probe]</span></span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a>                <span class="dv">1</span>, <span class="co"># [NA]</span></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a>                Hyb_FA)</span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a>            eff <span class="ot">&lt;-</span> <span class="fu">cbind</span>(eff,</span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a>                <span class="fu">data.frame</span>(<span class="at">id=</span>ref_tiles<span class="sc">$</span>id[ns],</span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a>                    <span class="at">probe_rc=</span><span class="fu">toString</span>(probes_rc[count]),</span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a>                    <span class="at">target=</span>ref_tiles<span class="sc">$</span>target_site[ns],</span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a>                    <span class="at">dFAm=</span>eff[, <span class="st">"FAm"</span>] <span class="sc">-</span> Hyb_FA,</span>
<span id="cb39-25"><a href="#cb39-25" aria-hidden="true" tabindex="-1"></a>                    <span class="at">stringsAsFactors=</span><span class="cn">FALSE</span>))</span>
<span id="cb39-26"><a href="#cb39-26" aria-hidden="true" tabindex="-1"></a>            results <span class="ot">&lt;-</span> <span class="fu">rbind</span>(results, eff)</span>
<span id="cb39-27"><a href="#cb39-27" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb39-28"><a href="#cb39-28" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb39-29"><a href="#cb39-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb39-30"><a href="#cb39-30" aria-hidden="true" tabindex="-1"></a>    w <span class="ot">&lt;-</span> <span class="fu">which</span>(results<span class="sc">$</span>dFAm <span class="sc">&gt;</span> <span class="sc">-</span><span class="dv">20</span>)</span>
<span id="cb39-31"><a href="#cb39-31" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(w) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb39-32"><a href="#cb39-32" aria-hidden="true" tabindex="-1"></a>        <span class="co"># only record the strongest cross-hybridization in each group</span></span>
<span id="cb39-33"><a href="#cb39-33" aria-hidden="true" tabindex="-1"></a>        results <span class="ot">&lt;-</span> results[w,]</span>
<span id="cb39-34"><a href="#cb39-34" aria-hidden="true" tabindex="-1"></a>        u <span class="ot">&lt;-</span> <span class="fu">unique</span>(results<span class="sc">$</span>id)</span>
<span id="cb39-35"><a href="#cb39-35" aria-hidden="true" tabindex="-1"></a>        keep <span class="ot">&lt;-</span> <span class="fu">integer</span>()</span>
<span id="cb39-36"><a href="#cb39-36" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(u)) {</span>
<span id="cb39-37"><a href="#cb39-37" aria-hidden="true" tabindex="-1"></a>            w <span class="ot">&lt;-</span> <span class="fu">which</span>(results<span class="sc">$</span>id<span class="sc">==</span>u[j])</span>
<span id="cb39-38"><a href="#cb39-38" aria-hidden="true" tabindex="-1"></a>            keep <span class="ot">&lt;-</span> <span class="fu">c</span>(keep, w[<span class="fu">which.max</span>(results<span class="sc">$</span>dFAm[w])])</span>
<span id="cb39-39"><a href="#cb39-39" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb39-40"><a href="#cb39-40" aria-hidden="true" tabindex="-1"></a>        results <span class="ot">&lt;-</span> results[keep,]</span>
<span id="cb39-41"><a href="#cb39-41" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb39-42"><a href="#cb39-42" aria-hidden="true" tabindex="-1"></a>        <span class="co"># append more non-targets to mismatches</span></span>
<span id="cb39-43"><a href="#cb39-43" aria-hidden="true" tabindex="-1"></a>        p <span class="ot">&lt;-</span> <span class="fu">pairwiseAlignment</span>(results<span class="sc">$</span>probe_rc,</span>
<span id="cb39-44"><a href="#cb39-44" aria-hidden="true" tabindex="-1"></a>            results<span class="sc">$</span>target,</span>
<span id="cb39-45"><a href="#cb39-45" aria-hidden="true" tabindex="-1"></a>            <span class="at">type=</span><span class="st">"global-local"</span>)</span>
<span id="cb39-46"><a href="#cb39-46" aria-hidden="true" tabindex="-1"></a>        probes<span class="sc">$</span>mismatches[i] <span class="ot">&lt;-</span> <span class="fu">paste</span>(probes<span class="sc">$</span>mismatches[i],</span>
<span id="cb39-47"><a href="#cb39-47" aria-hidden="true" tabindex="-1"></a>            <span class="fu">paste</span>(results<span class="sc">$</span>id,</span>
<span id="cb39-48"><a href="#cb39-48" aria-hidden="true" tabindex="-1"></a>                <span class="st">" ("</span>, <span class="fu">round</span>(<span class="dv">100</span><span class="sc">*</span>results<span class="sc">$</span>HybEff, <span class="dv">1</span>), <span class="st">"</span></span>
<span id="cb39-49"><a href="#cb39-49" aria-hidden="true" tabindex="-1"></a><span class="st">                round(results$ddG1, 2), "</span>kcal<span class="sc">/</span>mol,<span class="st">",</span></span>
<span id="cb39-50"><a href="#cb39-50" aria-hidden="true" tabindex="-1"></a><span class="st">                round(results$dFAm, 1), "</span></span>
<span id="cb39-51"><a href="#cb39-51" aria-hidden="true" tabindex="-1"></a>                <span class="fu">substring</span>(<span class="fu">reverseComplement</span>(<span class="fu">DNAStringSet</span>(<span class="fu">pattern</span>(p))),</span>
<span id="cb39-52"><a href="#cb39-52" aria-hidden="true" tabindex="-1"></a>                    1L),</span>
<span id="cb39-53"><a href="#cb39-53" aria-hidden="true" tabindex="-1"></a>                <span class="st">"/"</span>, <span class="fu">substring</span>(<span class="fu">subject</span>(p), 1L), <span class="st">")"</span>,</span>
<span id="cb39-54"><a href="#cb39-54" aria-hidden="true" tabindex="-1"></a>                <span class="at">sep=</span><span class="st">""</span>,</span>
<span id="cb39-55"><a href="#cb39-55" aria-hidden="true" tabindex="-1"></a>                <span class="at">collapse=</span><span class="st">" "</span>),</span>
<span id="cb39-56"><a href="#cb39-56" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep=</span><span class="st">""</span>)</span>
<span id="cb39-57"><a href="#cb39-57" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb39-58"><a href="#cb39-58" aria-hidden="true" tabindex="-1"></a>        <span class="co"># score -= 0.2 + 1.2^dFAm</span></span>
<span id="cb39-59"><a href="#cb39-59" aria-hidden="true" tabindex="-1"></a>        probes<span class="sc">$</span>score[i] <span class="ot">&lt;-</span> probes<span class="sc">$</span>score[i] <span class="sc">-</span></span>
<span id="cb39-60"><a href="#cb39-60" aria-hidden="true" tabindex="-1"></a>            <span class="fu">sum</span>(<span class="fu">ifelse</span>(results<span class="sc">$</span>dFAm <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">20</span>, <span class="dv">0</span>, <span class="fl">0.2</span> <span class="sc">+</span></span>
<span id="cb39-61"><a href="#cb39-61" aria-hidden="true" tabindex="-1"></a>                <span class="fl">1.2</span><span class="sc">^</span><span class="fu">ifelse</span>(results<span class="sc">$</span>dFAm <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="dv">0</span>, results<span class="sc">$</span>dFAm)))</span>
<span id="cb39-62"><a href="#cb39-62" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb39-63"><a href="#cb39-63" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb39-64"><a href="#cb39-64" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setTxtProgressBar</span>(pBar, i<span class="sc">/</span><span class="fu">dim</span>(probes)[<span class="dv">1</span>])</span>
<span id="cb39-65"><a href="#cb39-65" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>  |===========================================================================| 100%</code></pre>
<p>Finally, we can examine the top scoring single probe after updating
the results using the comprehensive database. The new best scoring probe
targets a site over 300 nucleotides away from the original best scoring
probe. The design set was missing many non-targets that were found by
searching the comprehensive database. Note that the target group,
<em>Sphingopyxis</em>, was present in the comprehensive database and now
appears in the set of non-targets (<code>mismatches</code>). This has no
net impact on the scores however, as all of the candidate probes are
effected equally since they are all perfect match to sequences in this
group.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the original best scoring probe after searching the comprehensive database</span></span>
<span><span class="va">probes</span><span class="op">[</span><span class="va">o</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="op">]</span></span></code></pre></div>
<pre><code>      identifier start start_aligned permutations        score
826 Sphingopyxis   859          1019            3 -7.75865....
                    probe.1                 probe.2
826 CCGCGATTAGGATGTCAAACGCT CCGCGATCAGGATGTCAAATGCT
                    probe.3 probe.4 efficiency.1 efficiency.2
826 CCGCGATCAGGATGTCAAACGCT    &lt;NA&gt;    0.5014489    0.5506237
    efficiency.3 efficiency.4    FAm.1    FAm.2    FAm.3    FAm.4
826    0.6315190           NA 35.04488 36.57339 39.17166       NA
    coverage.1 coverage.2 coverage.3 coverage.4
826        0.4        0.2        0.4         NA
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         mismatches
826 unclassified_Sphingomonadaceae (63.2
AGCGTTTGACATCCTGATCGCGG) unclassified_Sphingomonadales (9.1
mol,-17.8
(13.2
ref_Sphingopyxis (63.2
AGCGTTTGACATCCTGATCGCGG) ref_Zymomonas (63.2
4.2
6.25kcal/mol,-19.5
ref_Cucumibacter (25.9
AGCTCTTGACATCCTGATCGCGG) ref_Celeribacter (15
mol,-13.5
5.64kcal/mol,-13.5
ref_Maritimibacter (15
AACCCTTGACATCCTGATCGCGG) ref_Pseudoruegeria (15
mol,-13.5
5.64kcal/mol,-13.5
(15
ref_Altererythrobacter (24.4
AGCCTTTGACATCCTGGTCGCGG) ref_Croceicoccus (24.4
mol,-8.8
6.32kcal/mol,-18.2
(8.7</code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the new best scoring probe after searching the comprehensive database</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">probes</span><span class="op">$</span><span class="va">score</span>, <span class="op">-</span><span class="fl">1</span><span class="op">*</span><span class="va">probes</span><span class="op">$</span><span class="va">permutations</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">probes</span><span class="op">$</span><span class="va">coverage</span><span class="op">[</span>,<span class="op">]</span><span class="op">)</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">probes</span><span class="op">[</span><span class="va">o</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="op">]</span></span></code></pre></div>
<pre><code>      identifier start start_aligned permutations        score
489 Sphingopyxis   522           679            2 -5.93961....
                      probe.1                   probe.2 probe.3 probe.4
489 CACTGACCTCTCCAAGATTCAAGTT CACTGACCTCTCCAAGATTCTAGTT    &lt;NA&gt;    &lt;NA&gt;
    efficiency.1 efficiency.2 efficiency.3 efficiency.4    FAm.1
489    0.5700456    0.6445774           NA           NA 37.04415
       FAm.2    FAm.3    FAm.4 coverage.1 coverage.2 coverage.3
489 39.31453       NA       NA        0.8        0.2         NA
    coverage.4
489         NA
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   mismatches
489 Porphyrobacter (12.3
AGCTAGAATCTTGGAGAGGCGAGTG) Blastomonas (9.2
mol,-18.6
(54.1
unclassified_Sphingomonadaceae (12.3
CGCTAGAATCTTGGAGAGGCGAGTG) Spirochaeta (12.6
mol,-16.1
(57
ref_Croceicoccus (65.8
GACTAGAATCTTGGAGAGGTCAGTG) ref_Novosphingobium (65.8
4.7
5.91kcal/mol,-13</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="dual-probe-design-steps">Dual Probe Design Steps<a class="anchor" aria-label="anchor" href="#dual-probe-design-steps"></a>
</h2>
<div class="section level3">
<h3 id="designing-the-optimal-dual-probe-set">Designing the Optimal Dual Probe Set<a class="anchor" aria-label="anchor" href="#designing-the-optimal-dual-probe-set"></a>
</h3>
<p>Since no single probe was sufficient to completely distinguish all
non-targets, we can further increase specificity by using two probes in
combination. There are two methods to accomplish the design of dual
probes:</p>
<ol style="list-style-type: decimal">
<li>If using a comprehensive sequence set during design, then we could
have simply set the parameter <code>numProbeSets</code> to a positive
number from its default of zero. In doing so the
<code>DesignProbes</code> function will search through potential
combinations of two probes. The value of <code>numProbeSets</code> will
determine how many dual probe sets <code>DesignProbes</code> will
design. For example:</li>
</ol>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">probes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignProbes.html">DesignProbes</a></span><span class="op">(</span><span class="va">tiles</span>, identifier<span class="op">=</span><span class="st">"Sphingopyxis"</span>,</span>
<span>    start<span class="op">=</span><span class="fl">120</span>, end<span class="op">=</span><span class="fl">1450</span>,</span>
<span>    numProbeSets<span class="op">=</span><span class="fl">100</span><span class="op">)</span> <span class="co"># note numProbeSets &gt; 0</span></span></code></pre></div>
<pre><code>Sphingopyxis (678):
  |=======================================================================| 100%

Time difference of 270 secs</code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">probes</span><span class="op">)</span> <span class="co"># now there are 100 potential probe sets</span></span></code></pre></div>
<pre><code>[1] 100 21</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>If searching through non-targets using a comprehensive reference
database (as outlined in the sections above) then it is simpler to
continue from where we left off above, because we already finished
designing all single probes and searching for potential non-targets. In
this case, all we need to do is search for the combination of two probes
with the minimal mismatch overlap, as described below.</li>
</ol>
<p>To continue, we first compute the score for using each pair of probes
as a dual probe set. The new score for using two probes is based off the
minimal cross-hybridization between each pair of probes and every
non-target. For example, if probe #1 cross-hybridizes with a non-target,
but probe #2 does not, then the non-target is not a threat when the two
probes are used in conjunction because the hybridization of both probes
is required for a positive identification.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>numProbeSets <span class="ot">&lt;-</span> <span class="dv">100</span> <span class="co"># number of probe sets to generate</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>s <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">dim</span>(probes)[<span class="dv">1</span>] <span class="sc">&gt;</span> numProbeSets, numProbeSets, <span class="fu">dim</span>(probes)[<span class="dv">1</span>])</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>MMs <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(probes<span class="sc">$</span>mismatches[o[<span class="dv">1</span><span class="sc">:</span>s]], <span class="st">"mol,"</span>, <span class="at">fixed=</span><span class="cn">TRUE</span>)</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>ls <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(MMs, length))</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>ls <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(ls <span class="sc">&gt;</span> <span class="dv">0</span>, ls <span class="sc">-</span> <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>index <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(ls), ls)</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(index) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>    MMs <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(<span class="fu">unlist</span>(MMs), <span class="st">"</span></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a><span class="st">    MMs &lt;- unlist(strsplit(unlist(MMs), "</span>) <span class="st">", fixed=TRUE))</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a><span class="st">    effs &lt;- as.numeric(MMs[seq(2, length(MMs), 3)])</span></span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a><span class="st">    MMs &lt;- MMs[seq(1, length(MMs), 3)]</span></span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a><span class="st">    MMs &lt;- unlist(strsplit(MMs, "</span> (<span class="st">", fixed=TRUE))</span></span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a><span class="st">    MMs &lt;- MMs[seq(1, length(MMs), 2)]</span></span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a><span class="st">m &lt;- matrix(0, nrow=s, ncol=s, dimnames=list(o[1:s], o[1:s])) # scores</span></span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a><span class="st">n &lt;- matrix(0, nrow=s, ncol=s, dimnames=list(o[1:s], o[1:s])) # counts</span></span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a><span class="st">for (i in 1:(s - 1)) {</span></span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a><span class="st">    w1 &lt;- which(index==i)</span></span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a><span class="st">    if (length(w1)==0)</span></span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a><span class="st">        next</span></span>
<span id="cb49-22"><a href="#cb49-22" aria-hidden="true" tabindex="-1"></a><span class="st">    for (j in (i + 1):s) {</span></span>
<span id="cb49-23"><a href="#cb49-23" aria-hidden="true" tabindex="-1"></a><span class="st">        w2 &lt;- which(index==j)</span></span>
<span id="cb49-24"><a href="#cb49-24" aria-hidden="true" tabindex="-1"></a><span class="st">        if (length(w2)==0)</span></span>
<span id="cb49-25"><a href="#cb49-25" aria-hidden="true" tabindex="-1"></a><span class="st">            next</span></span>
<span id="cb49-26"><a href="#cb49-26" aria-hidden="true" tabindex="-1"></a><span class="st">        overlap_MMs &lt;- match(MMs[w1], MMs[w2])</span></span>
<span id="cb49-27"><a href="#cb49-27" aria-hidden="true" tabindex="-1"></a><span class="st">        w &lt;- which(!is.na(overlap_MMs))</span></span>
<span id="cb49-28"><a href="#cb49-28" aria-hidden="true" tabindex="-1"></a><span class="st">        n[i, j] &lt;- length(w)</span></span>
<span id="cb49-29"><a href="#cb49-29" aria-hidden="true" tabindex="-1"></a><span class="st">        if (length(w) &gt; 0)</span></span>
<span id="cb49-30"><a href="#cb49-30" aria-hidden="true" tabindex="-1"></a><span class="st">            m[i, j] &lt;- sum(tapply(c(effs[w1[w]], effs[w2[overlap_MMs[w]]]),</span></span>
<span id="cb49-31"><a href="#cb49-31" aria-hidden="true" tabindex="-1"></a><span class="st">                rep(1:length(w), 2),</span></span>
<span id="cb49-32"><a href="#cb49-32" aria-hidden="true" tabindex="-1"></a><span class="st">                function(x) return(-1.2^ifelse(min(x) &gt; 0, 0, min(x)))))</span></span>
<span id="cb49-33"><a href="#cb49-33" aria-hidden="true" tabindex="-1"></a><span class="st">    }</span></span>
<span id="cb49-34"><a href="#cb49-34" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span></code></pre></div>
<p>As expected, the score for using two probes in combination is
significantly better than using any single probe alone. This can be seen
in Figure 2, where the off-diagonal colors represent the use of two
probes, and the diagonal elements represent the use of a single probe.
Some combinations of two probes results in poor scores (seen in
red/magenta), while others give high scores (blue/green).</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the matrix of dual-probe scores</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="va">m</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span> <span class="co"># make symmetric</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">probes</span><span class="op">$</span><span class="va">score</span><span class="op">[</span><span class="va">o</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">s</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">O</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">o</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">s</span><span class="op">]</span><span class="op">)</span> <span class="co"># unsort</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRamp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">rainbow</a></span><span class="op">(</span><span class="fl">101</span>, start<span class="op">=</span><span class="fl">0</span>, end<span class="op">=</span><span class="fl">0.35</span>, v<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span>, bias<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">100</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span><span class="op">/</span><span class="fl">255</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">rgb</span>, <span class="va">cols</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">cols</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, <span class="va">cols</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span><span class="va">O</span>, <span class="va">O</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">:</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, yaxt<span class="op">=</span><span class="st">'n'</span>, xaxt<span class="op">=</span><span class="st">'n'</span>, col<span class="op">=</span><span class="va">cols</span>,</span>
<span>    xlab<span class="op">=</span><span class="st">"Probe #1 Target Site Position"</span>, ylab<span class="op">=</span><span class="st">"Probe #2 Target Site Position"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.01</span><span class="op">)</span>, <span class="va">probes</span><span class="op">$</span><span class="va">start</span><span class="op">[</span><span class="va">o</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">s</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">O</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">s</span>, length.out<span class="op">=</span><span class="fl">101</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fl">0.01</span><span class="op">)</span>, <span class="va">probes</span><span class="op">$</span><span class="va">start</span><span class="op">[</span><span class="va">o</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">s</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">O</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">s</span>, length.out<span class="op">=</span><span class="fl">101</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>There are many combinations of two probes that have fairly similar
scores, so now the challenge becomes how to choose the best two probes.
One criteria is that the first and second probe’s target sites must be
separated by at least 50 nucleotides to mitigate helper-oligonucleotide
effects. Other criteria include the number of probe permutations
required, percent coverage of the group, and individual probe’s scores.
Next we will take these considerations into account when choosing the
top dual-probe sets.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose the best probe sets</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dimnames</span>(m)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">outer</span>(probes<span class="sc">$</span>permutations[o[<span class="dv">1</span><span class="sc">:</span>s]],</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>    probes<span class="sc">$</span>permutations[o[<span class="dv">1</span><span class="sc">:</span>s]],</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">FUN=</span><span class="st">"+"</span>)</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>c <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">1</span><span class="sc">*</span><span class="fu">outer</span>(<span class="fu">rowSums</span>(<span class="fu">as.matrix</span>(probes<span class="sc">$</span>coverage[o[<span class="dv">1</span><span class="sc">:</span>s],]), <span class="at">na.rm=</span><span class="cn">TRUE</span>),</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ifelse</span>(<span class="fu">rep</span>(s, s)<span class="sc">==</span><span class="dv">1</span>,</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">sum</span>(probes<span class="sc">$</span>coverage[o[<span class="dv">1</span><span class="sc">:</span>s],], <span class="at">na.rm=</span><span class="cn">TRUE</span>),</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">rowSums</span>(<span class="fu">as.matrix</span>(probes<span class="sc">$</span>coverage[o[<span class="dv">1</span><span class="sc">:</span>s],]), <span class="at">na.rm=</span><span class="cn">TRUE</span>)),</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">FUN=</span><span class="st">"*"</span>)</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>ss <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">1</span><span class="sc">*</span><span class="fu">outer</span>(probes<span class="sc">$</span>score[o[<span class="dv">1</span><span class="sc">:</span>s]],</span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>    probes<span class="sc">$</span>score[o[<span class="dv">1</span><span class="sc">:</span>s]],</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">FUN=</span><span class="st">"+"</span>)</span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a><span class="co"># order probes by dual-probe score, permutations, coverage, and individual scores</span></span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>o <span class="ot">&lt;-</span> <span class="fu">order</span>(m <span class="sc">+</span> n<span class="sc">/</span><span class="dv">5</span>, p, c, ss) <span class="co"># score = 0.2*n + 1.2^dFAm</span></span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a>j <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>first <span class="ot">&lt;-</span> second <span class="ot">&lt;-</span> <span class="fu">integer</span>()</span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>scores <span class="ot">&lt;-</span> <span class="fu">numeric</span>()</span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a>mismatches <span class="ot">&lt;-</span> <span class="fu">character</span>()</span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>numProbeSets) {</span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> ((j <span class="sc">+</span> <span class="dv">1</span>) <span class="sc">&gt;</span> <span class="fu">length</span>(o))</span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a>        <span class="cf">break</span></span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> (j <span class="sc">+</span> <span class="dv">1</span>)<span class="sc">:</span><span class="fu">length</span>(o)) {</span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a>        w_o <span class="ot">&lt;-</span> <span class="fu">c</span>((o[j] <span class="sc">-</span> <span class="dv">1</span>) </span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a>            (o[j] <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb51-26"><a href="#cb51-26" aria-hidden="true" tabindex="-1"></a>        f <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(d[[<span class="dv">1</span>]][w_o[<span class="dv">1</span>]])</span>
<span id="cb51-27"><a href="#cb51-27" aria-hidden="true" tabindex="-1"></a>        r <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(d[[<span class="dv">2</span>]][w_o[<span class="dv">2</span>]])</span>
<span id="cb51-28"><a href="#cb51-28" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb51-29"><a href="#cb51-29" aria-hidden="true" tabindex="-1"></a>        start_F <span class="ot">&lt;-</span> probes<span class="sc">$</span>start[f]</span>
<span id="cb51-30"><a href="#cb51-30" aria-hidden="true" tabindex="-1"></a>        start_R <span class="ot">&lt;-</span> probes<span class="sc">$</span>start[r]</span>
<span id="cb51-31"><a href="#cb51-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">abs</span>(start_F <span class="sc">-</span> start_R) <span class="sc">&gt;</span> (<span class="dv">50</span> <span class="sc">+</span> <span class="fu">nchar</span>(probes<span class="sc">$</span>probe[f][<span class="dv">1</span>])))</span>
<span id="cb51-32"><a href="#cb51-32" aria-hidden="true" tabindex="-1"></a>            <span class="cf">break</span> <span class="co"># &gt; 50 nt separation</span></span>
<span id="cb51-33"><a href="#cb51-33" aria-hidden="true" tabindex="-1"></a>    <span class="er">}</span></span>
<span id="cb51-34"><a href="#cb51-34" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb51-35"><a href="#cb51-35" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">abs</span>(start_F <span class="sc">-</span> start_R) <span class="sc">&gt;</span> (<span class="dv">50</span> <span class="sc">+</span> <span class="fu">nchar</span>(probes<span class="sc">$</span>probe[f][<span class="dv">1</span>]))) {</span>
<span id="cb51-36"><a href="#cb51-36" aria-hidden="true" tabindex="-1"></a>        first <span class="ot">&lt;-</span> <span class="fu">c</span>(first, f)</span>
<span id="cb51-37"><a href="#cb51-37" aria-hidden="true" tabindex="-1"></a>        second <span class="ot">&lt;-</span> <span class="fu">c</span>(second, r)</span>
<span id="cb51-38"><a href="#cb51-38" aria-hidden="true" tabindex="-1"></a>        scores <span class="ot">&lt;-</span> <span class="fu">c</span>(scores, <span class="sc">-</span>m[o[j]]<span class="sc">/</span><span class="dv">100</span>)</span>
<span id="cb51-39"><a href="#cb51-39" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb51-40"><a href="#cb51-40" aria-hidden="true" tabindex="-1"></a>        w_F <span class="ot">&lt;-</span> <span class="fu">which</span>(index<span class="sc">==</span>w_o[<span class="dv">1</span>])</span>
<span id="cb51-41"><a href="#cb51-41" aria-hidden="true" tabindex="-1"></a>        w_R <span class="ot">&lt;-</span> <span class="fu">which</span>(index<span class="sc">==</span>w_o[<span class="dv">2</span>])</span>
<span id="cb51-42"><a href="#cb51-42" aria-hidden="true" tabindex="-1"></a>        overlap_MMs <span class="ot">&lt;-</span> <span class="fu">match</span>(MMs[w_F], MMs[w_R])</span>
<span id="cb51-43"><a href="#cb51-43" aria-hidden="true" tabindex="-1"></a>        w_S <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(overlap_MMs))</span>
<span id="cb51-44"><a href="#cb51-44" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">length</span>(w_S)<span class="sc">==</span><span class="dv">0</span>)</span>
<span id="cb51-45"><a href="#cb51-45" aria-hidden="true" tabindex="-1"></a>            <span class="cf">next</span></span>
<span id="cb51-46"><a href="#cb51-46" aria-hidden="true" tabindex="-1"></a>        EFFs <span class="ot">&lt;-</span> <span class="fu">tapply</span>(<span class="fu">c</span>(effs[w_F[w_S]], effs[w_R[overlap_MMs[w_S]]]),</span>
<span id="cb51-47"><a href="#cb51-47" aria-hidden="true" tabindex="-1"></a>                <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(w_S), <span class="dv">2</span>),</span>
<span id="cb51-48"><a href="#cb51-48" aria-hidden="true" tabindex="-1"></a>                min)</span>
<span id="cb51-49"><a href="#cb51-49" aria-hidden="true" tabindex="-1"></a>        <span class="co"># record set mismatches</span></span>
<span id="cb51-50"><a href="#cb51-50" aria-hidden="true" tabindex="-1"></a>        w1 <span class="ot">&lt;-</span> <span class="fu">which</span>(EFFs <span class="sc">&gt;=</span> <span class="sc">-</span><span class="dv">20</span>)</span>
<span id="cb51-51"><a href="#cb51-51" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">length</span>(w1) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb51-52"><a href="#cb51-52" aria-hidden="true" tabindex="-1"></a>            mismatches <span class="ot">&lt;-</span> <span class="fu">c</span>(mismatches, <span class="fu">paste</span>(MMs[w_F][w_S][w1],</span>
<span id="cb51-53"><a href="#cb51-53" aria-hidden="true" tabindex="-1"></a>                <span class="st">" ("</span>, <span class="fu">formatC</span>(EFFs[w1], <span class="at">digits=</span><span class="dv">1</span>, <span class="at">width=</span><span class="dv">1</span>, <span class="at">format=</span><span class="st">"f"</span>),</span>
<span id="cb51-54"><a href="#cb51-54" aria-hidden="true" tabindex="-1"></a>                <span class="st">"</span></span>
<span id="cb51-55"><a href="#cb51-55" aria-hidden="true" tabindex="-1"></a><span class="st">    }</span></span>
<span id="cb51-56"><a href="#cb51-56" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span>
<span id="cb51-57"><a href="#cb51-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-58"><a href="#cb51-58" aria-hidden="true" tabindex="-1"></a><span class="st">pSets &lt;- cbind(probes[first,], probes[second,])</span></span>
<span id="cb51-59"><a href="#cb51-59" aria-hidden="true" tabindex="-1"></a><span class="st">ns &lt;- names(pSets)</span></span>
<span id="cb51-60"><a href="#cb51-60" aria-hidden="true" tabindex="-1"></a><span class="st">ns[1:10] &lt;- paste(ns[1:10], "</span>one<span class="st">", sep="</span>_<span class="st">")</span></span>
<span id="cb51-61"><a href="#cb51-61" aria-hidden="true" tabindex="-1"></a><span class="st">ns[11:20] &lt;- paste(ns[11:20], "</span>two<span class="st">", sep="</span>_<span class="st">")</span></span>
<span id="cb51-62"><a href="#cb51-62" aria-hidden="true" tabindex="-1"></a><span class="st">names(pSets) &lt;- ns</span></span>
<span id="cb51-63"><a href="#cb51-63" aria-hidden="true" tabindex="-1"></a><span class="st">pSets$score_set &lt;- scores</span></span>
<span id="cb51-64"><a href="#cb51-64" aria-hidden="true" tabindex="-1"></a><span class="st">pSets$mismatches_set &lt;- mismatches</span></span>
<span id="cb51-65"><a href="#cb51-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-66"><a href="#cb51-66" aria-hidden="true" tabindex="-1"></a><span class="st">pSets[1, -which(names(pSets) </span></span></code></pre></div>
<pre><code>    identifier_one start_one start_aligned_one permutations_one
653   Sphingopyxis       686               844                2
       score_one                probe_one.1                probe_one.2
653 -50.1945.... CCCGGACAGCTAGTTATCATCGTTTA CCCGGACAGATAGTTATCATCGTTTA
    probe_one.3 probe_one.4 efficiency_one.1 efficiency_one.2
653        &lt;NA&gt;        &lt;NA&gt;        0.8086710        0.6370537
    efficiency_one.3 efficiency_one.4 FAm_one.1 FAm_one.2 FAm_one.3
653               NA               NA  45.12303  38.95117        NA
    FAm_one.4 coverage_one.1 coverage_one.2 coverage_one.3
653        NA            0.8            0.2             NA
    coverage_one.4 identifier_two start_two start_aligned_two
653             NA   Sphingopyxis      1195              1376
    permutations_two    score_two         probe_two.1 probe_two.2
653                1 -44.1990.... CCGCCTTCATGCTCTCGAG        &lt;NA&gt;
    probe_two.3 probe_two.4 efficiency_two.1 efficiency_two.2
653        &lt;NA&gt;        &lt;NA&gt;        0.6540405               NA
    efficiency_two.3 efficiency_two.4 FAm_two.1 FAm_two.2 FAm_two.3
653               NA               NA  40.71256        NA        NA
    FAm_two.4 coverage_two.1 coverage_two.2 coverage_two.3
653        NA              1             NA             NA
    coverage_two.4 score_set
653             NA 0.1546974
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     mismatches_set
653 Porphyrobacter (-4.4
unclassified_Sphingomonadaceae (0.0
ref_ref_Altererythrobacter (5.7
ref_ref_Erythrobacter (5.7
(5.7
ref_ref_Sandarakinorhabdus (-3.9
ref_ref_Sphingomonas (5.7
(-5.2</code></pre>
<p>There are now multiple candidate dual probe sets for consideration,
all of which target variable regions in the alignment. Notice that the
top dual probe set does not include the best scoring single probe. Using
two probes in conjunction was sufficient to eliminate many non-targets,
which are listed in the column <em>mismatches_set</em>. However, even
using two probes several non-targets still remain, and careful
consideration should be given to which probe set has the least liklihood
of causing a false positive in the sample. For example, several of the
potential non-targets are rarely found in drinking water and pose little
threat of causing a false positive identification for this sample.</p>
</div>
<div class="section level3">
<h3 id="visualizing-the-target-sites">Visualizing the Target Sites<a class="anchor" aria-label="anchor" href="#visualizing-the-target-sites"></a>
</h3>
<p>Often it is useful to visualize the target sites where the probes
will hybridize on both the target and non-targets side-by-side. We can
accomplish this by highlighting the region of both probes in the
sequence set with <code>BrowseSeqs</code>. After querying the database
for the sequences of interest, we can selectively color the target site
regions given by <code>start_aligned_one</code> and
<code>start_aligned_two</code> outputs of <code>DesignProbes</code>.
Next we name and order the sequences in the set such that the target
(<em>Sphingopyxis</em>) appears at the top.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span>, nameBy<span class="op">=</span><span class="st">"identifier"</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># move the target group to the top of the sequence set</span></span>
<span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dna</span><span class="op">)</span><span class="op">==</span><span class="st">"Sphingopyxis"</span><span class="op">)</span></span>
<span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">dna</span><span class="op">[</span><span class="va">w</span><span class="op">]</span>, <span class="va">dna</span><span class="op">[</span><span class="op">-</span><span class="va">w</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">dna</span>, colorPatterns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">pSets</span><span class="op">$</span><span class="va">start_aligned_one</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    <span class="va">pSets</span><span class="op">$</span><span class="va">start_aligned_one</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">pSets</span><span class="op">$</span><span class="va">probe_one</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>,</span>
<span>    <span class="va">pSets</span><span class="op">$</span><span class="va">start_aligned_two</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    <span class="va">pSets</span><span class="op">$</span><span class="va">start_aligned_two</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">pSets</span><span class="op">$</span><span class="va">probe_two</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    highlight<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Examining the output, it is clear why these target sites were chosen
for the dual probes: at least one of the dual probes has a number of
mismatches to each of the non-target groups.</p>
</div>
<div class="section level3">
<h3 id="finishing-up">Finishing Up<a class="anchor" aria-label="anchor" href="#finishing-up"></a>
</h3>
<p>Finally, we can order the first and second probe and try them out in
an experiment! The probes should be labeled with different fluorphores
(e.g., fluorescein and Cy3). In cases where all non-targets cannot be
eliminated, it may be possible to further increase specificity by using
a competitor oligonucleotide complementary to the non-target
permutations given in the <em>mismatches</em> columns. Hybridization
conditions could be further optimized based on the results of a
formamide gradient experiment. If the two probes have markedly different
optimal formamide concentrations then a dual-hybridization approach can
be employed by washing with the more stringent condition followed by the
second.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>All of the output in this vignette was produced under the following
conditions:</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
