<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DECIPHER">
<title>Finding Chimeric Sequences • DECIPHER</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Finding Chimeric Sequences">
<meta property="og:description" content="DECIPHER">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DECIPHER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.29.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ArtOfAlignmentInR.html">The Art of Multiple Sequence Alignment in R</a>
    <a class="dropdown-item" href="../articles/ClassifySequences.html">Classify Sequences</a>
    <a class="dropdown-item" href="../articles/ClusteringSequences.html">Upsize your clustering with Clusterize</a>
    <a class="dropdown-item" href="../articles/DECIPHERing.html">Getting Started DECIPHERing</a>
    <a class="dropdown-item" href="../articles/DesignMicroarray.html">Design Microarray Probes</a>
    <a class="dropdown-item" href="../articles/DesignPrimers.html">Design Group-Specific Primers</a>
    <a class="dropdown-item" href="../articles/DesignProbes.html">Design Group-Specific FISH Probes</a>
    <a class="dropdown-item" href="../articles/DesignSignatures.html">Design Primers That Yield Group-Specific Signatures</a>
    <a class="dropdown-item" href="../articles/FindChimeras.html">Finding Chimeric Sequences</a>
    <a class="dropdown-item" href="../articles/FindingGenes.html">The Magic of Gene Finding</a>
    <a class="dropdown-item" href="../articles/FindingNonCodingRNAs.html">The Double Life of RNA: Uncovering Non-Coding RNAs</a>
    <a class="dropdown-item" href="../articles/GrowingTrees.html">Growing phylogenetic trees with TreeLine</a>
    <a class="dropdown-item" href="../articles/RepeatRepeat.html">Repeat, Repeat!</a>
    <a class="dropdown-item" href="../articles/normal.html">normal</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Finding Chimeric Sequences</h1>
                        <h4 data-toc-skip class="author">Erik S.
Wright</h4>
            
            <h4 data-toc-skip class="date">2023-08-24</h4>
      
      
      <div class="d-none name"><code>FindChimeras.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This document illustrates how to detect chimeric sequences using the
<strong>DECIPHER</strong> package through the use of the
<code>FindChimeras</code> function. The function finds chimeric
sequences present in a database (<code>dbFile</code>) of sequences by
making use of a reference database (<code>dbFileReference</code>) of
good sequences. The examples in this document are directed towards
finding chimeric 16S rRNA sequences, although a similar strategy could
be used with any type of sequence. For 16S sequences the functionality
of <code>FindChimeras</code> is implemented as a web tool available at
<a href="http://DECIPHER.codes" class="external-link uri">http://DECIPHER.codes</a>.</p>
</div>
<div class="section level2">
<h2 id="chimera-finding-problems">Chimera Finding Problems<a class="anchor" aria-label="anchor" href="#chimera-finding-problems"></a>
</h2>
<p>Chimeras are a common type of sequence anomaly that result from the
use of PCR amplification. Chimeras masquerade as normal sequences, but
actually are made up of several different sequence parts that do not
exist together in reality. Undetected chimeras can have insidious
effects, such as causing incorrect assessments of sequence diversity or
primer/probe specificity. Chimeras are very difficult to detect, and
several detection methods have been proposed with varying degrees of
success. The basic problem can be described as the search for one or
more regions of a sequence that do not belong with the rest of the
sequence.</p>
<p>The <code>FindChimeras</code> algorithm works by finding suspect
sequence fragments that are uncommon in the group where the sequence
belongs, but very common in another group where the sequence does not
belong. Each sequence in the <code>dbFile</code> is tiled into short
sequence segments called fragments. If the fragments are infrequent in
their respective group in the <code>dbFileReference</code> then they are
considered suspect. If enough suspect fragments from a sequence meet the
specified constraints then the sequence is flagged as a chimera.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>To get started we need to load the <strong>DECIPHER</strong> package,
which automatically loads several other required packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DECIPHER</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: Biostrings</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     findMatches</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: XVector</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biostrings'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     strsplit</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: RSQLite</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: parallel</span></span></code></pre>
<p>Help for any <strong>DECIPHER</strong> function, such as
<code>FindChimeras</code>, can be accessed through:</p>
<pre><code>&gt; ? FindChimeras</code></pre>
<p>If <strong>DECIPHER</strong> is installed on your system, the code in
each example can be obtained via:</p>
<pre><code>&gt; browseVignettes("DECIPHER")</code></pre>
</div>
<div class="section level2">
<h2 id="obtainingbuilding-a-reference-database">Obtaining/Building A Reference Database<a class="anchor" aria-label="anchor" href="#obtainingbuilding-a-reference-database"></a>
</h2>
<p>The first step necessary to use <code>FindChimeras</code> is to
obtain a reference database (<code>dbFileReference</code>) of good
sequences. In order for the algorithm to work accurately, it is crucial
that the reference database is free of any chimeras. The examples given
in this document use 16S rRNA sequences, and a suitable reference
database can be found at <a href="http://DECIPHER.codes/Download.html" class="external-link uri">http://DECIPHER.codes/Download.html</a>.</p>
<p>If a reference database is not present then it is feasible to create
one if a significantly large repository of sequences is available. By
using the input database as the reference database it is possible to
remove chimeras from the database. Iteratively repeating the process can
eventually result in a clean reference database that is chimera
free.</p>
<p>The reference database is required to have several fields of
information that are not found in a newly created sequence database:</p>
<ol style="list-style-type: decimal">
<li>bases and nonbases as provided by the function
<code>IdLengths</code>.</li>
<li>origin as provided by the function <code>FormGroups</code>. The
origin column contains the taxonomic rank above the level shared by the
whole group. For example, the origin of the group <em>Nanoarchaeum</em>
would be <em>Root; Archaea; Nanoarchaeota</em>. Without the taxonomic
classifications it would be possible to substitute numbers. For example,
the rank 3.1.4.2 with a group named 4 that included 4.1 and 4.2 would
have origin 3.1. These numbers could be generated using
<code>Clusterize</code> or <code>TreeLine</code> with multiple
<code>cutoffs</code>.</li>
</ol>
<p>Below is an example of building a reference database <em>de novo</em>
from a FASTA file of ribosomal RNA (rRNA) sequences downloaded from
SILVA (<a href="http://www.arb-silva.de/" class="external-link uri">http://www.arb-silva.de/</a>). Of course, the reference
database must match the type of the query sequences, so in the remainder
of this vignette we use a comprehensive SSU rRNA (16S) database. The
information in this section is only given as an example of how to build
a reference database from scratch.</p>
<pre><code>&gt; dbRef &lt;- dbConnect(SQLite(), "&lt;&lt;path to reference database&gt;&gt;")
&gt; Seqs2DB("&lt;&lt;path to reference sequences&gt;&gt;", "FASTA", dbRef, "")
645151 total sequences in table Seqs.
Time difference of 172.81 secs</code></pre>
<p>When importing a GenBank file the rank (classification) information
is automatically imported from the ORGANISM field. The classification is
simply each level of taxonomic rank separated by semicolons. In this
case the classification must be parsed from each sequence’s description
information following the “&gt;” symbol.</p>
<pre><code>&gt; ns &lt;- dbGetQuery(dbRef, "select description from Seqs")
&gt; ns$rank &lt;- gsub("(.+) (.+)?", "\\2", ns$description)
&gt; ns$rank &lt;- gsub("(.+)?;(.*)$", "\\2\nRoot;\\1", ns$rank)
&gt; ns$description &lt;- gsub("(.+) (.+)?", "\\1", ns$description)
&gt; Add2DB(ns, dbRef)
Expression:
update Seqs set description = :description where row_names = :row_names

Expression:
alter table Seqs add column rank TEXT

Expression:
update Seqs set rank = :rank where row_names = :row_names

Added to table Seqs:  "description" and "rank".

Time difference of 883.32 secs</code></pre>
<p>It is necessary to perform two more steps before we can use the
reference database for chimera finding:</p>
<pre><code>&gt; lengths &lt;- IdLengths(dbRef, add2tbl=TRUE)
  |===============================================================| 100%
Lengths counted for 645151 sequences.
Added to Seqs:  "bases", "nonbases", and "width".

Time difference of 3924.83 secs

&gt; groups &lt;- FormGroups(dbRef, add2tbl=TRUE)
  |===============================================================| 100%
Updating column: "identifier"...
Updating column: "origin"...
Formed 1760 distinct groups.
Added to table Seqs: "identifier", "origin".

Time difference of 19860.67 secs</code></pre>
<p>We can now screen the reference database for chimeras. This process
can be repeated until no new chimeras can be found in the reference
database, at which point the reference database will be ready for
checking new sequences.</p>
<pre><code>&gt; chimeras &lt;- FindChimeras(dbRef, dbFileReference=dbRef, add2tbl=TRUE)
Found 102 possible chimeras.
Added to table Seqs:  "chimera".
Time difference of 13387.34 secs</code></pre>
<p>Note that sequences in the reference database must be in the same
orientation as the query sequences. The function
<code>OrientNucleotides</code> can be used to reorient the query
sequences, if necessary.</p>
</div>
<div class="section level2">
<h2 id="steps-to-find-chimeras">Steps to Find Chimeras<a class="anchor" aria-label="anchor" href="#steps-to-find-chimeras"></a>
</h2>
<div class="section level3">
<h3 id="assigning-sequences-to-reference-groups">Assigning Sequences to Reference Groups<a class="anchor" aria-label="anchor" href="#assigning-sequences-to-reference-groups"></a>
</h3>
<div class="section level4">
<h4 id="training-the-classifier">Training the Classifier<a class="anchor" aria-label="anchor" href="#training-the-classifier"></a>
</h4>
<p>In order to assign new sequences to a group in the reference
database, it is necessary to first classify the sequences. We can use
the functions <code>LearnTaxa</code> and <code>IdTaxa</code> to assign
groups to the query sequences. See the “Classify Sequences” vignette for
a more detailed description of this process.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">dbRef</span>, <span class="st">"select identifier, origin from Seqs"</span><span class="op">)</span></span>
<span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">ids</span><span class="op">$</span><span class="va">origin</span>, <span class="va">ids</span><span class="op">$</span><span class="va">identifier</span>, sep<span class="op">=</span><span class="st">";"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbRef</span>, type<span class="op">=</span><span class="st">"DNAStringSet"</span>, remove<span class="op">=</span><span class="st">"all"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">trainingSet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LearnTaxa.html">LearnTaxa</a></span><span class="op">(</span><span class="va">dna</span>, <span class="va">ids</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="creating-a-query-sequence-database">Creating a Query Sequence Database<a class="anchor" aria-label="anchor" href="#creating-a-query-sequence-database"></a>
</h4>
<p>Chimera finding requires two database tables, one for the reference
sequences and one for the query sequences. Now that we have finished
creating the chimera-free reference sequence database and training a
classifer to its taxonomy, it is not time to make the query sequence
database. This can be accomplished by specifying a new table within the
reference sequence database or creating a new database file for the
query sequences, as shown here.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">'&lt;&lt;path to query database&gt;&gt;'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="st">'&lt;&lt;path to query sequences&gt;&gt;'</span>, <span class="st">'FASTA'</span>, <span class="va">dbConn</span>, <span class="st">''</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="assigning-to-groups">Assigning to Groups<a class="anchor" aria-label="anchor" href="#assigning-to-groups"></a>
</h4>
<p>We can now classify each of the query sequences and assign them to
the corresponding group in the reference database. Note that the query
sequences need to be in the same orientation as sequences in the
reference sequence database.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span>, remove<span class="op">=</span><span class="st">"all"</span><span class="op">)</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IdTaxa.html">IdTaxa</a></span><span class="op">(</span><span class="va">query</span>, <span class="va">trainingSet</span>, strand<span class="op">=</span><span class="st">"top"</span>, threshold<span class="op">=</span><span class="fl">0</span>, processors<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">groups</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">taxon</span>, n<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>identifier<span class="op">=</span><span class="va">groups</span>, row.names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">groups</span><span class="op">)</span>, stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Add2DB.html">Add2DB</a></span><span class="op">(</span><span class="va">groups</span>, <span class="va">dbConn</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="finding-chimeras">Finding Chimeras<a class="anchor" aria-label="anchor" href="#finding-chimeras"></a>
</h3>
<p>Finally, we can use the <code>FindChimeras</code> function to search
for chimeras in our sequences.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># full-length sequences</span></span>
<span><span class="va">chimeras</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindChimeras.html">FindChimeras</a></span><span class="op">(</span><span class="va">dbConn</span>, dbFileReference<span class="op">=</span><span class="va">dbRef</span>, add2tbl<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code>Group of Sequence (Number of Sequences):
  |===============================================================| 100%

Searching Other Groups:
  |===============================================================| 100%

Found 19 possible chimeras.
Added to table Seqs:  "chimera".
Time difference of 981.21 secs</code></pre>
<p>If any chimeras were found then they are stored in the
<code>chimeras data.frame</code>. There are several ways to determine
which sequences were found to be chimeric.</p>
<pre><code>&gt; # outputs the number of chimeras that were found
&gt; cat("Number of chimeras =", dim(chimeras)[1])
Number of chimeras = 19 
&gt; # get the indices of chimeric sequences in the set
&gt; as.numeric(row.names(chimeras))
 [1] 147 89  87  90  88  132 139 97  5  120 123 13 
[13] 118 114 173 174 175 18 124
&gt; # allows viewing the chimera results in a web browser
&gt; BrowseDB(dbConn)
[1] TRUE
&gt; # returns the FASTA description of the chimeric sequences
&gt; dbGetQuery(dbConn, "select description from Seqs where chimera is not null")
1       uncultured bacterium; Fin_CL-100633_OTU-22.
2       uncultured bacterium; Fin_CL-030731_OTU-12.
3        uncultured bacterium; Fin_CL-050647_OTU-9.
4       uncultured bacterium; Pro_CL-100643_OTU-10.
5        uncultured bacterium; LO1_CL-100643_OTU-3.
6        uncultured bacterium; LO3_CL-100641_OTU-3.
7       uncultured bacterium; UWH_CL-010711_OTU-13.
8   uncultured bacterium; Chlplus_CL-120529_OTU-35.
9       uncultured bacterium; UWH_CL-110523_OTU-35.
10      uncultured bacterium; UWH_CL-100636_OTU-11.
11 uncultured bacterium; Chlminus_CL-090519_OTU-14.
12      uncultured bacterium; UWL_CL-080514_OTU-34.
13      uncultured bacterium; UWL_CL-080545_OTU-12.
14       uncultured bacterium; Pro_CL-100633_OTU-7.
15      uncultured bacterium; UWL_CL-110645_OTU-15.
16        uncultured bacterium; UWL_CL-09061_OTU-5.
17       uncultured bacterium; Fin_CL-03079_OTU-21.
18      uncultured bacterium; UWL_CL-110518_OTU-11.
19      uncultured bacterium; UWL_CL-110548_OTU-32.</code></pre>
<pre><code>&gt; dbDisconnect(dbRef)
[1] TRUE
&gt; dbDisconnect(dbConn)
[1] TRUE</code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Erik Wright.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
